---
output: 
    html_document:
        css: style.css
chunk_output_type: console
header-includes:
- \pagenumbering{gobble}
params:
    set_title: "Article figures"
title: "`r params$set_title`"
---

```{r setup, include = FALSE, echo = FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, autodep = knitr::dep_prev())

source(here::here('r/prerequisites.R'))

message('Loading required datasets, please wait ...')
## load all required input files here
chrm_annotated_data <- fread(here('data/original_data/chrM_annotated.txt'))
gene_info_data <- fread(here('data/original_data/gene_info.txt'))
homopolymer_data <- fread(here('data/original_data/chrM_homopolymers.txt'))

## precalculated data from indel/SNV/tRNA hotspots tests and sample-level mtDNA status annotations
hotspots_indel_data <- fread(here('data/processed_data/hotspots_indel_precalculated.txt'))
hotspots_snv_data <- fread(here('data/processed_data/hotspots_snv_precalculated.txt'))
hotspots_trna_data <- fread(here('data/processed_data/hotspots_trna_precalculated.txt'))
hotspots_trna_pcawg_data <- fread(here('data/processed_data/hotspots_trna_pcawg_precalculated.txt'))
tcga_sample_hotspot_status_data <- fread(here('data/processed_data/sample_hotspot_status_tcga_precalculated.txt'))
impact_sample_hotspot_status_data <- fread(here('data/processed_data/sample_hotspot_status_impact_precalculated.txt'))

## precalculated results from rnaseq analysis
rnaseq_r25q_gsea_data <- fread(here('data/processed_data/rnaseq_r25q_gsea_hallmark_precalculated.txt'))
rnaseq_truncating_gsea_data <- fread(here('data/processed_data/rnaseq_truncating_gsea_hallmark_precalculated.txt'))
rnaseq_truncating_histogram_data <- fread(here('data/processed_data/rnaseq_truncating_histogram_precalculated.txt'))
rnaseq_vus_gsea_data <- fread(here('data/processed_data/rnaseq_vus_gsea_hallmark_precalculated.txt'))
rnaseq_vus_histogram_data <- fread(here('data/processed_data/rnaseq_vus_histogram_precalculated.txt'))
rnaseq_truncating_gsea_lowvaf_data <- fread(here('data/processed_data/rnaseq_truncating_lowvaf_gsea_hallmark_precalculated.txt'))
rnaseq_truncating_gsea_highvaf_data <- fread(here('data/processed_data/rnaseq_truncating_highvaf_gsea_hallmark_precalculated.txt'))

## generated from r/do/get_frac_callable.R 
tcga_frac_callable_data <- fread(here('data/processed_data/frac_callable_precalculated.txt'))

## TCGA (provisional) colorectal clinical data
colorectal_sample_clin_data <- fread(here('data/ext/colorectal_bcr_clinical_data_sample.txt.gz'),skip=4,header=T)
colorectal_patient_clin_data <- fread(here('data/ext/colorectal_bcr_clinical_data_patient.txt.gz'),skip=4,header=T)

# https://pubmed.ncbi.nlm.nih.gov/29316426/
pangi_clin_data <- fread(here('data/ext/pangi_clinical_data_msi.txt'))

# https://mitimpact.css-mendel.it/
mitimpact_data <- fread(here('data/ext/MitImpact_db_2.9.txt')) 

# https://www.mitomap.org/MITOMAP/MitoTipInfo
mitotip_data <- fread(here('data/ext/MitoTIP_August2017.txt'))

# https://www.synapse.org/#!Synapse:syn1713813
tcga_tmb_data <- fread(here('data/ext/per_sample_mutation_rates.tsv')) 

# https://gdc.cancer.gov/about-data/publications/pancanatlas
tcga_cdr_supptab1 <- fread(here('data/ext/TCGA-CDR-SupplementalTableS1.txt'))

# https://gdc.cancer.gov/about-data/publications/pancanatlas
tcga_absolute <- fread(here('data/ext/TCGA_mastercalls.abs_tables_JSedit.fixed.txt'))

# https://www.synapse.org/#!Synapse:syn4978511
tcga_cms <- fread(here('data/ext/cms_labels_public_all.txt'))

# https://gdc.cancer.gov/about-data/publications/pancanatlas
tcga_pancan_clin <- fread(here('data/ext/clinical_PANCAN_patient_with_followup.tsv'))

# http://gdac.broadinstitute.org/runs/stddata__latest/samples_report/GBMLGG_Notifications.html
tcga_tx_history <- fread(here('data/ext/tcga_tx_history.txt'))

# https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/center-codes
tcga_center_codes <- fread(here('data/ext/tcga_center_codes.tsv'))

# mutation and clinical data from PCAWG (NB: PCAWG mutations include VAF as low as 1%, and control-region genes)
pcawg_clin_data <- fread(here('data/original_data/data_samples_pcawg.txt.gz'))
pcawg_maf_data <- fread(here('data/original_data/data_mutations_pcawg.txt.gz'))

# non-cancer health control mtDNA mutation data from HelixMTdb 
helix_maf_data <- fread(here('data/original_data/data_mutations_helix.txt.gz'))

# TCGA somatic mutations (pan-cancer) among 468 MSK-IMPACT panel cancer genes
mc3_maf_data <- fread(here('data/original_data/data_mutations_tcga_cancergenes.txt.gz'))

# TCGA mtDNA mutation calls (somatic, unknown-tuncating), and associated sample info
tcga_clin_data <- fread(here('data/original_data/data_samples_tcga.txt.gz'))
tcga_maf_data <- fread(here('data/original_data/data_mutations_tcga.txt.gz'))

# somatic and truncating variants detected in TCGA Normal samples 
tcga_maf_normaltrunc_data <- fread(here('data/original_data/data_mutations_tcga_normal_truncating.txt'))

# MSK-IMPACT mutation calls (somatic, truncating, and likely-somatic) and sample info
impact_maf_data <- fread(here('data/original_data/data_mutations_impact.txt.gz'))
impact_clin_data <- fread(here('data/original_data/data_samples_impact.txt.gz'))

# Nuclear driver mutations in MSK-IMPACT samples
impact_maf_nuclear_data <- fread(here('data/original_data/data_mutations_impact_cancergenes_oncokb.txt.gz'))

# clinical patient/sample clinical data from MSKCC CRC tumors (https://pubmed.ncbi.nlm.nih.gov/29316426/) 
crc_msk_2017_sample_clin_data <- fread(here('data/ext/crc_msk_2017_clinical_sample.txt.gz'),skip=4,header=T)
crc_msk_2017_patient_clin_data <- fread(here('data/ext/crc_msk_2017_clinical_patient.txt.gz'),skip=4,header=T)

# more recent data pull of MSK-IMPACT patient survival data than MSKCC CRC 2017 paper
patient_clin_data_impact <- fread(here('data/ext/impact_data_clinical_patient.txt.gz'),skip=4,header=T)

# IMPACT CRC sample tumor purity
impact_colorectal_purity <- fread(here('data/ext/impact_colorectal_purity.txt'))

# tumor/normal sample mtDNA coverage data
message('Loading mtDNA coverage matrices ...')
cov_t <- fread(here('data/original_data/coverage_matrix_tcga_tumor_only.txt.gz'),header=T)
cov_n <- fread(here('data/original_data/coverage_matrix_tcga_normal_only.txt.gz'),header=T)
cov_tn <- fread(here('data/original_data/coverage_matrix_tcga.txt.gz'),header=T)
cov20_t <- fread(here('data/original_data/coverage_matrix_tcga_tumor_only20.txt.gz'),header=T)
cov20_n <- fread(here('data/original_data/coverage_matrix_tcga_normal_only20.txt.gz'),header=T)
cov20_tn <- fread(here('data/original_data/coverage_matrix_tcga20.txt.gz'),header=T)
cov_t_impact <- fread(here('data/original_data/coverage_matrix_impact_tumor_only.txt.gz'))

# sample effective gene lengths (length per gene in each sample with sufficient coverage to call variants)
tcga_sample_gene_length_data <- fread(here('data/processed_data/tcga_sample_gene_lengths_precalculated.txt.gz'))
tcga_sample_gene_length20_data <- fread(here('data/processed_data/tcga_sample_gene_lengths20_precalculated.txt.gz'))
tcga_samples_callable_per_pos_data <- fread(here('data/processed_data/tcga_samples_callable_per_position_precalculated.txt'))

impact468region <- sum(gene_info_data$cds_length)
get_pdfs <- F
message('Done.')

```

***

# Figure 1

## 1b
```{r panel_1b, fig.height=5, fig.width=2.75, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# for any individual position, what % of samples
# on average are available to sequence it?
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

tcga_clin <- copy(tcga_clin_data)
pcawg_clin <- copy(pcawg_clin_data)
m <- data.table(pos=1:16569)
d <- copy(tcga_samples_callable_per_pos_data)
dt <- d[,c('pos','callable_t'),with=F]
setnames(dt,'callable_t','N')
dtn <- d[,c('pos','callable_tn'),with=F]
setnames(dtn,'callable_tn','N')

mu_t <- mean(dt$N,na.rm=T)
s_t <- sd(dt$N,na.rm=T)
coverage_rate_t <- paste0('mean=',prettyNum(mu_t,digits=2),'; mean+/-SD: ', prettyNum(mu_t-s_t,digits=2),'-',prettyNum(mu_t+s_t,digits=2))
mu_tn <- mean(dtn$N,na.rm=T)
s_tn <- sd(dtn$N,na.rm=T)
coverage_rate_tn <- paste0('mean=',prettyNum(mu_tn,digits=2),'; mean+/-SD: ',  prettyNum(mu_tn-s_tn,digits=2),'-',prettyNum(mu_tn+s_tn,digits=2))
mu_p <- nrow(pcawg_clin)
s_p <- 0

tmp <- data.table(dataset=c('TCGA\n(tumor only)','TCGA\n(tumor+normal)','PCAWG'), mu=c(mu_t,mu_tn,mu_p),s=c(s_t, s_tn, s_p))
tmp$lwr <- tmp$mu - tmp$s
tmp$upr <- tmp$mu + tmp$s
tmp$dataset <- factor(tmp$dataset, levels=c('TCGA\n(tumor only)','TCGA\n(tumor+normal)','PCAWG'))
tmp$Variants <- c('Frame-shift indels, nonsense mutations','Missense, rRNA, tRNA','(validation)')

cols <- c('#6aa6c6','#d2e8f0','#bfbfbf')
names(cols) <- tmp$Variants

p <- ggplot(tmp, aes(x=dataset,y=mu)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity',aes(fill=Variants)) + 
    scale_fill_manual(values=cols,name='Mutations called') +
    geom_errorbar(aes(min=lwr,max=upr),width=NA) +
    labs(x=NULL,y='N samples with 5+ reads\n(Averaged across each position)') +
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))
p <- extract_gglegend(p)
panel_1b <- plot_grid(p$plot, p$legend, ncol=1, nrow=2, rel_heights=c(3,1))
panel_1b
if(get_pdfs==T) ggsave(here('figures/panel_1b.pdf'),width=2.75,height=5)
```


## 1c
```{r panel_1c, fig.height=2.5, fig.width=2.4, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# What % of variants in WXS TCGA are also in WGS (PCAWG)? 
# also in RNA-Seq?
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

dp <- copy(pcawg_maf_data[TumorVAF >= 0.05 & Hugo_Symbol != 'Control_Region',])
dp <- dp[,c('Tumor_Sample_Barcode','ShortVariantID'),with=F]
pcawg_clin <- copy(pcawg_clin_data)

dp <- dp[grepl('TCGA-',Tumor_Sample_Barcode),]
setnames(dp,'Tumor_Sample_Barcode','PATIENT_ID')
dp <- merge(dp, pcawg_clin, by='PATIENT_ID',all.x=T)
dp$id <- paste(dp$Tumor_Sample_Barcode,dp$ShortVariantID)

dt <- copy(tcga_maf_data)
dt <- dt[,c('Tumor_Sample_Barcode','ShortVariantID','class','Variant_Classification','Variant_Type','TumorVAF','TumorVAF_rna','t_alt_count','t_depth','t_depth_rna'),with=F]
dt$pcawg_available <- dt$Tumor_Sample_Barcode %in% dp$Tumor_Sample_Barcode
dt$rna_available <- dt$t_depth_rna >= 5 & !is.na(dt$t_depth_rna)

dt <- dt[pcawg_available | rna_available,]
dt$id <- paste(dt$Tumor_Sample_Barcode,dt$ShortVariantID)

dt$in_pcawg <- 'n/a'
dt[id %in% dp$id & pcawg_available==T,in_pcawg:='Yes']
dt[id %nin% dp$id & pcawg_available==T,in_pcawg:='No']

dt$in_rna <- 'n/a'
dt[TumorVAF_rna > 0.05 & rna_available==T,in_rna:='Yes']
dt[TumorVAF_rna <= 0.05 & rna_available==T,in_rna:='No']

t1 <- table.freq(dt$in_pcawg[dt$pcawg_available==T])
t1$prop <- t1$N/sum(t1$N)
t1$label <- paste0(prettyNum(t1$prop*100,digits=3),'% (N=',t1$N,')')
t1$label[2] <- NA
t1$group <- 'WGS\n(PCAWG)'

t2 <- table.freq(dt$in_rna[dt$rna_available==T])
t2$prop <- t2$N/sum(t2$N)
t2$label <- paste0(prettyNum(t2$prop*100,digits=3),'% (N=',t2$N,')')
t2$label[2] <- NA
t2$group <- 'RNA-Seq\n(TCGA)'
x <- rbind(t1,t2)
x$prop <- 100*x$prop

cols <- c('#d2e8f0','#3f99bc')
names(cols) <- c('Yes','No')
panel_1c <- ggplot(x,aes(x=group,y=prop)) +
    scale_y_continuous(expand=c(0,0)) +
    geom_bar(stat='identity',aes(fill=value)) +
    scale_fill_manual(values=cols,name='Mutations\nvalidated:') +
    geom_text(aes(label=label),y=50,angle=90,size=3.5) +
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),legend.position='right') + 
    labs(x=NULL,y='% of mtDNA variants')
panel_1c
if(get_pdfs) ggsave(here('figures/panel_1c.pdf'),width=2.4,height=2.5)
```


## 1d
```{r panel_1d, fig.width=2.75, fig.height=2.25, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# panel: correlation between DNA/RNA heteroplasmy overall
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

d <- copy(tcga_maf_data)
d <- d[t_depth >= 30 & t_depth_rna >= 30,]
dd <- d[,c('TumorVAF','TumorVAF_rna'),with=F]
dd$Heteroplasmy_DNA <- dd$TumorVAF * 100
dd$Heteroplasmy_RNA <- dd$TumorVAF_rna * 100
pearson_r <- cor(dd$Heteroplasmy_DNA,dd$Heteroplasmy_RNA,method='pearson')
statistic <- function(r,n) {
    t <- (r * sqrt(n-2))/sqrt(1-r^2)
    pval <- 2*pt(t,df=n-2,lower.tail=F)  
    list(pval=pval, statistic=t)
}
panel_1d <- ggplot(dd,aes(x=Heteroplasmy_DNA,y=Heteroplasmy_RNA)) +
    scale_x_continuous(limits=c(0,100),breaks=seq(0,100,25)) +
    scale_y_continuous(limits=c(0,100),breaks=seq(0,100,25)) +
    geom_point(alpha=0.3,color='steelblue',pch=19) +
    geom_abline(slope=1,intercept=0,color='black') +
    theme_std(base_size=12) +
    geom_text(data=dd[1,],label=paste('r =',round(pearson_r,3)),aes(x=80,y=7.5),size=3) +
    labs(x='% DNA heteroplasmy',y='% RNA heteroplasmy')
panel_1d
if(get_pdfs) ggsave(here('figures/panel_1d.pdf'),width=2.75,height=2.25)

```


## 1e
```{r panel_1e, fig.width=2.6, fig.height=2, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# panel: compare average nu-TMB to mt-TMB overall
# + per cancertype for SI
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

clin <- copy(tcga_clin_data)
clin$Tumor_Sample_Barcode <- clin$PATIENT_ID
mc3 <- copy(mc3_maf_data)
mc3$Tumor_Sample_Barcode <- strtrim(mc3$Tumor_Sample_Barcode,12)

## get nuclear TMB based on the number of mutations in the 
## MC3 MAF and median pan-cancer bps-covered from https://www.synapse.org/#!Synapse:syn1713813
tmb <- copy(tcga_tmb_data)
covered_region <- median(tmb$covd_bps)
tbl <- table.freq(mc3$Tumor_Sample_Barcode)
tbl$tmb <- 1e6*(tbl$N/covered_region)
setnames(tbl,'tmb','tmb_nu')
setnames(tbl,'N','N_nu')
clin <- merge(clin, tbl, by.x='Tumor_Sample_Barcode',by.y='value', all.x=T)
clin[is.na(tmb_nu),tmb_nu:=0] 

## calculate nuclear mutation rates
tbl_cancer <- table.freq(mc3$Tumor_Sample_Barcode)
tbl_cancer$tmb_cancergene <- 1e6*(tbl_cancer$N/impact468region)
names(tbl_cancer) <- c('Tumor_Sample_Barcode','N_cancergene','tmb_cancergene')

clin <- merge(clin, tbl_cancer, by='Tumor_Sample_Barcode', all.x=T)
clin[is.na(N_cancergene),N_cancergene:=0]
clin[is.na(tmb_cancergene),tmb_cancergene:=0]

## now calculate the mito-TMB based on number of mtDNA mutations and the region covered
dt <- copy(tcga_maf_data)
dt <- dt[,c('Tumor_Sample_Barcode','ShortVariantID','class','Variant_Classification','Variant_Type',
             'TumorVAF','TumorVAF_rna','rna_available','t_alt_count','t_depth','Hugo_Symbol'),with=F]
dt <- dt[class %in% c('somatic'),]
regions <- copy(chrm_annotated_data)
tbl2 <- table.freq(dt$Tumor_Sample_Barcode)
tbl2$value <- strtrim(tbl2$value,12)
setnames(tbl2,'N','N_mt')
clin <- merge(clin, tbl2, by.x='Tumor_Sample_Barcode',by.y='value', all.x=T)
clin[is.na(N_mt),N_mt:=0]
clin$tmb_mt <- 1e6*(clin$N_mt/clin$capture_area_tumornormal)

## also generate correlations per tumortype/overall
getlabel <- function(dat) {
    r=tryCatch({
        cor(dat$tmb_nu,dat$tmb_mt,method='pearson')
    },error=function(e) {
        as.numeric(NA)
    })
    rlabel <- paste0('r = ',prettyNum(r,digits=2))
    list(labels=rlabel)
}

dat <- clin[coverage_tumornormal >= 0.9,c('tmb_nu','tmb_mt','SAMPLE_TYPE','maintype','PATIENT_ID'),with=F]
correlation_n <- nrow(dat)
info <- getlabel(dat)

panel_1e <- ggplot(dat, aes(x=tmb_nu,y=tmb_mt)) +
    geom_point(pch=19,color='#6aa6c6',alpha=0.3) +
    theme_std(base_size=12) +
    labs(x='Nuclear TMB',y='mito-TMB') + 
    geom_text(data=dat[1,], x=20,y=1500,label=info$label,size=3) 
panel_1e
if(get_pdfs) ggsave(here('figures/panel_1e.pdf'),width=2.6,height=2)



## repeat for SI by cancer type
dat <- clin[coverage_tumornormal >= 0.9,]
tbl <- table.freq(dat$maintype)
types <- tbl$value[tbl$N >= 100]
dat <- dat[maintype %in% types,]
cancertype_correlation_n <- table.freq(dat$maintype)
dat$capture_area_impact468 <- impact468region
dat$capture_area_noncancer <- covered_region - impact468region
info <- dat[,getlabel(.SD),by=maintype]
dat <- merge(dat, info, by='maintype', all.x=T)
dat$maintype_label <- paste0(dat$maintype,'\n',dat$labels)

getp <- function(query.maintype,dat) {
    tmp <- dat[maintype_label==query.maintype,]
    p <- ggplot(tmp, aes(x=tmb_nu,y=tmb_mt)) +
        geom_point(pch=19,color='#6aa6c6',alpha=0.3) +
        theme_std(base_size=10) +
        labs(x='Nuclear TMB',y='mito-TMB',subtitle=query.maintype)
    p
}

plist <- lapply(sortunique(dat$maintype_label),getp, dat)
panel_EDF3e <- plot_grid(plotlist=plist,nrow=2)

```


## 1f
```{r panel_1f, fig.width=8, fig.height=5, eval=T}

## ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
## compare TMB across individual cancer genes to individual mtDNA genes
## ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

clin <- copy(tcga_clin_data)
clin <- clin[coverage_tumornormal >= 0.9]
valid_samples <- clin$Tumor_Sample_Barcode
n_samples <- length(valid_samples)

mc3 <- copy(mc3_maf_data)
mc3 <- mc3[Tumor_Sample_Barcode %in% valid_samples,]

## Cancer genes:
## - add cancer role for nuclear genes
## - get N mutations; truncating; Non-truncating
genes <- copy(gene_info_data)
genes <- genes[,c('Hugo_Symbol','Role'),with=F]
genes[Role %nin% c('Oncogene','TSG'),Role:='n/a']

mc3 <- merge(mc3, genes, by='Hugo_Symbol', all.x=T)

tbl_nu <- table.freq(mc3$Hugo_Symbol)
tbl_nu_trunc <- table.freq(mc3$Hugo_Symbol[mc3$Variant_Classification %in% truncating_classes])
tbl_nu_missense <- table.freq(mc3$Hugo_Symbol[mc3$Variant_Classification %in% nontruncating_classes])
tbl_nu_silent <- table.freq(mc3$Hugo_Symbol[mc3$Variant_Classification %in% 'Silent'])

## mtDNA genes
## - get N mutations (all genes); truncating (protein-coding); Missense (protein-coding)
tcga <- copy(tcga_maf_data)
tcga <- tcga[class=='somatic' & Tumor_Sample_Barcode %in% valid_samples,]
tbl_mt <- table.freq(tcga$Hugo_Symbol)
tbl_mt_trunc <- table.freq(tcga$Hugo_Symbol[tcga$Variant_Classification %in% truncating_classes])
tbl_mt_missense <- table.freq(tcga$Hugo_Symbol[tcga$Variant_Classification %in% nontruncating_classes & 
                              tcga$Variant_Classification %nin% c('rRNA','tRNA')])
tbl_mt_silent <- table.freq(tcga$Hugo_Symbol[tcga$Variant_Classification %in% 'Silent'])

## merge results
tbl <- rbind(tbl_nu,tbl_mt)
tbl_trunc <- rbind(tbl_nu_trunc,tbl_mt_trunc)

setnames(tbl_trunc,'N','N_trunc')
tbl_missense <- rbind(tbl_nu_missense,tbl_mt_missense)
setnames(tbl_missense,'N','N_missense')
tbl_silent <- rbind(tbl_nu_silent,tbl_mt_silent)
setnames(tbl_silent,'N','N_silent')

tbl <- merge(tbl, tbl_trunc, by='value', all.x=T)
names(tbl) <- c('Hugo_Symbol','N','N_trunc')
tbl <- merge(tbl, tbl_missense, by.x='Hugo_Symbol', by.y='value', all.x=T)
tbl[is.na(tbl)] <- 0

## get info for nuclear genes
genes <- copy(gene_info_data)
genes <- genes[,c('Hugo_Symbol','Role','cds_length'),with=F]
genes[Role %nin% c('Oncogene','TSG'),Role:='n/a']
names(genes) <- c('symbol','role','length')
genes$genome <- 'Nuclear'
genes$length <- n_samples*genes$length

## get length available for sequencing per mt_tumornormalDNA gene
mt <- copy(tcga_sample_gene_length_data)
mt <- mt[sample_type=='tumor+normal',]
mt[,sample_type:=NULL]
mt <- mt[Tumor_Sample_Barcode %in% valid_samples,]
mt <- colSums(mt[,2:ncol(mt),with=F])
mt <- data.table(symbol=names(mt),length=mt)
mt[grepl('MT-N',symbol),role:='Complex I']
mt[grepl('MT-CY',symbol),role:='Complex III']
mt[grepl('MT-CO',symbol),role:='Complex IV']
mt[grepl('MT-ATP',symbol),role:='Complex V']
mt[grepl('MT-R',symbol),role:='rRNA']
mt[grepl('MT-T',symbol),role:='tRNA']
mt$genome <- 'Mitochondrial'

## continue from here
genes <- rbind(genes, mt) 
setnames(genes,'length','covd_bps')
genes <- merge(genes, tbl, by.x='symbol', by.y='Hugo_Symbol', all.x=T)
genes <- genes[symbol!='HIST2H3C']
genes[role %in% c('rRNA','tRNA'),N_trunc:=NA]
genes[role %in% c('rRNA','tRNA'),N_missense:=NA]


## calculate per-gene TMBs, save SI data
genes$tmb <- 1e6*(genes$N/genes$covd_bps)
genes$tmb_trunc <- 1e6*(genes$N_trunc/genes$covd_bps)
genes$tmb_missense <- 1e6*(genes$N_missense/genes$covd_bps)

## save the data for SI table
write.tsv(genes,here('data/processed_data/gene_mutation_rates.txt'))

## also calculate high-vaf truncating rates
tbl_mt_trunc_highvaf <- table.freq(tcga$Hugo_Symbol[tcga$Variant_Classification %in% truncating_classes & tcga$TumorVAF > 0.8])
tmp <- merge(tbl_mt_trunc_highvaf, mt, by.x='value', by.y='symbol', all.x=T)
tmp$tmb <- 1e6*(tmp$N/tmp$length)


## make plot
d1 <- genes[,c('symbol','genome','tmb','role'),with=F]
d1 <- d1[order(tmb,decreasing=T),]
d1$symbol <- factor(d1$symbol, levels=d1$symbol)
d1$x <- as.integer(d1$symbol)
d1$label1 <- d1$symbol
d1$label2 <- d1$symbol
d1$label1[3:nrow(d1)] <- NA

cols <- c('#bfbfbf','#6ba5c5')
names(cols) <- c('Nuclear','Mitochondrial')
plab <- plabel(wilcox.test(d1$tmb[d1$genome=='Nuclear'],d1$tmb[d1$genome=='Mitochondrial'])$p.value)

p_overall <- ggplot(d1, aes(x=x,y=tmb)) +
    scale_x_continuous(limits=c(0,505),breaks=seq(0,500,by=250)) + 
    scale_y_continuous(limits=c(0,400),breaks=seq(0,400,by=100)) + 
    geom_segment(data=d1,aes(x=x,xend=x,y=0,yend=tmb,color=genome)) +
    scale_color_manual(values=cols,name=NULL) + 
    geom_text_repel(aes(label=label1),size=2.5) +
    geom_text(data=d1[1,],label=plab,x=300,y=50,size=3) +
    theme_std(base_size=12) +
    theme(legend.position='none') +
    labs(x=NULL,y=NULL)

mtdna_positions <- range(d1$x[d1$genome=='Mitochondrial'])
d2 <- d1[1:mtdna_positions[2],]
d2[genome!='Mitochondrial' & symbol %nin% c('TP53','KRAS','PTEN','PIK3CA','CDKN2A','NRAS','B2M','BRAF'),label2:=NA]
d2[tmb > 150,tmb:=150]
p_closeup <- ggplot(d2, aes(x=x,y=tmb)) +
    scale_x_continuous(limits=c(0,130),breaks=c(seq(0,50,by=10),58,79,129)) + 
    scale_y_continuous(limits=c(0,150),expand=c(0,0), breaks=seq(0,125,by=25)) + 
    geom_bar(stat='identity',aes(fill=genome)) + 
    geom_text(aes(label=label2),color='black',size=2,angle=45,hjust=-0.1,vjust=-0.1) +
    scale_fill_manual(values=cols,name=NULL) + 
    theme_std(base_size=12) +
    theme(legend.position='bottom') + 
          #axis.line.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank(),
          #axis.line.y=element_blank(),axis.text.y=element_blank(),axis.ticks.y=element_blank() ) +
    labs(x='mtDNA + Nuclear cancer-assoc. genes',y='Mutations/Mb') 

p_null <- ggplot() + theme_std() + theme(axis.line.x=element_blank(),axis.line.y=element_blank())
p <- plot_grid(p_null, p_null, p_null, p_overall, ncol=4)
panel_1f <- plot_grid(p, p_closeup, nrow=2, ncol=1, rel_heights=c(1,2))
panel_1f
if(get_pdfs) ggsave(here('figures/panel_1f.pdf'),width=6.4,height=4.4)

```


## 1g
```{r panel_1g, fig.height=3.5, fig.width=2.6, eval=T}

## ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
## compare truncating-TMB among mtDNA and TSGs
## ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

x <- genes[role %in% 'TSG' | genome=='Mitochondrial',c('symbol','role','tmb_trunc','genome'),with=F]
x <- x[!is.na(tmb_trunc),]
setnames(x,'tmb_trunc','tmb')

## make plot
cols <- c('#bfbfbf','#3f99bc')
names(cols) <- c('Nuclear','Mitochondrial')
yinfo <- break_axis(x$tmb,maxlower=25,lowerticksize=5,upperticksize=25,ratio_lower_to_upper=0.4)
x$y <- yinfo$newy
x$label <- x$symbol
x[(tmb < 15 & genome=='Nuclear'), label:=NA] #| (tmb < 10 & genome=='Mitochondrial'),label:=NA]
x[symbol %in% c('MT-ATP6','MT-ATP8'),label:=symbol]
x[genome=='Nuclear',genome:='Nuclear TSGs']
x[genome=='Mitochondrial',genome:='mtDNA\n(protein-coding)']
tst <- wilcox.test(x$tmb[x$genome=='mtDNA\n(protein-coding)'], x$tmb[x$genome=='Nuclear TSGs'])
plab <- plabel(tst$p.value)
x$genome <- factor(x$genome)
foldchange_truncating <- median(x$tmb[x$genome!='Nuclear TSGs'])/median(x$tmb[x$genome=='Nuclear TSGs'])
foldchange_truncating <- paste0('Fold-change=',prettyNum(foldchange_truncating,digits=2),', ',plab)

panel_1g <- ggplot(x,aes(x=genome,y=y)) +      
    scale_y_continuous(limits=c(yinfo$limits[1],37),breaks=yinfo$breaks,labels=yinfo$labels) + 
    geom_point(position=position_jitter(seed=42,width=0.2),pch=19,color='#bfbfbf') + 
    geom_boxplot(outlier.shape=NA,width=0.3,fill=NA) +
    geom_signif(comparisons=list(c(1,2)),annotations=plab,y_position=35,textsize=3,size=0.25,
                tip_length=NA) +
    geom_text_repel(aes(label=label),position=position_jitter(seed=42,width=0.2),
                    min.segment.length=0.01,size=2.5,segment.size=0.25) + 
    theme_std(base_size=12) +
    labs(x=NULL,y='Truncating/Mb') +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
panel_1g
if(get_pdfs) ggsave(here('figures/panel_1g.pdf'),width=2.6,height=3.5)
```


## 1h
```{r panel_1h, fig.height=3.5, fig.width=2.6, eval=T}

## ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
## compare missense-TMB among mtDNA and Oncogenes
## ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

x <- genes[role %in% 'Oncogene' | genome=='Mitochondrial',c('symbol','role','tmb_missense','genome'),with=F]
x <- x[!is.na(tmb_missense),]
setnames(x,'tmb_missense','tmb')

## make plot
cols <- c('#bfbfbf','#3f99bc')
names(cols) <- c('Nuclear','Mitochondrial')
yinfo <- break_axis(x$tmb,maxlower=100,lowerticksize=25,upperticksize=50,ratio_lower_to_upper=0.4)
x$y <- yinfo$newy
x$label <- x$symbol
x[(tmb < 20 & genome=='Nuclear'), label:=NA]
x[symbol %in% c('MT-ATP6','MT-ATP8'),label:=symbol]
x[genome=='Nuclear',genome:='Nuclear\nOncogenes']
x[genome=='Mitochondrial',genome:='mtDNA\n(protein-coding)']
tst <- wilcox.test(x$tmb[x$genome=='mtDNA\n(protein-coding)'], x$tmb[x$genome=='Nuclear\nOncogenes'])
plab <- plabel(tst$p.value)
x$genome <- factor(x$genome)
foldchange_missense <- median(x$tmb[x$genome!='Nuclear\nOncogenes'])/median(x$tmb[x$genome=='Nuclear\nOncogenes'])
foldchange_missense <- paste0('Fold-change=',prettyNum(foldchange_missense,digits=2),', ',plab)

panel_1h <- ggplot(x,aes(x=genome,y=y)) +      
    scale_y_continuous(limits=c(yinfo$limits[1],150),breaks=yinfo$breaks,labels=yinfo$labels) + 
    geom_point(position=position_jitter(seed=42,width=0.2),pch=19,color='#bfbfbf') + 
    geom_boxplot(outlier.shape=NA,width=0.3,fill=NA) +
    geom_signif(comparisons=list(c(1,2)),annotations=plab,y_position=140,textsize=3,size=0.25,
                tip_length=NA) +
    geom_text_repel(aes(label=label),position=position_jitter(seed=42,width=0.2),
                    min.segment.length=0.01,size=2.5,segment.size=0.25) + 
    theme_std(base_size=12) +
    labs(x=NULL,y='Missense/Mb') +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
panel_1h
if(get_pdfs) ggsave(here('figures/panel_1h.pdf'),width=2.6,height=3.5)

```


## 1i
```{r panel_1i, fig.width=1.6, fig.height=3.25, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# % samples with truncating mutations
# in tumor-only vs germline,
# compare the rate in germline with the 200k data
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## tcga data
clin <- copy(tcga_clin_data)
clin <- clin[coverage_tumor >= 0.9]
d <- copy(tcga_maf_data)
d <- d[Variant_Classification %in% truncating_classes & Tumor_Sample_Barcode %in% clin$Tumor_Sample_Barcode,]
x_somatic_trunc <- length(unique(d$Tumor_Sample_Barcode))
n_overall <- nrow(clin) 
ci_tcga_somatic <- binom.confint(x_somatic_trunc, n_overall, method='exact')
ci_tcga_somatic <- adt(ci_tcga_somatic)
ci_tcga_somatic$class <- 'TCGA Somatic'


## helix data
x_helix_trunc <- 291 ## total number of samples with truncating variants

n_helix <- 195983 ## total number of samples 
ci_helix <- binom.confint(x_helix_trunc, n_helix, method='exact') ## assumes max of 1 truncating per sample
ci_helix <- adt(ci_helix)
ci_helix$class <- '196K Normal\nsamples'


## get % affected in each
ci <- rbind(ci_tcga_somatic, ci_helix)
ci$class <- factor(ci$class, levels=rev(c('TCGA Somatic','196K Normal\nsamples')))
ci <- ci[order(class),]
ci$mean <- 100*ci$mean
ci$lower <- 100*ci$lower
ci$upper <- 100*ci$upper
setnames(ci,'mean','pct')


## include p-values:
tst <- prop.test(ci$x[1:2],ci$n[1:2])
plab <- plabel(tst$p.value)
pctinfo <- break_axis(ci$pct,maxlower=0.25,lowerticksize=0.05,minupper=8,upperticksize=1,
                       ratio_lower_to_upper=0.5)
lwrinfo <- break_axis(ci$lower,maxlower=0.25,lowerticksize=0.05,minupper=8,upperticksize=1,
                       ratio_lower_to_upper=0.5)
uprinfo <- break_axis(ci$upper,maxlower=0.25,lowerticksize=0.05,minupper=8,upperticksize=1,
                       ratio_lower_to_upper=0.5)
ci$newpct <- pctinfo$newy
ci$newlower <- lwrinfo$newy
ci$newupper <- uprinfo$newy

panel_1i <- ggplot(ci,aes(x=class,y=newpct)) + 
    scale_y_continuous(expand=c(0,0),labels=pctinfo$labels,
                       limits=c(pctinfo$limits[1],0.475),breaks=pctinfo$breaks) +
    geom_bar(stat='identity',fill='#3f99bc') +
    geom_errorbar(aes(min=newlower,max=newpct),width=NA,color='white') +
    geom_errorbar(aes(min=newpct,max=newupper),width=NA,color='black') +
    geom_signif(comparisons=list(c(1,2)),annotations=plab,textsize=2.5,size=0.25,tip_length=NA) +
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),legend.position='none') +
    labs(x=NULL,y='% cases with truncating variant')
panel_1i
if(get_pdfs) ggsave(here('figures/panel_1i.pdf'),width=1.6,height=3.25)
```


***

# Fig. 2

```{r eval=T, echo=F}
## define helper functions used in multiple panels

combinatorial_poisson_test <- function(categories,counts,times,result,upper_field,ygap) {
    compare <- function(comparisons,result,categories,counts,times) {
        #browser()
        a <- comparisons$a; b=comparisons$b
        x <- unlist(result[get(categories) %in% c(a,b),(counts),with=F])
        T <- unlist(result[get(categories) %in% c(a,b),(times),with=F])
        p.value <- tryCatch({
            poisson.test(x,T,alternative='two.sided')$p.value
        },error=function(e) {
            as.numeric(NA)
        })
        comparisons$p.value <- p.value
        comparisons
    }
    x <- as.data.table(expand.grid(a=result[[categories]],b=result[[categories]],stringsAsFactors=F))
    x$i <- 1:nrow(x)
    reorder <- function(x) {
        comparisons <- sort(c(x$a,x$b))
        x$a <- comparisons[1]
        x$b <- comparisons[2]
        x
    }
    x <- x[,reorder(.SD),by=i]
    x$id <- paste(x$a,x$b)
    x <- x[!duplicated(id),]
    x <- x[a!=b,]
    x[,id:=NULL]
    x <- x[order(a,b),]
    x$i <- 1:nrow(x)
    x <- x[,compare(.SD,result,categories,counts,times),by=i]
    x <- x[order(a,b),]
    max_y <- ceiling(max(result[[upper_field]]+ygap/4))
    x$ypos <- max_y + ygap*seq(0,(nrow(x)-1))
    x$label <- sapply(x$p.value,plabel,USE.NAMES=F)
    x
}


get_variant_rates_per_complex <- function(l_tumornormal, d, clin, min.coverage=0.9) {
    l_tumornormal <- l_tumornormal[sample_type=='tumor+normal',]
    l_tumornormal[,sample_type:=NULL]
    l_tumornormal <- as.data.table(reshape2::melt(l_tumornormal,id.vars='Tumor_Sample_Barcode'))
    names(l_tumornormal) <- c('Tumor_Sample_Barcode','Hugo_Symbol','l')
    l_tumornormal <- merge(l_tumornormal, genes, by.x='Hugo_Symbol', by.y='symbol', all.x=T)
    l_tumornormal <- l_tumornormal[!is.na(length),]
    l_tumornormal$complex <- as.character(NA)
    l_tumornormal[grepl('MT-ND',Hugo_Symbol),complex:='I']
    l_tumornormal[grepl('MT-CYB',Hugo_Symbol),complex:='III']
    l_tumornormal[grepl('MT-CO',Hugo_Symbol),complex:='IV']
    l_tumornormal[grepl('MT-ATP',Hugo_Symbol),complex:='V']
    setnames(l_tumornormal,'l','l_tumornormal')
    l <- l_tumornormal


    ## for each sample and gene, annotate if it is mutated
    d <- d[,c('Tumor_Sample_Barcode','Hugo_Symbol','Variant_Classification','class'),with=F]
    d <- d[class=='somatic',]

    x_trunc <- xtabs(~Tumor_Sample_Barcode + Hugo_Symbol,data=d[Variant_Classification %in% truncating_classes])
    x_trunc <- adt(reshape2::melt(x_trunc))
    setnames(x_trunc,'value','truncating')

    x_nontrunc <- xtabs(~Tumor_Sample_Barcode + Hugo_Symbol,data=d[Variant_Classification %in% nontruncating_classes])
    x_nontrunc <- adt(reshape2::melt(x_nontrunc))
    setnames(x_nontrunc,'value','nontruncating')

    x_silent <- xtabs(~Tumor_Sample_Barcode + Hugo_Symbol,data=d[Variant_Classification=='Silent'])
    x_silent <- adt(reshape2::melt(x_silent))
    setnames(x_silent,'value','silent')

    x <- merge(x_trunc, x_nontrunc, by=c('Tumor_Sample_Barcode','Hugo_Symbol'),all=T)
    x <- merge(x, x_silent, by=c('Tumor_Sample_Barcode','Hugo_Symbol'),all=T)

    l <- merge(l, x, by=c('Tumor_Sample_Barcode','Hugo_Symbol'), all.x=T)
    l[is.na(truncating),truncating:=0]
    l[is.na(nontruncating),nontruncating:=0]
    l[is.na(silent),silent:=0]


    ## annotate the cancer type per sample
    clin <- clin[,c('Tumor_Sample_Barcode','maintype','coverage_tumornormal'),with=F]
    l <- merge(l, clin, by='Tumor_Sample_Barcode', all.x=T)
    l <- l[!is.na(maintype),]

    ## subset for complexes in samples covered at 90%+
    l$prop_covered <- l$l_tumornormal / l$length
    dat <- l[prop_covered >= min.coverage,]
    dat[,length:=NULL]

    ## collapse across the multiple genes of each sample/complex
    collapse_sample_complex <- function(l) {
        total_truncating <- sum(l$truncating) ## across all genes
        total_nontruncating <- sum(l$nontruncating) ## across all genes    
        total_silent <- sum(l$silent) ## across all genes    
        list(total_truncating=total_truncating, total_nontruncating=total_nontruncating,total_silent=total_silent)
    }
    dat <- dat[,collapse_sample_complex(.SD),by=c('Tumor_Sample_Barcode','maintype','complex',
                                                  'l_tumornormal')]

    ## collapse the complex across all samples
    collapse_complex <- function(l) {
        total_truncating <- sum(l$total_truncating) ## across all genes
        total_nontruncating <- sum(l$total_nontruncating) ## across all genes
        total_silent <- sum(l$total_silent) ## across all genes    
        available_samples <- length(unique(l$Tumor_Sample_Barcode)) 
        total_complex_length_mb <- sum(l$l_tumornormal)/1e6
        list(total_truncating=total_truncating, total_nontruncating=total_nontruncating,
             total_silent=total_silent,
             available_samples=available_samples, 
             total_complex_length_mb=total_complex_length_mb
        )
    }
    dat2 <- dat[,collapse_complex(.SD),by=c('complex')]
    dat2 <- dat2[order(complex)]


    ## calculate the number of affected cases/Mb-available
    ci <- adt(pois.exact(dat2$total_truncating,dat2$total_complex_length_mb))
    ci$complex <- dat2$complex
    ci <- ci[,c('rate','lower','upper','complex'),with=F]
    names(ci) <- c('rate_trunc','lower_trunc','upper_trunc','complex')
    result <- merge(dat2, ci, by='complex')

    ci <- adt(pois.exact(dat2$total_nontruncating,dat2$total_complex_length_mb))
    ci$complex <- dat2$complex
    ci <- ci[,c('rate','lower','upper','complex'),with=F]
    names(ci) <- c('rate_nontrunc','lower_nontrunc','upper_nontrunc','complex')
    result <- merge(result, ci, by='complex')

    ci <- adt(pois.exact(dat2$total_silent,dat2$total_complex_length_mb))
    ci$complex <- dat2$complex
    ci <- ci[,c('rate','lower','upper','complex'),with=F]
    names(ci) <- c('rate_silent','lower_silent','upper_silent','complex')
    result <- merge(result, ci, by='complex')
    result <- result[!is.na(complex),]
    result
}


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# calculate mutaiton/Mb per complex
# - truncating
# - non-truncating
# - silent
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## for each sample, get its effective length per gene
genes <- copy(chrm_annotated_data)
genes <- table.freq(genes$symbol)
names(genes) <- c('symbol','length')

result_tcga <- get_variant_rates_per_complex(tcga_sample_gene_length_data, 
                                        tcga_maf_data, 
                                        tcga_clin_data,
                                        min.coverage=0.9)

## make equivalent sample-gene-length and coverage data for pcawg (excluding tcga)
clin <- copy(pcawg_clin_data)
valid_samples <- clin$PATIENT_ID[grepl('TCGA',clin$PATIENT_ID)==F]
chrm <- copy(chrm_annotated_data)
tbl <- table.freq(chrm$symbol)
x <- adt(expand.grid(PATIENT_ID=valid_samples,Hugo_Symbol=tbl$value))
x <- merge(x, tbl, by.x='Hugo_Symbol', by.y='value', all.x=T)
x <- reshape(x,idvar='PATIENT_ID',timevar='Hugo_Symbol',direction='wide')
setnames(x,'PATIENT_ID','Tumor_Sample_Barcode')
names(x) <- gsub('N[.]','',names(x))
x$sample_type <- 'tumor+normal'
clin <- copy(pcawg_clin_data)
clin <- clin[PATIENT_ID %in% valid_samples,]
clin$coverage_tumornormal <- 1
setnames(clin,'dcc_project_code','maintype')
mut <- copy(pcawg_maf_data[TumorVAF >= 0.05 & Hugo_Symbol != 'Control_Region',])
mut <- mut[Tumor_Sample_Barcode %in% valid_samples,]
mut$class <- 'somatic'

result_pcawg <- get_variant_rates_per_complex(x, mut, clin, min.coverage=0.9)

```

## 2a

```{r panel_2a, fig.height=6.3, fig.width=2.5, eval=T}

## plot truncating rates
x <- combinatorial_poisson_test('complex','total_truncating','total_complex_length_mb',
                                result_tcga,'upper_trunc',ygap=1)
f <- function(i,x) c(x$a[i],x$b[i])
comparisons <- lapply(1:nrow(x),f,x)

p1 <- ggplot(result_tcga,aes(x=complex,y=rate_trunc)) +
    geom_bar(stat='identity',fill='#3f99bc') +
    geom_errorbar(aes(min=lower_trunc,max=rate_trunc),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_trunc,max=upper_trunc),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x$ypos,annotations=x$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Complex',y='Truncating/Mb') +
    scale_y_continuous(expand=c(0,0), limits=c(0,25),breaks=seq(0,20,by=5)) 


## plot silent rates
x <- combinatorial_poisson_test('complex','total_silent','total_complex_length_mb',result_tcga,
                                'upper_silent',ygap=1)
f <- function(i,x) c(x$a[i],x$b[i])
comparisons <- lapply(1:nrow(x),f,x)

p2 <- ggplot(result_tcga,aes(x=complex,y=rate_silent)) +
    geom_bar(stat='identity',fill='#3f99bc') +
    geom_errorbar(aes(min=lower_silent,max=rate_silent),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_silent,max=upper_silent),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x$ypos,annotations=x$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Complex',y='Synonymous/Mb') +
    scale_y_continuous(expand=c(0,0),limits=c(0,25),breaks=seq(0,20,by=5)) 

panel_2a <- plot_grid(p1,p2,ncol=1,nrow=2)
panel_2a
if(get_pdfs) ggsave(here('figures/panel_2a.pdf'),width=2.5,height=6.3)
```


## 2b

```{r, panel_2b, fig.height=6.3, fig.width=2.5, eval=T}

## plot truncating rates
x <- combinatorial_poisson_test('complex','total_truncating','total_complex_length_mb',
                                result_pcawg,'upper_trunc',ygap=1)
f <- function(i,x) c(x$a[i],x$b[i])
comparisons <- lapply(1:nrow(x),f,x)

p1 <- ggplot(result_pcawg,aes(x=complex,y=rate_trunc)) +
    geom_bar(stat='identity',fill='#3f99bc') +
    geom_errorbar(aes(min=lower_trunc,max=rate_trunc),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_trunc,max=upper_trunc),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x$ypos,annotations=x$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Complex',y='Truncating/Mb') +
    scale_y_continuous(expand=c(0,0), limits=c(0,25),breaks=seq(0,20,by=5)) 


## plot silent rates
x <- combinatorial_poisson_test('complex','total_silent','total_complex_length_mb',result_pcawg,
                                'upper_silent',ygap=1)
f <- function(i,x) c(x$a[i],x$b[i])
comparisons <- lapply(1:nrow(x),f,x)

p2 <- ggplot(result_pcawg,aes(x=complex,y=rate_silent)) +
    geom_bar(stat='identity',fill='#3f99bc') +
    geom_errorbar(aes(min=lower_silent,max=rate_silent),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_silent,max=upper_silent),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x$ypos,annotations=x$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Complex',y='Synonymous/Mb') +
    scale_y_continuous(expand=c(0,0),limits=c(0,25),breaks=seq(0,20,by=5)) 

panel_2b <- plot_grid(p1,p2,ncol=1,nrow=2)
panel_2b
if(get_pdfs) ggsave(here('figures/panel_2b.pdf'),width=2.5,height=6.3)

```


## 2c

```{r panel_2c, fig.height=3.2, fig.width=5, eval=T}

d <- copy(tcga_maf_data)
d <- d[,c('Tumor_Sample_Barcode','Hugo_Symbol','Variant_Classification','TumorVAF','t_depth','t_alt_count'),with=F]
d[grepl('MT-ND',Hugo_Symbol),Complex:='CI']
d[grepl('MT-CY',Hugo_Symbol),Complex:='CIII']
d[grepl('MT-CO',Hugo_Symbol),Complex:='CIV']
d[grepl('MT-ATP',Hugo_Symbol),Complex:='CV']
d <- d[!is.na(Complex),]
d[Variant_Classification %in% truncating_classes,type:='Truncating']
d[Variant_Classification %in% 'Silent',type:='Silent']
dd <- d[t_depth >= 30 & !is.na(Complex) & !is.na(type),]

get_percentile <- function(dd) {
    dd <- dd[order(TumorVAF,decreasing=F)]
    dd$percentile <- (1:nrow(dd)) / nrow(dd)
    dd[,c('TumorVAF','percentile'),with=F]
}
res <- dd[,get_percentile(.SD),by=c('Complex','type')]
res$Heteroplasmy <- 100*res$TumorVAF

test_complex <- function(res) {
    x_trunc <- res$TumorVAF[res$type=='Truncating']
    x_silent <- res$TumorVAF[res$type=='Silent']
    p <- tryCatch({
        tst <- wilcox.test(x_trunc, x_silent, alternative='two.sided')
        tst$p.value
    },error=function(e) {
        as.numeric(NA)
    })
    list(p=p)
}
res2 <- res[,test_complex(.SD),by=Complex]    
res2$label <- paste0('p=',prettyNum(res2$p,digits=1))
res <- merge(res, res2[,c('Complex','label'),with=F], by='Complex', all.x=T)
res$Complex <- paste0(res$Complex,' (',res$label,')')
cols <- c('#3f99bc','#bfbfbf')
names(cols) <- c('Truncating','Silent')

panel_2c <- ggplot(res,aes(x=percentile,y=Heteroplasmy)) +
    scale_x_continuous(expand=c(0,0),limits=c(0,1),breaks=seq(0,1,by=0.25)) +
    scale_y_continuous(expand=c(0,0),limits=c(0,100),breaks=seq(0,100,by=25)) +
    geom_point(aes(color=type)) +
    scale_color_manual(values=cols,name='Variant') + 
    facet_wrap(facets=~Complex,ncol=2) +
    theme_std(base_size=12) + 
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(), panel.grid.major=element_line()) +
    labs(x='Variants (increasing order of VAF)',y='% VAF')
panel_2c
if(get_pdfs) ggsave(here('figures/panel_2c.pdf'),width=5,height=3.2)

```


## 2d,e

```{r panel_2d, fig.width=8, fig.height=4.7, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# 2D
# - % samples with truncating per cancer type
# - proportion barplot indicating the estimated prop of samples 
# 2E
# - the rate of truncating in CI vs others per type
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

get_barplot <- function(l,tcga=T) {
    setnames(l,'annotation_detailed_prefilter','label')
    min.n <- 20
    cols <- c('#e5e5e5',brewer.pal(6,'Set3'))
    l <- l[label!='Unknown']
    x <- xtabs(~maintype + truncating, data=l)
    x <- as.data.frame.matrix(x)
    out <- x
    Total <- rowSums(x)
    x <- x/Total
    x <- cbind(type=rownames(x),adt(x))
    xt <- data.table(type=x$type,N=Total)
    xt <- xt[N >= min.n,]
    x <- x[type %in% xt$type,]
    names(x) <- c('type','No truncating','Truncating')
    x <- x[order(`No truncating`),]
    x[,`No truncating`:=NULL]
    xm <- x
    xm$type <- factor(xm$type,levels=x$type)
    xt$type <- factor(xt$type,levels=x$type)
    xm$Percentage <- 100*xm$Truncating
    xm$variable <- 'Truncating'
    xm$variable <- factor(xm$variable, levels=c('Wildtype','Silent','Missense','tRNA','rRNA',
                                                '2+ types','Truncating'))
    p1 <- ggplot(xm,aes(x=type,y=Percentage)) +
        scale_y_continuous(expand=c(0,0),position='right') +
        geom_bar(stat='identity',fill='black') + 
        coord_flip() +
        theme_std(base_size=12) +
        theme(legend.position='bottom') +
        labs(x=NULL,y='% of cases with\ntruncating variants')
    info <- break_axis(xt$N,maxlower=250,lowerticksize=50,upperticksize=250,ratio_lower_to_upper=0.75)
    xt$y <- info$newy
    p2 <- ggplot(xt,aes(x=type,y=y)) +
        scale_y_continuous(expand=c(0,0),position='right',limits=info$limits,breaks=info$breaks,labels=info$labels) +
        geom_bar(stat='identity',fill='black') +
        coord_flip() +
        theme_std(base_size=12) +
        theme(axis.text.y=element_blank(),axis.ticks.y=element_blank(),
              axis.text.x=element_text(angle=90,vjust=1,hjust=0)) +
    labs(x=NULL,y='T/N Samples with\n90%+ coverage')


    ## barplot of the rate of truncating variants in CI vs CIII-IV for each cancer type
    chrm_annotated <- copy(chrm_annotated_data)
    tbl <- table.freq(chrm_annotated$symbol)
    names(tbl) <- c('Hugo_Symbol','length')

    ## for each sample, get its effective length per gene
    l <- copy(tcga_sample_gene_length_data)
    l <- l[sample_type=='tumor only',]
    l[,sample_type:=NULL]
    l <- as.data.table(reshape2::melt(l,id.vars='Tumor_Sample_Barcode',variable.factor=F))
    names(l) <- c('Tumor_Sample_Barcode','Hugo_Symbol','effective_length')
    l <- merge(l, tbl, by='Hugo_Symbol', all.x=T)
    l$complex <- as.character(NA)
    l[grepl('MT-ND',Hugo_Symbol),complex:='C1']
    l[grepl('MT-CYB',Hugo_Symbol),complex:='C3-5']
    l[grepl('MT-CO',Hugo_Symbol),complex:='C3-5']
    l[grepl('MT-ATP',Hugo_Symbol),complex:='C3-5']
    l <- l[!is.na(complex),]

    ## for each sample and gene, annotate if it is mutated
    d <- copy(tcga_maf_data)
    d <- d[,c('Tumor_Sample_Barcode','Hugo_Symbol','Variant_Classification'),with=F]
    d <- d[Variant_Classification %in% truncating_classes]
    l <- merge(l, d, by=c('Tumor_Sample_Barcode','Hugo_Symbol'), all.x=T)
    l$affected <- !is.na(l$Variant_Classification)
    l[,Variant_Classification:=NULL]
    l <- l[!is.na(complex),]

    ## annotate the cancer type per sample, subset for well-covered samples
    clin <- copy(tcga_clin_data)
    clin <- clin[,c('Tumor_Sample_Barcode','maintype'),with=F]
    l <- merge(l, clin, by='Tumor_Sample_Barcode', all.x=T)

    ## collapse complex for each sample, get % covered, and affected
    collapse_sample_complex <- function(l) {
        variants <- sum(l$affected)
        l <- l[!duplicated(Hugo_Symbol),]
        len <- sum(l$length)
        len_eff <- sum(l$effective_length)
        prop_covered <- len_eff/len
        list(prop_covered=prop_covered, 
             variants=variants,
             complex_length=len)
    }
    dat <- l[,collapse_sample_complex(.SD),by=c('maintype','Tumor_Sample_Barcode','complex')]
    dat <- dat[prop_covered >= 0.9,]
    dat[variants > 0,status:='affected']
    dat[variants == 0,status:='wildtype']
    dat$status <- factor(dat$status, levels=c('wildtype','affected'))
    dat_wide <- reshape(dat[,c('maintype','Tumor_Sample_Barcode','complex','status'),with=F],
                        idvar=c('maintype','Tumor_Sample_Barcode'),timevar='complex',direction='wide')
    names(dat_wide) <- gsub('-','_',names(dat_wide))
    dat_wide <- dat_wide[!is.na(status.C3_5) & !is.na(status.C1),]

    test_tumortype <- function(dat) {
        samples <- nrow(dat)
        status_c1 <- sum(dat$status.C1=='affected')
        status_c345 <- sum(dat$status.C3_5=='affected')
        m <- xtabs(~status.C1 + status.C3_5, data=dat)
        p.value <- mcnemar.test(m)$p.value
        list(samples=samples, status_c1=status_c1, status_c345=status_c345, p.value=p.value)
    }
    result <- dat_wide[,test_tumortype(.SD),by=maintype]
    result$maintype <- factor(result$maintype, levels=xm$type)
    result <- result[order(maintype),]
    result <- result[!is.na(maintype),]
    result$q.value <- p.adjust(result$p.value,method='BH')
    result$prop_c1_affected <- result$status_c1/result$samples
    result$prop_c345_affected <- result$status_c345/result$samples
    resultm <- as.data.table(reshape2::melt(result[,c('maintype','prop_c1_affected','prop_c345_affected'),with=F],id.vars=c('maintype')))
    resultm$variable <- gsub('prop_c1_affected','C1',resultm$variable)
    resultm$variable <- gsub('prop_c345_affected','C3-5',resultm$variable)
    resultm$maintype <- factor(resultm$maintype, levels=levels(xt$type))
    resultm$value <- 100*resultm$value

    cols <- c('#6ba5c5','#d2e7ef')
    names(cols) <- c('C1','C3-5')

    p3 <- ggplot(resultm,aes(x=maintype,y=value)) + 
        scale_y_continuous(expand=c(0,0),breaks=seq(0,30,by=10),limits=c(0,30),position='right') + 
        geom_bar(stat='identity',aes(fill=variable),position=position_dodge(width=0.75),width=0.75) +
        scale_fill_manual(values=cols,name='ETC Complex') + 
        theme_std(base_size=12) +
        theme(axis.text.y=element_blank(),axis.ticks.y=element_blank()) + 
        labs(x=NULL,y='% of affected\nsamples') +
        coord_flip()

    p <- plot_grid(p1, p2, p3, align='h', nrow=1, rel_widths=c(2,0.75,1.5))
    list(plot=p, data=out, complex_diff=result, cancertypes=levels(xm$type))
}

l <- copy(tcga_sample_hotspot_status_data)
tmp <- get_barplot(l)
panel_2d <- tmp$plot
panel_2d
if(get_pdfs) ggsave(here('figures/panel_2d.pdf'),width=8,height=4.7)

```


## 2f

```{r panel_2f, fig.height=5.4, fig.width=7, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# make lollipop plot for homopolymer indel hotspots
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## load chrM region annotations 
d <- copy(chrm_annotated_data)
d[,c('gene_start','gene_end','nt'):=NULL]
setnames(d,'symbol','gene')
d[grep('MT-ND',d$gene),type:='NADH dehydrogenase']
d[grep('MT-RNR',d$gene),type:='rRNA']
d[grep('MT-T',d$gene),type:='tRNA']
d[grep('MT-ATP',d$gene),type:='ATP synthase']
d[grep('MT-CO',d$gene),type:='Cytochrome c oxidase']
d[grep('MT-CYB',d$gene),type:='Cytochrome b']

## add in Control Region
CR1 <- data.table(gene='D-loop1',pos=1:576,type='Control Region',strand=-1)
CR2 <- data.table(gene='D-loop2',pos=16024:16569,type='Control Region',strand=-1)
d <- rbind(d,CR1)
d <- rbind(d,CR2)

collapse_gene <- function(d) {
    start <- min(d$pos)
    end <- max(d$pos)
    width <- end - start
    list(start=start,end=end,width=width)
}
map <- d[,collapse_gene(.SD),by=c('gene','strand')]

## map of gene positions for plot
map$pos <- (map$start + map$end)/2
map$label <- paste0(map$gene,' (',map$start,'-',map$end,')')
map <- map[,c('gene','pos','strand','label'),with=F]

## for un-categorized positions, add 'None'
chrM_length <- 16569
missing_positions <- (1:chrM_length)[1:chrM_length %nin% d$pos]
unkforward <- data.table(gene='None',pos=missing_positions,type='No gene',strand=1)
unkreverse <- data.table(gene='None',pos=missing_positions,type='No gene',strand=-1)
d <- rbind(d,unkforward)
d <- rbind(d,unkreverse)
d$strand <- factor(d$strand,levels=c(1,-1))
d$type <- factor(d$type) 

## get complete base-level annotation for forward and reverse strands
d_forward <- d[strand==1,]
d_forward <- d_forward[order(d_forward$pos),]
dat_forward <- data.table(pos=1:chrM_length)
dat_forward <- merge(dat_forward, d_forward, by='pos', all.x=T)
dat_forward$strand[is.na(dat_forward$strand)] <- 1
dat_forward[is.na(dat_forward$gene),type:='No gene']
dat_forward[is.na(dat_forward$gene),gene:='None']
dat_forward$gene <- factor(dat_forward$gene,levels=unique(dat_forward$gene))
dat_forward$gene_index <- as.integer(dat_forward$gene)
dat_forward$gene_different <- c(F,diff(dat_forward$gene_index)!=0)
dat_forward$unique_gene_index <- cumsum(dat_forward$gene_different)

d_reverse <- d[strand==-1,]
d_reverse <- d_reverse[order(d_reverse$pos),]
dat_reverse <- data.table(pos=1:chrM_length)
dat_reverse <- merge(dat_reverse, d_reverse, by='pos', all.x=T)
dat_reverse$strand[is.na(dat_reverse$strand)] <- -1
dat_reverse[is.na(dat_reverse$gene),type:='No gene']
dat_reverse[is.na(dat_reverse$gene),gene:='None']
dat_reverse$gene <- factor(dat_reverse$gene,levels=unique(dat_reverse$gene))
dat_reverse$gene_index <- as.integer(dat_reverse$gene)
dat_reverse$gene_different <- c(F,diff(dat_reverse$gene_index)!=0)
dat_reverse$unique_gene_index <- cumsum(dat_reverse$gene_different)

collapse <- function(d) {
    out <- d[1,]
    out$xstart <- min(d$pos)
    out$xend <- max(d$pos)
    out    
}
dat_forward_collapsed <- dat_forward[,collapse(.SD),by=unique_gene_index]
dat_reverse_collapsed <- dat_reverse[,collapse(.SD),by=unique_gene_index]

## set color scheme for plot
### color the different genes based on their functions in MT genome
region_colors <- c('#BEBEBD',     # tRNA
              '#8150A0',     # NADH dehydrogenase
              '#FAA41A',     # rRNA
              '#3953A4',     # Cytochrome c oxidase
              '#6ABD45',     # ATP synthase
              '#F6EB16',     # Cytochrome b
              '#F8BFCA')     # Control Region 

names(region_colors) <- c('tRNA','NADH dehydrogenase','rRNA','Cytochrome c oxidase','ATP synthase','Cytochrome b','Control Region')

## tweak these to position the circles correctly
y_scaling_factor <- 500
y_offset <- 0.3 ## change inner radius size
mid_ypos <- (y_offset*0.8+y_offset*0.9)/2
delta <- y_offset - mid_ypos
min_ypos <- mid_ypos - delta/1.5
max_ypos <- mid_ypos + delta/1.5

l00 <- (0/y_scaling_factor)+y_offset
l10 <- (10/y_scaling_factor)+y_offset
l20 <- (20/y_scaling_factor)+y_offset
l30 <- (30/y_scaling_factor)+y_offset
l40 <- (40/y_scaling_factor)+y_offset
l50 <- (50/y_scaling_factor)+y_offset
l100 <- (100/y_scaling_factor)+y_offset

## collapse the expanded genes into single rows of table
dat <- rbind(dat_forward_collapsed,dat_reverse_collapsed)
dat$ystart <- ifelse(dat$strand==-1,min_ypos,mid_ypos)
dat$yend <- ifelse(dat$strand==-1,mid_ypos,max_ypos)
dat[gene=='None',type:=NA]
dat[gene=='None',gene:=NA]
dat$gene_exists <- ifelse(is.na(dat$gene),NA,'yes')
dat$gene_exists <- factor(dat$gene_exists,levels='yes')
dat[is.na(gene),label:=NA]
dat[is.na(gene),ystart:=NA]
dat[is.na(gene),yend:=NA]
dat <- dat[!is.na(gene),]

## load homopolymer indel hotspot data
homopolymer_indel_hotspots <- copy(hotspots_indel_data)
setnames(homopolymer_indel_hotspots,'mutant','samples_mutated')
setnames(homopolymer_indel_hotspots,'covered','samples_covered')
setnames(homopolymer_indel_hotspots,'homopolymer_element','nt')
setnames(homopolymer_indel_hotspots,'homopolymer_width','width')

cts <- homopolymer_indel_hotspots[,c('Hugo_Symbol','homopolymer','Start_Position','End_Position',
                                     'samples_mutated','samples_covered','width','nt','q.value'),with=F]
setnames(cts,'samples_mutated','n')
cts$n <- cts$n/y_scaling_factor + y_offset
cts[,Start_Position:=Start_Position+1]
cts$hotspot <- 'Not significant'
cts$hotspot[cts$q.value < 0.01] <- 'q<0.01'
cts$label <- cts$homopolymer
cts[hotspot %nin% c('q<0.01'),label:=NA]


## plot
nt_cols <- c('#df65b0','#1d91c0','#78c679','#cb181d')
names(nt_cols) <- c('A','C','G','T')

p <- ggplot(dat) +
    geom_segment(data=cts[1,],x=l00,xend=l00,y=1,yend=16569,color='grey',linetype=2,size=0.5) +
    geom_segment(data=cts[1,],x=l50,xend=l50,y=1,yend=16569,color='grey',linetype=2,size=0.5) +
    geom_segment(data=cts[1,],x=0.8*y_offset,xend=0.8*y_offset,y=1,yend=16569,color='black',size=0.5) +
    geom_segment(data=cts[1,],x=0.9*y_offset,xend=0.9*y_offset,y=1,yend=16569,color='black',size=0.5) +

    geom_segment(data=cts[hotspot!='q<0.01'],x=y_offset,aes(xend=n,y=Start_Position,yend=Start_Position),
                 color='grey',size=0.25,alpha=0.5) +
    geom_point(data=cts[hotspot!='q<0.01'],aes(x=n,y=Start_Position,color=nt,size=width),pch=19,alpha=0.5) +

    geom_segment(data=cts[hotspot=='q<0.01'],x=y_offset,
                 aes(xend=n,y=Start_Position,yend=Start_Position),color='black',size=0.25) +
    geom_point(data=cts[hotspot=='q<0.01'],
               aes(x=n,y=Start_Position,fill=nt,size=width),pch=21,color='black') +

    geom_rect(aes(ymin=xstart,ymax=xend,xmin=ystart,xmax=yend,fill=type),color='black',size=0.5) + 
    scale_color_manual(values=nt_cols,name=NULL) + 
    scale_fill_manual(values=c(region_colors,nt_cols),name=NULL) + 
    scale_x_continuous(limits=c(0,max(cts$n))) +
    scale_y_continuous(position='left') +
    geom_text_repel(data=cts,color='black',aes(x=n,y=Start_Position,label=label),size=2.5) +
    theme_std(base_size=12) + 
    theme(
        axis.line.x=element_blank(), axis.ticks.x=element_blank(), axis.text.x=element_blank(), 
        axis.line.y=element_blank(),axis.ticks.y=element_blank(), axis.text.y=element_blank(), 
        legend.position='right') +
    coord_polar(theta="y") +
    labs(x=NULL,y=NULL,title=NULL)
p <- extract_gglegend(p)
panel_2f <- plot_grid(p$plot,p$legend,ncol=2,rel_widths=c(4,1))
panel_2f
if(get_pdfs) ggsave(here('figures/panel_2f.pdf'),width=7,height=5.4)

```


## 2g
```{r panel_2g, fig.width=3.8, fig.height=2.6, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# panel showing where single nucleotide repeats arise and 
# how frequently affected
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ms <- copy(hotspots_indel_data)
setnames(ms,'mutant','samples_mutated')
setnames(ms,'covered','samples_covered')
setnames(ms,'homopolymer_element','nt')
setnames(ms,'homopolymer_width','width')
ms$nlog10q <- -log10(ms$q.value)

ms <- ms[,c('Hugo_Symbol','nt','width','samples_mutated','samples_covered','prop','q.value','homopolymer'),with=F]
ms$Hugo_Symbol <- factor(ms$Hugo_Symbol, levels=rev(sortunique(ms$Hugo_Symbol)))
ms[q.value < 0.01,label:=homopolymer]
ms$nlog10q <- -log10(ms$q.value)

panel_2g <- ggplot(ms, aes(x=Hugo_Symbol,y=width)) + 
    geom_point(aes(size=nlog10q,color=nt),
               position=position_jitter(seed=42,width=0.1,height=0.1)) +
    geom_text_repel(aes(label=label),size=3,min.segment.length=1,seed=42) +
    scale_color_manual(values=nt_cols,name='Nucleotide') + 
    theme_std(base_size=12) +
    theme(legend.position='right') +
    coord_flip() +
    labs(x=NULL,y='Nucleotide repeat length (bp)')
panel_2g
if(get_pdfs) ggsave(here('figures/panel_2g.pdf'),width=3.8,height=2.6)

```

***

## 3a

```{r panel_3a, fig.height=4, fig.width=4, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# pie chart showing majority of variants are VUS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

d <- copy(tcga_maf_data)

## create levels for pie-chart
d[Variant_Classification %in% truncating_classes,type:='Truncating']
d[Variant_Classification %in% nontruncating_classes & Variant_Classification=='rRNA',type:='rRNA']
d[Variant_Classification %in% nontruncating_classes & Variant_Classification=='tRNA',type:='tRNA']
d[Variant_Classification %in% nontruncating_classes & Variant_Classification %nin% c('tRNA','rRNA'),type:='Missense']
d[Variant_Classification %in% 'Silent',type:='Silent']

hotspots <- copy(hotspots_snv_data)
hs <- hotspots$Start_Position[hotspots$q.value < 0.01]
d$hotspot <- as.character(NA)
d[type %in% c('Missense','rRNA','tRNA') & Variant_Type=='SNP' & Start_Position %in% hs,hotspot:='Q < 0.01']
d[type %in% c('Missense','rRNA','tRNA') & Variant_Type=='SNP' & Start_Position %nin% hs,hotspot:='Not significant']

## make plot
tbl <- table.freq(d$type)
tbl$prop <- tbl$N / sum(tbl$N)
tbl$label <- paste0(tbl$value,' (',prettyNum(tbl$prop*100,digits=2),'%)')
tbl$value <- factor(tbl$value, levels=c('Missense','rRNA','tRNA','Truncating','Silent'))
tbl <- tbl[order(value)]
tbl$label <- factor(tbl$label, levels=tbl$label)

## theme for pie chart
piechart_theme <- theme_minimal()+
    theme_std(base_size=12) + 
    theme(
      axis.text.x=element_blank(),
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      panel.border = element_blank(),
      panel.grid=element_blank(),
      axis.ticks = element_blank(),
      axis.line.x = element_blank(),
      axis.line.y = element_blank()
)

## pie chart
cols <- c('#357a33','#4DAF4A','#94cf92','#E41A1C','#377EB8')
names(cols) <- levels(tbl$label)
variants_overall <- paste0('N=',sum(tbl$N))

panel_3a <- ggplot(tbl, aes(x="", y=prop, fill=label))+
    geom_bar(width = 1, stat = "identity") + 
    coord_polar("y", start=0) +
    piechart_theme +
    scale_fill_manual(values=cols,name=paste0('mtDNA mutation type\n(',variants_overall,')'))
panel_3a
if(get_pdfs) ggsave(here('figures/panel_3a.pdf'),width=4,height=4)

```


## 3b 
```{r panel_3b, fig.height=5, fig.width=2.5, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# predicted pathogenicity of observed/not-observed 
# non-truncating protein-coding
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

d <- copy(tcga_maf_data)
helix <- copy(helix_maf_data)
mitimpact <- copy(mitimpact_data)

## load the APOGEE data from MitImpact, annotate hotspots, check if they're predicted more pathogenic
mitimpact$ShortVariantID <- paste0(mitimpact$Ref,mitimpact$Start,mitimpact$Alt)
mitimpact$APOGEE_boost_mean_prob <- as.numeric(gsub(',','.',mitimpact$APOGEE_boost_mean_prob))
apogee <- mitimpact[,c('Gene_symbol','OXPHOS_complex','ShortVariantID','Start','APOGEE_boost_mean_prob',
                       'APOGEE_boost_consensus','ClinVar_Oct2017_CLINSIG'),with=F]
apogee <- apogee[order(Start)]
setnames(apogee,'APOGEE_boost_mean_prob','prob')
setnames(apogee,'APOGEE_boost_consensus','consensus')

## get frequency of observed somatic SNVs
d <- d[Variant_Type=='SNP',]
d <- d[Variant_Classification %in% nontruncating_classes,] ## make sure to exclude silent
variants_ever_in_germline <- sortunique(d$ShortVariantID[d$ShortVariantID %in% helix$ShortVariantID])
d[ShortVariantID %in% variants_ever_in_germline,class:='Observed in germline']
d[ShortVariantID %nin% variants_ever_in_germline,class:='Somatic only']
getN <- function(d) {
    N <- length(unique(d$PATIENT_ID))
    list(N=N)
}
info <- d[,getN(.SD),by=c('class','ShortVariantID')]
x <- merge(apogee, info, by='ShortVariantID', all.x=T)
x[is.na(N),N:=0]
x[N==0,class:='Never observed']

## get prop pathogenic
summarize <- function(x) {
    path <- sum(x$consensus=='P')
    total <- nrow(x)
    list(path=path,total=total)
}
res <- x[,summarize(.SD),by=class]
ci <- binom.confint(res$path,res$total,method='exact')
res$prop <- ci$mean*100
res$lwr <- ci$lower*100
res$upr <- ci$upper*100
res <- res[order(prop,decreasing=F),]
res$class <- factor(res$class, levels=res$class)

getp <- function(ind1,ind2, res) {
    pval <- prop.test(res$path[c(ind1,ind2)],res$total[c(ind1,ind2)],alternative='two.sided')$p.value
    plab <- plabel(pval)
    plab
}
p12 <- getp(1,2,res)
p13 <- getp(1,3,res)
p23 <- getp(2,3,res)
setnames(res,'path','pathogenic')

panel_3b <- ggplot(res,aes(x=class,y=prop)) +
    scale_y_continuous(expand=c(0,0),limits=c(0,55),breaks=seq(0,40,by=10)) +
    geom_bar(stat='identity',fill='#68b8d6') +
    geom_errorbar(aes(min=lwr,max=prop),width=NA,color='white') +
    geom_errorbar(aes(min=prop,max=upr),width=NA,color='black') +
    geom_text(aes(x=class,label=paste0('N=',total)),angle=90,hjust=0,vjust=0.5,y=0.5,color='white') +
    geom_signif(comparisons=list(c(1,2),c(1,3),c(2,3)), annotations=c(p12,p13,p23),
                tip_length=0,y_position=c(45,48,51)) +
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)) +
    labs(x='Unique mtDNA variants\n(protein-coding VUS)',y='% predicted pathogenic')
panel_3b
if(get_pdfs) ggsave(here('figures/panel_3b.pdf'),width=4,height=6)


```


## 3c

```{r panel_3c, fig.height=6.3, fig.width=2.5, eval=T}

## plot nontruncating rates
x_vus <- combinatorial_poisson_test('complex','total_nontruncating','total_complex_length_mb',result_tcga,
                                'upper_nontrunc',ygap=4)
f <- function(i,x) c(x_vus$a[i],x_vus$b[i])
comparisons <- lapply(1:nrow(x_vus),f,x_vus)

p1 <- ggplot(result_tcga,aes(x=complex,y=rate_nontrunc)) +
    geom_bar(stat='identity',fill='#6ba4c4') +
    geom_errorbar(aes(min=lower_nontrunc,max=rate_nontrunc),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_nontrunc,max=upper_nontrunc),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x_vus$ypos,annotations=x_vus$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Non-truncating',y='Mutations/Mb') +
    scale_y_continuous(expand=c(0,0),limits=c(0,100),breaks=seq(0,80,by=20)) 

## plot silent rates
x <- combinatorial_poisson_test('complex','total_silent','total_complex_length_mb',result_tcga,
                                'upper_silent',ygap=1)
f <- function(i,x) c(x$a[i],x$b[i])
comparisons <- lapply(1:nrow(x),f,x)

p2 <- ggplot(result_tcga,aes(x=complex,y=rate_silent)) +
    geom_bar(stat='identity',fill='#6ba4c4') +
    geom_errorbar(aes(min=lower_silent,max=rate_silent),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_silent,max=upper_silent),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x$ypos,annotations=x$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Synonymous',y='Mutations/Mb') +
    scale_y_continuous(expand=c(0,0),limits=c(0,25),breaks=seq(0,20,by=5)) 

panel_3c <- plot_grid(p1,p2,ncol=1,align='v')
panel_3c
if(get_pdfs) ggsave(here('figures/panel_3c.pdf'),width=2.5,height=6.3)
```


## 3d

```{r panel_3d, fig.height=6.3, fig.width=2.5, eval=T}

## plot nontruncating rates
x <- combinatorial_poisson_test('complex','total_nontruncating','total_complex_length_mb',result_pcawg,
                                'upper_nontrunc',ygap=4)
f <- function(i,x) c(x$a[i],x$b[i])
comparisons <- lapply(1:nrow(x),f,x)

p1 <- ggplot(result_pcawg,aes(x=complex,y=rate_nontrunc)) +
    geom_bar(stat='identity',fill='#6ba4c4') +
    geom_errorbar(aes(min=lower_nontrunc,max=rate_nontrunc),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_nontrunc,max=upper_nontrunc),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x$ypos,annotations=x$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Non-truncating',y='Mutations/Mb') +
    scale_y_continuous(expand=c(0,0),limits=c(0,115),breaks=seq(0,100,by=20)) 

## plot silent rates
x <- combinatorial_poisson_test('complex','total_silent','total_complex_length_mb',result_pcawg,
                                'upper_silent',ygap=1)
f <- function(i,x) c(x$a[i],x$b[i])
comparisons <- lapply(1:nrow(x),f,x)

p2 <- ggplot(result_pcawg,aes(x=complex,y=rate_silent)) +
    geom_bar(stat='identity',fill='#6ba4c4') +
    geom_errorbar(aes(min=lower_silent,max=rate_silent),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_silent,max=upper_silent),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x$ypos,annotations=x$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Synonymous',y='Mutations/Mb') +
    scale_y_continuous(expand=c(0,0),limits=c(0,25),breaks=seq(0,20,by=5)) 

panel_3d <- plot_grid(p1,p2,ncol=1,align='v')
panel_3d
if(get_pdfs) ggsave(here('figures/panel_3d.pdf'),width=2.5,height=6.3)

```


## 3e

```{r panel_3e, fig.height=2.75, fig.width=6, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# panel showing enrichment for mutations vs number of samples mutated
# highlight points with the type of variant (PC, tRNA, rRNA)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

d <- copy(tcga_maf_data)
d <- d[class %in% c('somatic')]
d <- d[Variant_Classification %nin% 'Silent']
d$rug <- 'Missense'
d[Variant_Classification %in% truncating_classes,rug:='Truncating']
d[Variant_Classification=='rRNA',rug:='rRNA']
d[Variant_Classification=='tRNA',rug:='tRNA']
d[Variant_Classification %in% c('Frame_Shift_Del','Frame_Shift_Ins'),Variant_Classification:='Frame-shift indel']
d[Variant_Classification %in% c('In_Frame_Del','In_Frame_Ins'),Variant_Classification:='In-frame indel']
d[Variant_Classification %in% 'Frame-shift indel' & homopolymer!='', ShortVariantID:=homopolymer]

summarize_sample <- function(d) {
    samples <- length(unique(d$Tumor_Sample_Barcode))
    rug <- unique(d$rug)
    class <- unique(d$Variant_Classification)
    list(n=samples, rug=rug, class=class)
}
info <- d[,summarize_sample(.SD),by=ShortVariantID]
info <- info[order(n,decreasing=T),]
info$position <- 1:nrow(info)
info <- info[n >= 3,]
info$type <- info$rug
info[type!='Truncating',type:='Non-truncating']

cols <- brewer.pal(7,'Set1')[c(1,2,3)]
cols2 <- c(cols[1],'#bfbfbf')
names(cols2) <- c('Truncating','Non-truncating')


p <- wilcox.test(info$position[info$type=='Truncating'],info$position[info$type=='Non-truncating'])$p.value
plab <- plabel(p)

p1 <- ggplot(info,aes(x=position,y=n)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_segment(data=info[type=='Non-truncating',],aes(x=position,xend=position,y=0,yend=n,color=type),size=0.25) +
    geom_segment(data=info[type=='Truncating',],aes(x=position,xend=position,y=0,yend=n,color=type),size=0.25) +
    scale_color_manual(values=cols2,name='Variant') + 
    geom_text(data=info[1,],label=plab,y=15,x=100,size=3) +
    theme_std(base_size=12) +
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
    labs(x=NULL,y='Cases affected')
p1 <- extract_gglegend(p1)


cols4 <- c(cols[1],'#bfbfbf',cols[2:3])
names(cols4) <- c('Truncating','Missense','tRNA','rRNA')

p2 <- ggplot() +
    scale_y_continuous(expand=c(0,0),limits=c(0,4),
                       labels=c('rRNA','tRNA','Missense','Truncating'),
                       breaks=c(0.5,1.5,2.5,3.5)) + 
    geom_segment(data=info[rug=='Truncating'],aes(x=position,xend=position),y=3,yend=4,size=0.25) +
    geom_segment(data=info[rug=='Missense'],aes(x=position,xend=position),y=2,yend=3,size=0.25) +
    geom_segment(data=info[rug=='tRNA'],aes(x=position,xend=position),y=1,yend=2,size=0.25) +
    geom_segment(data=info[rug=='rRNA'],aes(x=position,xend=position),y=0,yend=1,size=0.25) +
    theme_std(base_size=12) +
    theme(
          axis.text.x=element_blank(),axis.ticks.x=element_blank(),axis.line.x=element_blank(),
          axis.ticks.y=element_blank(),axis.line.y=element_blank(),
          legend.position='right') +
    labs(x='mtDNA variants',y=NULL)
p2 <- extract_gglegend(p2)
p <- plot_grid(p1$plot, p2$plot, align='v', rel_heights=c(1.5,1), ncol=1)
pl <- plot_grid(p1$legend, p2$legend, ncol=1)
panel_3e <- plot_grid(p, pl, rel_widths=c(4,1), ncol=2)
panel_3e
if(get_pdfs) ggsave(here('figures/panel_3e.pdf'),width=4.6,height=2.4)

```


## 3f

```{r panel_3f, fig.width=5, fig.height=2, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# panel showing enrichment for mutations vs number of samples mutated
# highlight points with the type of variant (PC, tRNA, rRNA)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

cols <- c('#e41a1c','#377eb8','#4daf4a','#bfbfbf')
names(cols) <- c('Protein-coding','rRNA','tRNA','Not significant')

hs <- copy(hotspots_snv_data)
hs$nlog10q <- -log10(hs$q.value)
setnames(hs,'protien_change','protein_change')
hs <- hs[,c('Start_Position','nlog10q','q.value','Hugo_Symbol','label','protein_change','mutations_at_pos'),with=F]
hs$class <- 'Not significant'
hs[q.value < 0.01 & !is.na(protein_change),class:='Protein-coding']
hs[q.value < 0.01 & grepl('MT-R',Hugo_Symbol),class:='rRNA']
hs[q.value < 0.01 & grepl('MT-T',Hugo_Symbol),class:='tRNA']
hs <- hs[order(q.value,decreasing=F)]
hs[q.value >= 0.01,label:=NA]
hs[grepl('MT-ND',Hugo_Symbol),Complex:='I']
hs[grepl('MT-CY',Hugo_Symbol),Complex:='III']
hs[grepl('MT-CO',Hugo_Symbol),Complex:='IV']
hs[grepl('MT-ATP',Hugo_Symbol),Complex:='V']
x <- hs[q.value < 0.01]

panel_3f <- ggplot(hs,aes(x=mutations_at_pos,y=nlog10q)) +
    geom_point(aes(color=class),position=position_jitter(width=0.2,seed=42)) +
    geom_text_repel(aes(color=class,label=label),size=2.5) +
    scale_color_manual(values=cols,name='SNV Hotspot') +
    theme_std(base_size=12) +
    scale_x_continuous(limits=c(3,13),breaks=seq(3,13,by=1)) +
    labs(x='# of mutations in tumor (absent in normal)',y='-log10(Q-value)')
panel_3f
if(get_pdfs) ggsave(here('figures/panel_3f.pdf'),width=6,height=2.5)

```


## 3g

```{r, panel_3g, fig.height=4, fig.width=9, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# tRNA hotspots panel
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## find any positions in aligned trnas with positive/negative selection
## based on permutation tests

res <- copy(hotspots_trna_data)
collapse_pos <- function(res) {
    qval <- res$q.value[1]
    total_mutated <- sum(res$samples_mutated)
    list(qval=qval, total_mutated=total_mutated)
}
info <- res[,collapse_pos(.SD),by=c('trna_region','trna_pos','index')]
info[trna_region=='',trna_region:='n/a']
info <- info[order(index)]
region_levels <- c(unique(info$trna_region[info$trna_region!='n/a']),'n/a')
region_cols <- c(brewer.pal(8,'Paired'),'black')
names(region_cols) <- region_levels
mylabels <- info$trna_pos
mylabels[2:(length(mylabels)-1)] <- ''
info$nlog10q <- -log10(info$qval)

scaling_factor <- 5
pd <- info[,c('trna_region','trna_pos','total_mutated','nlog10q'),with=F]
suppressWarnings(pd <- melt(pd, id.var=c('trna_region','trna_pos')))
pd[variable=='total_mutated',value:=-1*value]
pd$trna_pos <- factor(pd$trna_pos, levels=info$trna_pos)
pd[variable=='nlog10q',value:=scaling_factor*value]
pd[variable=='total_mutated',variable:=trna_region]
pd$variable <- factor(pd$variable, levels=c(region_levels,'nlog10q'))

cols <- c(brewer.pal(8,'Accent'),'white','black')
names(cols) <- c(region_levels,'nlog10q')
mybreaks <- seq(-15,15,by=5)
mylabels <- c(seq(-15,0,by=5),seq(5,15,by=5)/scaling_factor)
panel_3g <- ggplot(pd,aes(x=trna_pos,value)) +
    geom_hline(yintercept=2*scaling_factor,color='red',linetype=2,size=0.5) +
    geom_hline(yintercept=0) + 
    scale_y_continuous(expand=c(0,0),breaks=mybreaks,labels=mylabels) +
    scale_x_discrete(labels=mylabels) + 
    scale_fill_manual(values=cols,name='tRNA region') + 
    geom_bar(stat='identity',aes(fill=variable),color='black') +
    theme_std(base_size=12) +
    labs(x='Aligned cloverleaf position',y='N cases          -log10(Q)') +
    theme( 
          axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          axis.line.x=element_blank(),legend.position='right')
panel_3g
if(get_pdfs) ggsave(here('figures/panel_3g.pdf'),width=9,height=4)

```


## 3h

```{r panel_3h, fig.height=3.8, fig.width=2.3, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# compare predicted pathogenicity of pos 31 vs others
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

mitotip <- copy(mitotip_data)
names(mitotip) <- c('Start_Position','ref','alt','predscore')
mitotip <- mitotip[ref %in% c('A','C','G','T') & alt %in% c('A','C','G','T'),]
mitotip$ShortVariantID <- paste0(mitotip$ref,mitotip$Start_Position,mitotip$alt)

d <- copy(tcga_maf_data)
tbl <- table.freq(d$ShortVariantID)
mitotip <- merge(mitotip, tbl, by.x='ShortVariantID', by.y='value', all.x=T)
mitotip[is.na(N),N:=0]

tmp <- copy(hotspots_trna_data)
tmp <- tmp[,c('pos','q.value','Hugo_Symbol','trna_pos'),with=F]
setnames(tmp,'pos','Start_Position')
result <- merge(mitotip, tmp, by='Start_Position', all.x=T)
#result[N==0,hotspot:='never observed']
result[N >= 0 & q.value >= 0.01,hotspot:='at other pos']
result[N >= 0 & q.value < 0.01,hotspot:='at pos 31']
result <- result[!is.na(q.value), ]
test <- function(group1,group2,result) {
    x1 <- result$predscore[result$hotspot==group1]
    x2 <- result$predscore[result$hotspot==group2]
    pval <- wilcox.test(x1,x2)$p.value
    pval <- paste0('P=',prettyNum(pval,digits=4))
    pval
}
#p12 <- test('never observed','at other pos',result)
#p13 <- test('never observed','at pos 31',result)
p23 <- test('at other pos','at pos 31',result)
#result$hotspot <- factor(result$hotspot, levels=c('never observed','at other pos','at pos 31'))
result$hotspot <- factor(result$hotspot, levels=c('at other pos','at pos 31'))
result$recurrent31 <- F
result[hotspot=='at pos 31' & N > 1,recurrent31:=T]
result[recurrent31==T,label:=paste(Hugo_Symbol,ShortVariantID)]

result2 <- result[hotspot=='at pos 31',]
otherpos <- result[hotspot=='at other pos']
nother <- nrow(otherpos)
ntosubset <- round(nother*0.05)
selected_points <- sample(1:nother,replace=F)[1:ntosubset]
otherpos <- otherpos[selected_points]
result2 <- rbind(result2, otherpos)

panel_3h <- ggplot(result2,aes(x=hotspot,y=predscore)) + 
    scale_y_continuous(limits=c(0,30),breaks=seq(0,30,by=5)) + 
    geom_point(data=result[recurrent31==F],position=position_jitter(width=0.2,seed=42),color='#BFBFBF') +
    geom_point(data=result[recurrent31==T],position=position_jitter(width=0.2,seed=42),size=2,color='red') +
    geom_boxplot(fill=NA,outlier.shape=NA) +
    geom_text_repel(aes(label=label),color='red',size=3) +
    geom_signif(comparisons=list(c(1,2)),annotation=c(p23),y_position=24+c(0),tip_length=NA) + 
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.50)) +
    labs(x='All tRNA mutations ...',y='Predicted pathogenicity\n(MitoTIP score)')
panel_3h
if(get_pdfs) ggsave(here('figures/panel_3h.pdf'),width=2.3,height=3.8)

```


## 3j

```{r panel_3j, fig.width=6, fig.height=3.5, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# panel comparing R25Q to Truncating in colorectal cancer
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

d1 <- copy(rnaseq_r25q_gsea_data)
d1 <- d1[type=='Colorectal Cancer']
d1 <- d1[,c('pathway','padj','NES'),with=F]
d1$nlog10q <- -log10(d1$padj)
d1$label <- d1$pathway
d1[padj >= 0.01,label:=NA]
d1[NES < 0 & pathway %nin% c('TNFA_signaling_via_NFKB','Hypoxia','Interferon_gamma_response','Apoptosis',
                             'Glycolysis','Complement','IL6_JAK_STAT3_signaling','IL2_STAT5_signaling'),label:=NA]
cols <- c('#6ba5c5','#bfbfbf','#ea948d')
names(cols) <- c('Down in MT-ND1 R25Q','Not significant','Up in MT-ND1 R25Q')
d1$direction <- 'Not significant'
d1[NES > 0 & padj < 0.01,direction:='Up in MT-ND1 R25Q']
d1[NES < 0 & padj < 0.01,direction:='Down in MT-ND1 R25Q']

panel_3j <- ggplot(d1,aes(x=NES,y=nlog10q)) + 
    geom_point(aes(color=direction),size=3) +
    scale_color_manual(values=cols,name='Change in expression') + 
    theme_std(base_size=12) +
    geom_text_repel(aes(label=label),color='black',size=3) +
    labs(x='Normalized enrichment score\n(ND1 R25Q vs WT)',y='-log10(Q)')
panel_3j

if(get_pdfs) ggsave(here('figures/panel_3j.pdf'),width=6,height=3.5)
```



***

# Figure 4

## 4a

```{r panel_4a, fig.width=6, fig.height=4.7, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# proportion barplot indicating the estimated prop of samples 
# in each cancertype affected by mtdna variants and their breakdown
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

get_barplot <- function(l,tcga=T) {
    setnames(l,'annotation_detailed_prefilter','label')
    min.n <- 20
    cols <- c('#e5e5e5',brewer.pal(6,'Set3'))

    x <- xtabs(~maintype + label, data=l[label!='Unknown'])
    x <- as.data.frame.matrix(x)
    out <- x
    Total <- rowSums(x)
    x <- x/Total
    x <- cbind(type=rownames(x),adt(x))
    xt <- data.table(type=x$type,N=Total)
    xt <- xt[N >= min.n,]
    x <- x[type %in% xt$type,]
    setnames(x,'2+ types, non-truncating/silent','2+ types')
    x <- x[order(Wildtype,Silent,Missense,rRNA,tRNA,`2+ types`,Truncating),]
    xm <- as.data.table(reshape2::melt(x,id.vars='type'))
    xm$type <- factor(xm$type,levels=x$type)
    xt$type <- factor(xt$type,levels=x$type)
    xm$variable <- factor(xm$variable, levels=c('Wildtype','Silent','Missense','tRNA','rRNA',
                                                '2+ types','Truncating'))
    xm$Percentage <- 100*xm$value

    p1 <- ggplot(xm,aes(x=type,y=Percentage)) +
        scale_y_continuous(expand=c(0,0),position='right') +
        geom_bar(stat='identity',aes(fill=variable)) + 
        scale_fill_manual(values=cols,name=NULL) + 
        coord_flip() +
        theme_std(base_size=12) +
        theme(legend.position='bottom') +
        labs(x=NULL,y='Percentage of samples')
    p1 <- extract_gglegend(p1)

    info <- break_axis(xt$N,maxlower=250,lowerticksize=50,upperticksize=250,ratio_lower_to_upper=0.75)
    xt$y <- info$newy
    p2 <- ggplot(xt,aes(x=type,y=y)) +
        scale_y_continuous(expand=c(0,0),position='right',limits=info$limits,breaks=info$breaks,labels=info$labels) +
        geom_bar(stat='identity') +
        coord_flip() +
        theme_std(base_size=12) +
        theme(axis.text.y=element_blank(),axis.ticks.y=element_blank(),
              axis.text.x=element_text(angle=90,vjust=1,hjust=0)) +
    labs(x=NULL,y='T/N Samples with\n90%+ coverage')

    p <- plot_grid(p1$plot, p2, align='h',rel_widths=c(4,1))
    p <- plot_grid(p, p1$legend, ncol=1,rel_heights=c(4,1))
    list(plot=p, data=out, cancertypes=levels(xm$type))
}

l <- copy(tcga_sample_hotspot_status_data)
tmp <- get_barplot(l)
panel_4a_cancertypes <- tmp$cancertypes
panel_4a <- tmp$plot
panel_4a
if(get_pdfs) ggsave(here('figures/panel_4a.pdf'),width=6,height=4.7)

```

## 4b

```{r panel_4b, fig.width=7.3,fig.height=3.9,eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# barplot expression across types per geneset
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

x <- copy(rnaseq_truncating_gsea_data)
x <- x[,c('pathway','type','padj','NES'),with=F]
x$stat <- -log10(x$padj) * sign(x$NES)
m <- x[,c('pathway','type','stat'),with=F]
summarize_signature <- function(m) {
    mid <- median(m$stat)
    n <- sum(abs(m$stat) > 2) #/ nrow(m)
    m <- m[order(stat,decreasing=F),]
    m$n <- n
    m$mid <- mid
    m$pos <- ((1:nrow(m))/nrow(m))
    m
}
m <- m[,summarize_signature(.SD), by=pathway]
m <- m[order(mid,pos,decreasing=F),]
m <- m[n >= 3,]
m$pathway <- factor(m$pathway, levels=unique(m$pathway))
m$label <- m$pathway
m$pos <- as.integer(m$pathway) + m$pos - 0.5
info <- m[!duplicated(pathway),]
panel_4b <- ggplot(m, aes(x=pos,y=stat)) +
    scale_y_continuous(limits=c(-5,5)) + 
    geom_bar(data=info, stat='identity',aes(x=pathway, y=mid, fill=n),color='black') + 
    geom_point(pch=19,size=0.8) +
    scale_fill_gradient(low='white',high='#4675b4',name='N cancer types\nsignificant') + 
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)) + 
    labs(x=NULL,y='Sign(NES) * -log10(Q-value)')
panel_4b
if(get_pdfs) ggsave(here('figures/panel_4b.pdf'),width=10,height=5)

```


## 4c

```{r panel_4c, fig.width=7.1, fig.height=4}

cols <- c('#6ba5c5','#bfbfbf','#ea948d')
names(cols) <- c('Down in truncating','Not significant','Up in truncating')

hist1 <- m[pathway %in% 'Oxidative_phosphorylation',]
hist1$label <- hist1$type
hist1[abs(stat) <= 2, label:=NA]
hist1$direction <- 'Not significant'
hist1[stat > 2,direction:='Up in truncating']
hist1[stat < -2,direction:='Down in truncating']
hist1$geneset <- 'Oxidative phosphorylation'

hist2 <- m[pathway %in% 'TNFA_signaling_via_NFKB',]
hist2$label <- hist2$type
hist2[abs(stat) <= 2, label:=NA]
hist2$direction <- 'Not significant'
hist2[stat > 2,direction:='Up in truncating']
hist2[stat < -2,direction:='Down in truncating']
hist2$geneset <- 'TNFa signaling via NFkB'

hist <- rbind(hist1, hist2)
hist$geneset <- factor(hist$geneset, levels=c('TNFa signaling via NFkB','Oxidative phosphorylation'))
info <- copy(tcga_sample_hotspot_status_data)
x <- xtabs(~maintype + annotation_prefilter, data=info[annotation_prefilter!='Unknown'])
x <- as.data.frame.matrix(x)
x$N <- rowSums(x)
x$prop_truncating <- x$Truncating / x$N
x <- cbind(type=rownames(x), adt(x))
x <- x[,c('type','N','prop_truncating'),with=F]
hist <- merge(hist, x, by='type', all.x=T)
hist$percent <- 100*hist$prop_truncating

panel_4c <- ggplot(hist, aes(x=percent, y=stat)) +
    scale_y_continuous(expand=c(0,0),limits=c(-4.5,4.5),breaks=seq(-4,4,by=2)) +
    scale_x_continuous(expand=c(0,0),breaks=seq(0,40,by=10),limits=c(0,40)) +
    geom_point(aes(size=N,color=direction)) +
    geom_text_repel(aes(label=label,color=direction),min.segment.length=0.01) +
    scale_color_manual(values=cols,name='Direction') + 
    theme_std(base_size=12) +
    theme(legend.position='bottom') +
    geom_hline(yintercept=0,linetype=1) +
    labs(x='Percent of cases with truncating variants',y='-log10(Q)') +
    facet_wrap(facets=~geneset,nrow=1)
panel_4c
if(get_pdfs) ggsave(here('figures/panel_4c.pdf'),width=7.1,height=4)

```



## 4d

```{r panel_4d, fig.height=3.1,fig.width=6.7, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# volcano of univarite survival across cancer types
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

clin <- copy(tcga_cdr_supptab1)
clin <- clin[,c('bcr_patient_barcode','OS','OS.time','PFI','PFI.time'),with=F]
info <- copy(tcga_sample_hotspot_status_data)
setnames(info,'annotation_postfilter','mtstatus')
info <- info[mtstatus %in% c('Wildtype','VUS','Truncating')]
info$PATIENT_ID <- strtrim(info$Tumor_Sample_Barcode,12) 
info <- info[,c('PATIENT_ID','maintype','mtstatus'),with=F]
info <- merge(info, clin, by.x='PATIENT_ID', by.y='bcr_patient_barcode', all.x=T)
days_per_month <- 365.25/12

info$OS_MONTHS <- as.numeric(info$OS.time) / days_per_month
setnames(info,'OS','OS_STATUS')
info$OS_STATUS <- as.integer(info$OS_STATUS)
info[,c('OS.time'):=NULL]
info$PFI_MONTHS <- as.numeric(info$PFI.time) / days_per_month
setnames(info,'PFI','PFI_STATUS')
info$PFI_STATUS <- as.integer(info$PFI_STATUS)
info[,c('PFI.time'):=NULL]

info$mtstatus2 <- 'WT'
info[mtstatus %in% c('VUS','Truncating'),mtstatus2:='Affected']
info$mtstatus2 <- factor(info$mtstatus2, levels=c('WT','Affected'))
info <- info[!is.na(OS_STATUS) & !is.na(OS_MONTHS),]
info_for_PFI_volcano <- info

## survival plot for univariate, combined VUS and Truncating
OS_for_type <- function(type, dat) {
    tmp <- dat[maintype %in% type,]
    N <- nrow(tmp)
    prop <- sum(tmp$mtstatus2!='WT') / N
    res.cox <- coxph(Surv(OS_MONTHS, OS_STATUS) ~ mtstatus2, data=tmp, iter.max=100)
    out <- as.data.frame(summary(res.cox)$coef)
    names(out) <- c('coef','exp(coef)','se(coef)','z','p')
    out <- cbind(out, adt(exp(confint(res.cox))))
    list(type=type,logHR=out$coef, HR=out$`exp(coef)`, lwr=out$`2.5 %`, upr=out$`97.5 %`,p=out$p, N=N, prop=prop)
}

## VUS vs WT
info_vus_vs_wt <- info[mtstatus %in% c('Wildtype','VUS')]
x <- as.data.frame.matrix(xtabs(~maintype + mtstatus2, data=info_vus_vs_wt))
min.n <- 20
valid_types <- rownames(x[x$WT >= min.n & x$Affected >= min.n,])
res_vus_vs_wt <- rbindlist(lapply(valid_types, OS_for_type, dat=info_vus_vs_wt))
res_vus_vs_wt$q <- p.adjust(res_vus_vs_wt$p,method='BH')
res_vus_vs_wt$comparison <- 'VUS vs WT'


## Truncating vs WT
info_trunc_vs_wt <- info[mtstatus %in% c('Wildtype','Truncating')]
x <- as.data.frame.matrix(xtabs(~maintype + mtstatus2, data=info_trunc_vs_wt))
min.n <- 20
valid_types <- rownames(x[x$WT >= min.n & x$Affected >= min.n,])
res_trunc_vs_wt <- rbindlist(lapply(valid_types, OS_for_type, dat=info_trunc_vs_wt))
res_trunc_vs_wt$q <- p.adjust(res_trunc_vs_wt$p,method='BH')
res_trunc_vs_wt$comparison <- 'Truncating vs WT'

res <- rbind(res_vus_vs_wt, res_trunc_vs_wt)
res$nlog10q <- -log10(res$q)
res$Effect <- 'Not significant'
res[q < 0.1 & logHR < 0, Effect:='Protective']
res[q < 0.1 & logHR > 0, Effect:='Deleterious']
res[Effect!='Not significant',label:=type]

cols <- c('#6ba5c5','#bfbfbf','#ea948d')
names(cols) <- c('Protective','Not significant','Deleterious')

panel_4d <- ggplot(res,aes(x=logHR, y=nlog10q)) +
    scale_x_continuous(expand=c(0,0),breaks=seq(-1,1,by=0.5),limits=c(-1.2,1.2)) +
    geom_point(aes(size=N,fill=prop),color='black',pch=21) +
    geom_text_repel(aes(label=label),min.segment.length=0.01) +
    scale_fill_gradient(low='white',high='#4675b4',name='Prop cases\nwith variant') + 
    geom_vline(xintercept=0,linetype=1) +
    theme_std(base_size=12) +
    theme(legend.position='bottom') +
    labs(x='Effect size',y='-log10(Q)') +
    facet_wrap(facets=~comparison,ncol=2)
panel_4d
if(get_pdfs) ggsave(here('figures/panel_4d.pdf'),width=7,height=4)

```

## 4e 

```{r panel_4e, fig.height=3.9,fig.width=5, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Colorectal Kaplan-Meier for OS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

tcga_clin <- copy(tcga_clin_data)
valid_patients <- tcga_clin$PATIENT_ID

## TCGA clinical data
info <- copy(tcga_sample_hotspot_status_data)
setnames(info,'annotation_prefilter','mtstatus_prefilter')
setnames(info,'annotation_postfilter','mtstatus')
info <- info[mtstatus %in% c('Wildtype','VUS','Truncating') & maintype=='Colorectal Cancer']
info$PATIENT_ID <- strtrim(info$Tumor_Sample_Barcode,12) 
info <- info[,c('PATIENT_ID','mtstatus'),with=F]

## load clinical data from tcga provisional, subset it for the mc3 samples
clin_sample <- copy(colorectal_sample_clin_data)
clin_pt <- copy(colorectal_patient_clin_data)
clin <- merge(clin_sample, clin_pt, by='PATIENT_ID', all.x=T)
clin$PATIENT_ID <- strtrim(clin$SAMPLE_ID,12)
clin <- merge(info, clin, by='PATIENT_ID', all.x=T)
clin <- clin[PATIENT_ID %in% valid_patients,]

## annotate with mutation status from mc3
maf <- copy(mc3_maf_data)
maf <- maf[Hugo_Symbol %in% c('KRAS','NRAS','HRAS','BRAF','APC','SMAD4','TP53')]
maf$important <- maf$oncogenic %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
    maf$`is-a-hotspot`=='Y' | maf$Variant_Classification %in% c('Nonsense_Mutation','Frame_Shift_Del','Frame_Shift_Ins',
                                                                'Splice_Site') & maf$Hugo_Symbol %in% c('APC','SMAD4','TP53')
maf <- maf[important==T,]
maf$PATIENT_ID <- strtrim(maf$Tumor_Sample_Barcode,12)
ras_samples <- maf$PATIENT_ID[maf$Hugo_Symbol %in% c('KRAS','NRAS','HRAS')]
raf_samples <- maf$PATIENT_ID[maf$Hugo_Symbol %in% c('BRAF')]
apc_samples <- maf$PATIENT_ID[maf$Hugo_Symbol %in% c('APC')]
smad4_samples <- maf$PATIENT_ID[maf$Hugo_Symbol %in% c('SMAD4')]
tp53_samples <- maf$PATIENT_ID[maf$Hugo_Symbol %in% c('TP53')]
clin$RAS <- clin$PATIENT_ID %in% ras_samples
clin$RAF <- clin$PATIENT_ID %in% raf_samples
clin$APC <- clin$PATIENT_ID %in% apc_samples
clin$SMAD4 <- clin$PATIENT_ID %in% smad4_samples
clin$TP53 <- clin$PATIENT_ID %in% tp53_samples

x <- clin
x$mtstatus <- factor(x$mtstatus, levels=c('Wildtype','VUS','Truncating'))
x[SITE_OF_TUMOR_TISSUE=='[Not Available]',SITE_OF_TUMOR_TISSUE:=NA]
x <- x[OS_STATUS %in% c('LIVING','DECEASED') & !is.na(OS_MONTHS),]
x$OS_MONTHS <- as.numeric(x$OS_MONTHS)
x[OS_STATUS=='DECEASED',OS_STATUS:='1']
x[OS_STATUS=='LIVING',OS_STATUS:='0']
x$OS_STATUS <- as.integer(x$OS_STATUS)

msi <- copy(pangi_clin_data)
x <- merge(x, msi, by='PATIENT_ID', all.x=T)
x <- x[!is.na(msi_status),]
x$msi_status <- factor(x$msi_status, levels=c('MSS','MSI-L','MSI-H'))
x[AJCC_PATHOLOGIC_TUMOR_STAGE=='Stage I',STAGE:='I']
x[AJCC_PATHOLOGIC_TUMOR_STAGE %in% c('Stage II','Stage IIA','Stage IIB','Stage IIC'),STAGE:='II']
x[AJCC_PATHOLOGIC_TUMOR_STAGE %in% c('Stage III','Stage IIIA','Stage IIIB','Stage IIIC'),STAGE:='III']
x[AJCC_PATHOLOGIC_TUMOR_STAGE %in% c('Stage IV','Stage IVA','Stage IVB'),STAGE:='IV']
x <- x[STAGE!='IV']

x$STAGE <- factor(x$STAGE, levels=c('I','II','III'))
x[SITE_OF_TUMOR_TISSUE=='Rectum',site:='Rectum']
x[PRIMARY_SITE_PATIENT %in% c('Ascending Colon','Cecum','Hepatic Flexure','Transverse Colon'),site:='Right']
x[PRIMARY_SITE_PATIENT %in% c('Descending Colon','Sigmoid Colon','Splenic Flexure'),site:='Left']

absolute <- copy(tcga_absolute)
absolute <- absolute[,c('array','purity'),with=F]
x <- merge(x, absolute, by.x='SAMPLE_ID', by.y='array', all.x=T)

cms <- copy(tcga_cms)
cms <- cms[,c('sample','CMS_network'),with=F]           
names(cms) <- c('PATIENT_ID','cms_type')
cms <- cms[cms_type %nin% c('UNK','NOLBL'),]
x <- merge(x, cms, by='PATIENT_ID', all.x=T)

## survival plot for univariate, combined VUS and Truncating
surv_object <- Surv(time=x$OS_MONTHS, event=x$OS_STATUS)
fit <- survfit(surv_object ~ mtstatus, data = x)
panel_4e <- ggsurvplot(fit, data = x, pval = TRUE, censor=T, 
                       censor.shape=124, censor.size=2)
panel_4e
if(get_pdfs) ggsave(here('figures/panel_4e.pdf'),width=5,height=3.9)

```


## 4f

```{r panel_4fold, fig.width=7.9, fig.height=7.2}
## Cox-PH regression for effect of VUS/truncating vs wild-type on survival
## (results, P13, line 190)

x <- x[!is.na(purity),]

## Cox-PH regression for effect of VUS/truncating vs wild-type on survival
## (results, P13, line 190)
res.cox <- coxph(Surv(OS_MONTHS, OS_STATUS) ~ mtstatus + AGE + STAGE + site + 
                 RAS + RAF + APC + SMAD4 + TP53 + SEX + msi_status + cms_type + purity, data = x, iter.max=100)

out <- summary(res.cox)$coef
out <- cbind(covariate=rownames(out),adt(out))
out <- cbind(out, adt(exp(confint(res.cox))))
n_wt <- sum(x$mtstatus=='Wildtype')
n_vus <- sum(x$mtstatus=='VUS')
n_trunc <- sum(x$mtstatus=='Truncating')
n_age <- nrow(x)
n_stageI <- sum(x$STAGE=='I')
n_stageII <- sum(x$STAGE=='II')
n_stageIII <- sum(x$STAGE=='III')
n_siteLeft <- sum(x$site=='Left',na.rm=T)
n_siteRectum <- sum(x$site=='Rectum',na.rm=T)
n_siteRight <- sum(x$site=='Right',na.rm=T)
n_RAS <- sum(x$RAS==T)
n_RAF <- sum(x$RAF==T)
n_APC <- sum(x$APC==T)
n_SMAD4 <- sum(x$SMAD4==T)
n_TP53 <- sum(x$TP53==T)
n_SEXFemale <- sum(x$SEX=='Female')
n_SEXMale <- sum(x$SEX=='Male')
n_msi_statusMSS <- sum(x$msi_status=='MSS')
n_msi_statusMSIl <- sum(x$msi_status=='MSI-L')
n_msi_statusMSIh <- sum(x$msi_status=='MSI-H')
n_cms_typeCMS1 <- sum(x$cms_type=='CMS1',na.rm=T)
n_cms_typeCMS2 <- sum(x$cms_type=='CMS2',na.rm=T)
n_cms_typeCMS3 <- sum(x$cms_type=='CMS3',na.rm=T)
n_cms_typeCMS4 <- sum(x$cms_type=='CMS4',na.rm=T)
n_purity <- sum(!is.na(x$purity))

dat <- list(mtstatusWT=n_wt, mtstatusVUS=n_vus, mtstatusTruncating=n_trunc, AGE=n_age, 
            STAGEIII=n_stageIII, STAGEII=n_stageII, STAGEI=n_stageI, 
            siteLeft=n_siteLeft, siteRectum=n_siteRectum, siteRight=n_siteRight, 
            RASTRUE=n_RAS, RAFTRUE=n_RAF, APCTRUE=n_APC, SMAD4TRUE=n_SMAD4, TP53TRUE=n_TP53, 
            SEXFemale=n_SEXFemale, SEXMale=n_SEXMale,
            msi_statusMSS=n_msi_statusMSS, `msi_statusMSI-L`=n_msi_statusMSIl,
            `msi_statusMSI-H`=n_msi_statusMSIh, 
            cms_typeCMS1=n_cms_typeCMS1, cms_typeCMS2=n_cms_typeCMS2, 
            cms_typeCMS3=n_cms_typeCMS3, cms_typeCMS4=n_cms_typeCMS4, purity=n_purity) 
Ns <- data.table(covariate=names(dat), N=unlist(dat))
Ns$pos <- 1:nrow(Ns)

dat <- merge(Ns, out[,c(1,3,7,8,6),with=F], by='covariate', all.x=T)
dat$covariate <- gsub(' ','_',dat$covariate)
dat$covariate <- gsub('-','_',dat$covariate)
names(dat) <- c('covariate','N','pos','HR','lwr','upr','p')
dat[is.na(HR),covariate:=paste0(covariate,'_Reference')]
dat[is.na(HR),HR:=1]
dat <- dat[order(pos),]

dat$covariate <- gsub('mtstatus','',dat$covariate)
dat$covariate <- gsub('AGE','Age',dat$covariate)
dat$covariate <- gsub('STAGE','',dat$covariate)
dat$covariate <- gsub('site','',dat$covariate)
dat$covariate <- gsub('TRUE','',dat$covariate)
dat$covariate <- gsub('SEX','',dat$covariate)
dat$covariate <- gsub('msi_status','',dat$covariate)
dat$covariate <- gsub('cms_type','',dat$covariate)

dat$covariate <- factor(dat$covariate, levels=rev(dat$covariate))
dat$Significance <- 'No'
dat[upr < 1 | lwr > 1, Significance:='Yes']
dat[is.na(p),Significance:='NA (Reference)']
cols <- c('#bfbfbf','black','#4675b4')
names(cols) <- c('NA (Reference)','No','Yes')

panel_4f <- ggplot(dat,aes(x=covariate, y=HR)) +
    geom_hline(yintercept=1) +
    geom_point(aes(size=N,color=Significance),fill=NA,pch=21) +
    geom_errorbar(aes(min=lwr,max=upr,color=Significance),width=0.5) +
    scale_color_manual(values=cols,name='Significance') +
    coord_flip() +
    theme_std(base_size=12) +
    scale_y_continuous(trans=log10_trans(),limits=c(0.01,100),
                       breaks=c(0.01,0.1,1,10,100),labels=c(0.01,0.1,1.0,10,100),position='right') +
    labs(x=NULL,y='Hazard ratio (overall survival)')
panel_4f

if(get_pdfs) ggsave(here('figures/panel_4f.pdf'),width=7.9,height=7.2)
```


***

# Extended data figures

## EDF 1a
```{r panel_EDF1, fig.width=8, fig.height=16, echo=F, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# TCGA cohort demographics
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## tcga data
info <- copy(tcga_clin_data)
info$sample_type <- substr(info$Tumor_Sample_Barcode,14,15)
info[sample_type=='01',sample_type:='Primary Solid Tumor']
info[sample_type=='02',sample_type:='Recurrent Solid Tumor']
info[sample_type=='03',sample_type:='Primary Blood Derived Cancer - Peripheral Blood']
info[sample_type=='06',sample_type:='Metastatic']
info$normal_type <- substr(info$Matched_Norm_Sample_Barcode,14,15)
info[normal_type=='10',normal_type:='Blood Derived Normal']
info[normal_type=='11',normal_type:='Solid Tissue Normal']
info[normal_type=='12',normal_type:='Buccal Cell Normal']
clin1 <- copy(tcga_cdr_supptab1)
clin1 <- clin1[,c('bcr_patient_barcode','age_at_initial_pathologic_diagnosis','gender','race','histological_type','ajcc_pathologic_tumor_stage'),with=F]
clin2 <- copy(tcga_pancan_clin)
clin2 <- clin2[,c('bcr_patient_barcode','pathologic_stage'),with=F]

tx_history <- copy(tcga_tx_history)
prior_tx_yes_patients <- tx_history$Tumor_Sample_Barcode[tx_history$prior_tx=='yes']
prior_tx_unknown_patients <- tx_history$Tumor_Sample_Barcode[tx_history$prior_tx=='unknown']
info[PATIENT_ID %in% prior_tx_yes_patients, prior_tx:='Yes']
info[PATIENT_ID %in% prior_tx_unknown_patients, prior_tx:='n/a']
info[is.na(prior_tx),prior_tx:='No']
info$prior_tx <- factor(info$prior_tx, levels=c('n/a','Yes','No'))

## add additional clinical info
info <- merge(info, clin1, by.x='PATIENT_ID', by.y='bcr_patient_barcode', all.x=T)
info <- merge(info, clin2, by.x='PATIENT_ID', by.y='bcr_patient_barcode', all.x=T)
info[is.na(gender),gender:='n/a']
info[race %in% c('[Not Available]','[Not Evaluated]','[Unknown]'),race:='n/a']
info[race %in% c('AMERICAN INDIAN OR ALASKA NATIVE','NATIVE HAWAIIAN OR OTHER PACIFIC ISLANDER'),race:='Other']
info[race=='WHITE',race:='White']
info[race=='BLACK OR AFRICAN AMERICAN',race:='Black']
info[race=='ASIAN',race:='Asian']

info[pathologic_stage %in% c('Stage I','Stage IA','Stage IB'),stage:='I']
info[pathologic_stage %in% c('Stage II','Stage IIA','Stage IIB','Stage IIC'),stage:='II']
info[pathologic_stage %in% c('Stage III','Stage IIIA','Stage IIIB','Stage IIIC'),stage:='III']
info[pathologic_stage %in% c('Stage IV','Stage IVA','Stage IVB','Stage IVC'),stage:='IV']
info[pathologic_stage %in% c('[Not Applicable]'),stage:='Does not apply']
info[pathologic_stage %in% c('','[Not Available]','[Unknown]') | is.na(pathologic_stage),stage:='n/a']
info[is.na(stage),stage:='Other']
info$stage <- factor(info$stage, levels=c('n/a','I','II','III','IV','Other','Does not apply'))
demographics_info <- copy(info)

hs <- copy(tcga_sample_hotspot_status_data)
hs <- hs[,c('Tumor_Sample_Barcode','maintype','annotation_postfilter','coverage_tumornormal'),with=F]
hs[coverage_tumornormal < 0.9, annotation_postfilter:='Unknown']
hs <- as.data.frame.matrix(xtabs(~maintype + annotation_postfilter, data=hs))
hs$Unknown <- NULL
total_covered <- rowSums(hs)
hs <- hs / total_covered
hs[total_covered <= 10,] <- NA
hs <- cbind(maintype=rownames(hs), as.data.table(hs))
hs$`Insufficient samples` <- 0
hs[is.na(Truncating),`Insufficient samples`:=1]
hs[is.na(hs)] <- 0
hs <- hs[order(1-`Insufficient samples`,Wildtype,decreasing=T),]
hs$maintype <- factor(hs$maintype, levels=(hs$maintype))
hs <- hs[order(maintype),]
hs <- melt(hs, id.var='maintype')
hs$variable <- factor(hs$variable, levels=c('Wildtype','VUS','Truncating','Insufficient samples'))
names(hs) <- c('maintype','variant','prop')


## variant
x <- hs
cols <- c('#bfbfbf',brewer.pal(8,'Accent')[c(6,3,2)])
names(cols) <- rev(levels(x$variant))
p_variant <- ggplot(x, aes(x=maintype,y=100*prop)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity',aes(fill=variant)) + 
    scale_fill_manual(values=cols,name='mtDNA genotype') + 
    theme_std(base_size=10) +
    theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1),legend.position='bottom') + 
    labs(x=NULL,y='mtDNA status (%)')
p_variant <- extract_gglegend(p_variant)
levels <- levels(hs$maintype)

## sample type
d <- info
d$type <- d$sample_type
d[type!='Metastatic',type:='Primary']
d[type %in% c('Metastatic'),type:='Metastasis']
x <- (xtabs(~maintype + type, data=d))
d <- rowSums(x)
d <- data.table(type=names(d),N=d)
d <- d[order(N,decreasing=F),]
x <- as.data.table(reshape2::melt(x))
#levels <- d$type
x$maintype <- factor(x$maintype, levels=levels)
x$type <- factor(x$type, levels=rev(c('Primary','Metastasis')))
cols <- brewer.pal(8,'Accent')[c(6,1)]
names(cols) <- levels(d$label)
p_type <- ggplot(x, aes(x=maintype,y=value)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity',aes(fill=type)) +
    scale_fill_manual(values=cols,name='Sample type') + 
    theme_std(base_size=10) +
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.position='bottom') +
    labs(x=NULL,y='Cases')
p_type <- extract_gglegend(p_type)

## mtDNA coverage
x <- info[,c('maintype','coverage_tumornormal'),with=F]
x$coverage <- cut(100*x$coverage_tumornormal,c(0,10,25,50,75,90,100),include.lowest=T)
x$coverage <- factor(x$coverage, levels=rev(levels(x$coverage)))
x <- (xtabs(~maintype + coverage, data=x))
x <- x / rowSums(x)
x <- as.data.table(reshape2::melt(x))
x$maintype <- factor(x$maintype, levels=levels)
cols <- brewer.pal(8,'PRGn')
names(cols) <- levels(x$coverage)
p_coverage <- ggplot(x, aes(x=maintype,y=100*value)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity',aes(fill=coverage)) + 
    scale_fill_manual(values=cols,name='mtDNA coverage') + 
    theme_std(base_size=10) +
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.position='bottom') +
    labs(x=NULL,y='Coverage (%)')
p_coverage <- extract_gglegend(p_coverage)

## age (later: age vs mtdna)
x <- info[,c('maintype','age_at_initial_pathologic_diagnosis'),with=F]
x$age_bin <- cut(x$age_at_initial_pathologic_diagnosis,c(0,20,35,45,55,65,75,85,130),include.lowest=T)
x$age_bin <- factor(x$age_bin, levels=rev(levels(x$age_bin)))
x <- (xtabs(~maintype + age_bin, data=x))
x <- x / rowSums(x)
x <- as.data.table(reshape2::melt(x))
x$maintype <- factor(x$maintype, levels=levels)
cols <- brewer.pal(8,'PRGn')
names(cols) <- levels(x$age_bin)
p_age <- ggplot(x, aes(x=maintype,y=100*value)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity',aes(fill=age_bin)) + 
    scale_fill_manual(values=cols,name='Age at diagnosis') + 
    theme_std(base_size=10) +
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.position='bottom') +
    labs(x=NULL,y='Age (%)')
p_age <- extract_gglegend(p_age)

## stage (later: age vs mtdna)
x <- info[,c('maintype','stage'),with=F]
x$stage <- factor(x$stage, levels=c('n/a','Does not apply','Other','IV','III','II','I'))
x <- (xtabs(~maintype + stage, data=x))
x <- x / rowSums(x)
x <- as.data.table(reshape2::melt(x))
x$maintype <- factor(x$maintype, levels=levels)
cols4 <- brewer.pal(9,'OrRd')[c(8,6,4,2)]
cols <- c('#bfbfbf','#f0027f','black',cols4)
names(cols) <- levels(x$stage)
p_stage <- ggplot(x, aes(x=maintype,y=100*value)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity',aes(fill=stage)) + 
    scale_fill_manual(values=cols,name='Pathologic stage') + 
    theme_std(base_size=10) +
    #theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1)) +
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.position='bottom') +
    labs(x=NULL,y='Stage (%)')
p_stage <- extract_gglegend(p_stage)

## gender
x <- info[,c('maintype','gender'),with=F]
x[gender=='FEMALE',gender:='Female']
x[gender=='MALE',gender:='Male']
x$gender <- factor(x$gender, levels=c('Female','Male','n/a'))
x <- (xtabs(~maintype + gender, data=x))
x <- x / rowSums(x)
x <- as.data.table(reshape2::melt(x))
x$maintype <- factor(x$maintype, levels=levels)
cols <- c('#386CB0','#A6DBA0','#bfbfbf')
names(cols) <- levels(x$gender)
p_gender <- ggplot(x, aes(x=maintype,y=100*value)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity',aes(fill=gender)) + 
    scale_fill_manual(values=cols,name='Gender') + 
    theme_std(base_size=10) +
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.position='bottom') +
    labs(x=NULL,y='Gender (%)')
p_gender <- extract_gglegend(p_gender)

## treatment
x <- info[,c('PATIENT_ID','maintype','prior_tx'),with=F]
collapse_type <- function(x) {
    n <- nrow(x)
    prior_tx <- sum(x$prior_tx=='Yes') / n 
    no_prior_tx <- sum(x$prior_tx=='No') / n
    unknown <- sum(x$prior_tx=='n/a') / n
    list(prior_tx=prior_tx, no_prior_tx=no_prior_tx, unknown=unknown)
}
res <- x[,collapse_type(.SD),by=c('maintype')]
res <- melt(res, id.var='maintype')
names(res) <- c('maintype','status','prop')
res$status <- factor(res$status, levels=c('no_prior_tx','prior_tx','unknown'))
res$maintype <- factor(res$maintype, levels=levels)

cols <- c('#BEAED4','#F0027F','#bfbfbf')
names(cols) <- levels(res$status)
p_prior_tx <- ggplot(res, aes(x=maintype,y=100*prop)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity',aes(fill=status)) + 
    scale_fill_manual(values=cols,name='History of treatment') + 
    theme_std(base_size=10) +
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.position='bottom') +
    labs(x=NULL,y='Treatment (%)')
p_prior_tx <- extract_gglegend(p_prior_tx)

## race
x <- info[,c('maintype','race'),with=F]
x$race <- factor(x$race, levels=(c('White','Black','Asian','Other','n/a')))
x <- (xtabs(~maintype + race, data=x))
x <- x / rowSums(x)
x <- as.data.table(reshape2::melt(x))
x$maintype <- factor(x$maintype, levels=levels)
cols <- c(brewer.pal(8,'Accent')[c(5:6,1,3)],'#bfbfbf')
names(cols) <- levels(x$race)
p_race <- ggplot(x, aes(x=maintype,y=100*value)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity',aes(fill=race)) + 
    scale_fill_manual(values=cols,name='Race') + 
    theme_std(base_size=10) +
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.position='bottom') +
    labs(x=NULL,y='Race (%)')
p_race <- extract_gglegend(p_race)

## combine plots
p_plot <- plot_grid(p_type$plot, p_stage$plot, p_prior_tx$plot, p_gender$plot,p_race$plot,p_age$plot, p_coverage$plot, p_variant$plot, ncol=1,align='v', rel_heights=c(1,1,1,1,1,1,1,2.4))
p_legend <- plot_grid(p_type$legend, p_stage$legend, p_prior_tx$legend, p_gender$legend,p_race$legend,p_age$legend, p_coverage$legend, p_variant$legend, ncol=1,align='v', rel_heights=c(1,1,1,1,1,1,1,1))
p <- plot_grid(p_plot, p_legend, ncol=1, rel_heights=c(1,1))
p
if(get_pdfs==T) ggsave(here('figures/panel_EDF1a.pdf'),width=8,height=12)
```


## EDF 1b
```{r panel_EDF1b,fig.height=4.5,fig.width=6.5, echo=F, eval=T}

# compare mtCN per cancer type, compare to coverage
sample_info <- copy(tcga_clin_data)
codes <- copy(tcga_center_codes)
codes$Code <- as.character(codes$Code)
codes[nchar(Code)==1,Code:=paste0('0',Code)]
codes <- codes[,c('Code','Short Name'),with=F]
sample_info$center <- substr(sample_info$Tumor_Sample_Barcode,27,28)
sample_info <- merge(sample_info, codes, by.x='center', by.y='Code',all.x=T)
sample_info[,center:=NULL]
setnames(sample_info,'Short Name','center')

summarize_type <- function(dat,value) {
    x <- dat[[value]]
    x <- x[!is.na(x)]
    mu <- mean(x)
    s <- sd(x)
    n <- length(x)
    sem <- s / sqrt(n)
    lwr <- mu - 1.96*sem
    upr <- mu + 1.96*sem
    out <- list(mu=mu,n=n,lwr=lwr,upr=upr)
    names(out) <- paste0(names(out),'_',value)
    out
}

info <- sample_info[,summarize_type(.SD,'coverage_tumor'),by=c('center','maintype')]
setnames(info,'n_coverage_tumor','n')
info <- info[n >= 20,]
center_tbl <- table.freq(info$center)
valid_centers <- center_tbl$value[center_tbl$N >= 5]
info <- info[center %in% valid_centers,]
cols <- brewer.pal(3,'Accent')
names(cols) <- unique(info$center)

p1 <- ggplot(info,aes(x=maintype,y=mu_coverage_tumor)) +
    geom_point(aes(fill=center),pch=21,color='black',size=3) +
    scale_fill_manual(values=cols,name='Center') + 
    theme_std(base_size=12) +
    labs(x=NULL,y='Avg tumor mtDNA coverage') + 
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
          panel.grid.major=element_line(color='grey92'),
          legend.position='none')

p2 <- ggplot(info,aes(x=mu_coverage_tumor)) +
    geom_density(aes(fill=center),alpha=0.75) +
    scale_fill_manual(values=cols,name='Center') + 
    theme_std(base_size=12) +
    theme(axis.text.y=element_blank(),axis.line.y=element_blank(),axis.ticks.y=element_blank()) + 
    coord_flip() +
    labs(x=NULL,y='Density')

p <- plot_grid(p1,p2,align='h',rel_widths=c(5,3),nrow=1)
p
if(get_pdfs==T) ggsave(here('figures/panel_EDF1b.pdf'),width=6.5,height=4.5)
```

## EDF 1c
```{r panel_EDF1c, fig.width=7, fig.height=3, echo=F, eval=T}

process <- function(cov) {
    mat <- as.matrix(cov[,2:ncol(cov),with=F])
    pos <- as.integer(colnames(mat))
    tbl <- colSums(mat,na.rm=T)
    data.table(pos=pos,samples_covd=as.integer(tbl))
}

res_t <- process(cov_t)
res_n <- process(cov_n)
res_tn <- process(cov_tn)
res20_t <- process(cov20_t)
res20_n <- process(cov20_n)
res20_tn <- process(cov20_tn)
n_pcawg <- 2658

res5 <- merge(res_t, res_n, by='pos', all=T)
names(res5) <- c('pos','Tumor','Normal')
res5 <- merge(res5, res_tn, by='pos', all=T)
setnames(res5,'samples_covd','Tumor+Normal')
res5$PCAWG <- n_pcawg
res5 <- reshape2::melt(res5,id.vars='pos')
res5$Depth <- 'Min depth = 5'

res20 <- merge(res20_t, res20_n, by='pos', all=T)
names(res20) <- c('pos','Tumor','Normal')
res20 <- merge(res20, res20_tn, by='pos', all=T)
setnames(res20,'samples_covd','Tumor+Normal')
res20$PCAWG <- n_pcawg
res20 <- reshape2::melt(res20,id.vars='pos')
res20$Depth <- 'Min depth = 20'

res <- rbind(res5, res20)
res$Depth <- factor(res$Depth, levels=c('Min depth = 5','Min depth = 20'))
res$variable <- factor(res$variable, levels=c('Tumor','Normal','Tumor+Normal','PCAWG'))
cols <- brewer.pal(4,'Set1')[1:4]
names(cols) <- c('Tumor','Normal','Tumor+Normal','PCAWG')
p <- ggplot(res,aes(x=pos,y=value)) +
    geom_line(aes(color=variable)) +
    scale_color_manual(values=cols,name='Sample Type') +
    facet_wrap(facets=~Depth,nrow=2) +
    theme_std(base_size=12) +
    theme(legend.position='right')
p
if(get_pdfs==T) ggsave(here('figures/panel_EDF1c.pdf'),width=7,height=3)
```


## EDF 1d
```{r panel_EDF1d, fig.width=6.6, fig.height=4.2, echo=F, eval=T} 

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# confirm that cancer types with more coverage dont have 
# a higher proportion of samples w/mutations
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## is there correlation of [% well-covered] and [% well-covered with mutations]?
x <- copy(tcga_sample_hotspot_status_data)
d <- copy(tcga_maf_data)
d <- d[class=='somatic',]
tbl <- table.freq(d$Tumor_Sample_Barcode)
x <- merge(x, tbl, by.x='Tumor_Sample_Barcode', by.y='value', all.x=T)
x[is.na(N), N:=0]

summarize_type <- function(x) {
    n_wellcovered <- sum(x$coverage_tumornormal >= 0.9)
    n_total <- nrow(x)
    n_wellcovered_with_somatic_muts <- sum(x$coverage_tumornormal >= 0.9 & x$N > 0)
    prop_wellcovered <- n_wellcovered / n_total
    prop_wellcovered_with_somatic_muts <- n_wellcovered_with_somatic_muts / n_wellcovered
    list(n_total=n_total, n_wellcovered=n_wellcovered, n_wellcovered_with_somatic_muts=n_wellcovered_with_somatic_muts,
         prop_wellcovered=prop_wellcovered, prop_wellcovered_with_somatic_muts=prop_wellcovered_with_somatic_muts)
}
info <- x[,summarize_type(.SD),by=maintype]
info <- info[n_wellcovered >= 30,]
r <- cor(info$prop_wellcovered, info$prop_wellcovered_with_somatic_muts)

fit <- lm(prop_wellcovered_with_somatic_muts ~ prop_wellcovered, data=info)
m <- summary(fit)$coef
p <- m[2,4]
plab <- plabel(p)
panel_EDF1d <- ggplot(info, aes(x=100*prop_wellcovered, y=100*prop_wellcovered_with_somatic_muts)) +
    geom_smooth(method='lm') +
    geom_point(aes(size=n_wellcovered)) +
    scale_y_continuous(limits=c(0,100)) + 
    scale_x_continuous(limits=c(0,100)) +
    geom_text_repel(aes(label=maintype)) +
    theme_std(base_size=12) +
    geom_text(data=info[1,], x=85, y=90,label=plab,color='red',size=6) + 
    labs(x='% of samples that are well-covered',y='% of well-covered samples\nwith somatic mutations')
panel_EDF1d
if(get_pdfs==T) ggsave(here('figures/panel_EDF1d.pdf'),width=6.6,height=4.2)

```


## EDF 2
```{r panel_EDF2, fig.width=7.5, fig.height=4, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# strand-specific signatures
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

dt <- copy(tcga_maf_data)
dt <- dt[Variant_Type=='SNP',]
dt <- dt[,c('Tumor_Sample_Barcode','ShortVariantID','Reference_Allele','Tumor_Seq_Allele2','flanking_bps'),with=F]
dt$ref_norm <- dt$Reference_Allele
dt$alt_norm <- dt$Tumor_Seq_Allele2
dt$strand <- 'L'

dt[ref_norm=='A' & alt_norm=='C',alt_norm:='g']
dt[ref_norm=='A' & alt_norm=='G',alt_norm:='c']
dt[ref_norm=='A' & alt_norm=='T',alt_norm:='a']
dt[ref_norm=='A',strand:='H']
dt[ref_norm=='A',ref_norm:='t']

dt[ref_norm=='G' & alt_norm=='A',alt_norm:='t']
dt[ref_norm=='G' & alt_norm=='C',alt_norm:='g']
dt[ref_norm=='G' & alt_norm=='T',alt_norm:='a']
dt[ref_norm=='G',strand:='H']
dt[ref_norm=='G',ref_norm:='c']

dt$ref_norm <- toupper(dt$ref_norm)
dt$alt_norm <- toupper(dt$alt_norm)
dt$sig <- paste(dt$ref_norm,dt$alt_norm,sep='>')
dt$trinuc <- paste0(substr(dt$flanking_bps,1,1), dt$sig,substr(dt$flanking_bps,3,3))

x <- adt(expand.grid(c('A','C','G','T'),c('C','T'),c('A','C','G','T'),c('A','C','T','G'),stringsAsFactors=F))
x <- x[Var2!=Var3,]
x$sig <- paste(x$Var2,x$Var3,sep='>')
x$trinuc <- paste0(x$Var1,x$sig,x$Var4)

tbl <- table.freq(dt$trinuc[dt$strand=='H'])
x <- merge(x, tbl, by.x='trinuc', by.y='value', all.x=T)
setnames(x,'N','N_H')
x[is.na(N_H),N_H:=0]

tbl <- table.freq(dt$trinuc[dt$strand=='L'])
x <- merge(x, tbl, by.x='trinuc', by.y='value', all.x=T)
setnames(x,'N','N_L')
x[is.na(N_L),N_L:=0]

x <- x[order(sig,Var1,Var4)]
tmp <- x[order(N_H,decreasing=T)]
label_H <- tmp$trinuc[1:4]
tmp <- x[order(N_L,decreasing=T)]
label_L <- tmp$trinuc[1:2]
labels <- c(label_H,label_L)

xm <- melt(x[,c('trinuc','sig','Var1','Var4','N_H','N_L'),with=F],id.var=c('trinuc','sig','Var1','Var4'))
xm$trinuc <- factor(xm$trinuc, levels=x$trinuc)
xm$label <- paste0(xm$Var1,xm$sig,xm$Var4)
xm[variable=='N_H' & label %nin% label_H,label:=NA]
xm[variable=='N_L' & label %nin% label_L,label:=NA]
xm$Strand <- gsub('N_','',xm$variable)

cols <- c('#739fbd','#fb998e')
names(cols) <- c('H','L')
cols2 <- brewer.pal(6,'Set3')

p1 <- ggplot(xm,aes(x=trinuc,y=value)) + 
    scale_y_continuous(expand=c(0,0)) +
    geom_bar(stat='identity',aes(fill=Strand),position=position_dodge(width=1)) +
    scale_fill_manual(values=cols,name='Strand') +
    geom_text_repel(aes(label=label),min.segment.length=0.001) +
    theme_std(base_size=16) +
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.position='bottom') +
    labs(x='96 substitution signatures',y='# observed')

p1 <- extract_gglegend(p1)

p2 <- ggplot(xm,aes(x=trinuc,y=1)) + 
    geom_tile(aes(fill=sig)) +
    scale_fill_manual(values=cols2,name='Mutation signature') +
    theme_nothing()+ 
    theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.position='bottom') 

panel_EDF2 <- plot_grid(p1$plot,p2,align='v',rel_heights=c(3,1),ncol=1)
panel_EDF2 <- plot_grid(panel_EDF2,p1$legend,rel_heights=c(4,1),ncol=1)
panel_EDF2
if(get_pdfs) ggsave(here('figures/panel_EDF2.pdf'),width=7.5,height=4)
```

## EDF 3a
```{r panel_EDF3a, echo=F, eval=T, fig.width=2.5, fig.height=5}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# % samples with truncating mutations
# in tumor-only vs blood-normal in tcga
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
library(binom)

## tcga data
clin <- copy(tcga_clin_data)
clin <- clin[coverage_tumor >= 0.9]
clin$normal_type <- substr(clin$Matched_Norm_Sample_Barcode,14,15)
clin[normal_type=='10',normal_type:='Blood Derived Normal']
clin[normal_type=='11',normal_type:='Solid Tissue Normal']


d <- copy(tcga_maf_data) 
d <- d[truncating==T & t_depth >= 20,]
x_somatic <- d$TumorVAF

dn <- copy(tcga_maf_normaltrunc_data)
normal_blood_derived <- clin$Matched_Norm_Sample_Barcode[clin$normal_type=='Blood Derived Normal']
normal_solid_tissue <- clin$Matched_Norm_Sample_Barcode[clin$normal_type=='Solid Tissue Normal']
x_normal_blood_derived <- dn$TumorVAF[dn$Tumor_Sample_Barcode %in% normal_blood_derived]
x_normal_solid_tissue <- dn$TumorVAF[dn$Tumor_Sample_Barcode %in% normal_solid_tissue]

x <- rbind(data.table(VAF=x_somatic,type='TCGA Somatic'), data.table(VAF=x_normal_blood_derived,type='TCGA Blood-derived normal'), data.table(VAF=x_normal_solid_tissue,type='TCGA Solid-tissue normal'))

#wilcox.test(x_somatic, x_normal_blood_derived)$p.value
#wilcox.test(x_somatic, x_normal_solid_tissue)$p.value
#wilcox.test(x_normal_blood_derived, x_normal_solid_tissue)$p.value

p <- ggplot(x, aes(x=type, y=VAF)) +
    scale_y_continuous(expand=c(0,0),limits=c(0,1),breaks=seq(0,1,by=0.25)) +
    geom_point(color='steelblue',position=position_jitter(width=0.15)) +
    geom_boxplot(fill=NA,outlier.shape=NA) +
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)) +
    labs(x=NULL,y='% heteroplasmy (truncating variants)')
p
if(get_pdfs) ggsave(here('figures/panel_EDF3a.pdf'),width=2.5,height=5)
```


## EDF 3b
```{r panel_EDF3b, echo=F, eval=T, fig.width=5.5, fig.height=2}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# % samples with truncating mutations
# in tumor-only vs blood-normal in tcga
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
library(binom)

## tcga data
clin <- copy(tcga_clin_data)
clin <- clin[coverage_tumor >= 0.9]
clin$normal_type <- substr(clin$Matched_Norm_Sample_Barcode,14,15)
clin[normal_type=='10',normal_type:='Blood Derived Normal']
clin[normal_type=='11',normal_type:='Solid Tissue Normal']
clin[normal_type=='12',normal_type:='Buccal Cell Normal']

d <- copy(tcga_maf_data)
d <- d[Variant_Classification %in% truncating_classes & Tumor_Sample_Barcode %in% clin$Tumor_Sample_Barcode,]
x_somatic_trunc <- length(unique(d$Tumor_Sample_Barcode))
n_overall <- nrow(clin)
ci_tcga_somatic <- binom.confint(x_somatic_trunc, n_overall, method='exact')
ci_tcga_somatic <- adt(ci_tcga_somatic)
ci_tcga_somatic$class <- 'TCGA Somatic'

## run for tcga-normals
frac <- copy(tcga_frac_callable_data)
frac <- frac[min_alt=='Min depth = 5',c(1,2,3),with=F]
frac <- dcast(frac, Tumor_Sample_Barcode ~ sample, value.var='frac')
clin <- merge(clin, frac, by='Tumor_Sample_Barcode', all.x=T)
dn <- copy(tcga_maf_normaltrunc_data)

normal_blood_derived <- clin$Matched_Norm_Sample_Barcode[clin$normal_type=='Blood Derived Normal' & clin$Normal >= 0.9]
normal_blood_derived <- normal_blood_derived[!is.na(normal_blood_derived)]
n_normal_blood <- length(normal_blood_derived)
x_normal_blood <- length(unique(dn$Tumor_Sample_Barcode[dn$Variant_Classification %in% truncating_classes & dn$Tumor_Sample_Barcode %in% normal_blood_derived]))
ci_normal_blood <- binom.confint(x_normal_blood, n_normal_blood, method='exact')
ci_normal_blood <- adt(ci_normal_blood)
ci_normal_blood$class <- 'TCGA Blood-derived normal'

solid_tissue_derived <- clin$Matched_Norm_Sample_Barcode[clin$normal_type=='Solid Tissue Normal' & clin$Normal >= 0.9]
solid_tissue_derived <- solid_tissue_derived[!is.na(solid_tissue_derived)]
n_solid_tissue <- length(solid_tissue_derived)
x_solid_tissue <- length(unique(dn$Tumor_Sample_Barcode[dn$Variant_Classification %in% truncating_classes & dn$Tumor_Sample_Barcode %in% solid_tissue_derived]))
ci_solid_tissue <- binom.confint(x_solid_tissue, n_solid_tissue, method='exact')
ci_solid_tissue <- adt(ci_solid_tissue)
ci_solid_tissue$class <- 'TCGA Solid-tissue normal'

## helix data
x_helix_trunc <- 291 ## total number of samples with truncating variants
n_helix <- 195983 ## total number of samples
ci_helix <- binom.confint(x_helix_trunc, n_helix, method='exact') ## assumes max of 1 truncating per sample
ci_helix <- adt(ci_helix)
ci_helix$class <- 'Helix Germline'

## get % affected in each
ci <- rbind(ci_tcga_somatic, ci_normal_blood, ci_solid_tissue, ci_helix)
ci$class <- factor(ci$class, levels=rev(c('TCGA Somatic','TCGA Solid-tissue normal','TCGA Blood-derived normal','Helix Germline')))
ci <- ci[order(class),]
ci$mean <- 100*ci$mean
ci$lower <- 100*ci$lower
ci$upper <- 100*ci$upper
setnames(ci,'mean','pct')

## include p-values:
getp <- function(vals,ci) {
    a <- vals[1]; b <- vals[2]
    tst <- prop.test(ci$x[c(a,b)],ci$n[c(a,b)])
    plab <- plabel(tst$p.value)
    list(a=ci$class[a],b=ci$class[b],plab=plab)
}
tmp <- rbindlist(lapply(list(c(1,2),c(1,3),c(1,4),c(2,3),c(2,4),c(3,4)), getp, ci))

pctinfo <- break_axis(ci$pct,maxlower=2.5,lowerticksize=0.5,minupper=8,upperticksize=1,ratio_lower_to_upper=0.5)
lwrinfo <- break_axis(ci$lower,maxlower=2.5,lowerticksize=0.5,minupper=8,upperticksize=1,ratio_lower_to_upper=0.5)
uprinfo <- break_axis(ci$upper,maxlower=2.5,lowerticksize=0.5,minupper=8,upperticksize=1,ratio_lower_to_upper=0.5)
ci$newpct <- pctinfo$newy
ci$newlower <- lwrinfo$newy
ci$newupper <- uprinfo$newy

p <- ggplot(ci,aes(x=class,y=newpct)) +
    scale_y_continuous(expand=c(0,0),labels=pctinfo$labels, limits=c(pctinfo$limits[1],4.5),breaks=pctinfo$breaks,position='right') +
    geom_bar(stat='identity',fill='#3f99bc') +
    geom_errorbar(aes(min=newlower,max=newpct),width=NA,color='white') +
    geom_errorbar(aes(min=newpct,max=newupper),width=NA,color='black') +
    theme_std(base_size=12) +
    theme(legend.position='none') +
    labs(x=NULL,y='% cases with truncating variant') +
    coord_flip()
p
if(get_pdfs) ggsave(here('figures/panel_EDF3b.pdf'),width=5.5,height=2)
```

## EDF 3c
```{r panel_edf3c, fig.width=2, fig.height=4, eval=T, echo=F} 

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# DNA/RNA heteroplasmy correlation among tumor-only truncating variants
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

dt <- copy(tcga_maf_data)
dt[Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del') & homopolymer!='',ShortVariantID:=homopolymer]
dt <- dt[,c('Tumor_Sample_Barcode','ShortVariantID','class','Variant_Classification','Variant_Type','TumorVAF','TumorVAF_rna','t_alt_count','t_depth','t_depth_rna','truncating','class'),with=F]
dt$rna_available <- dt$t_depth_rna >= 5 & !is.na(dt$t_depth_rna)
dt <- dt[rna_available==T,]
dt <- dt[truncating==T & class=='unknown-truncating',]
dt$id <- paste(dt$Tumor_Sample_Barcode,dt$ShortVariantID)
dt$in_rna <- 'n/a'
dt[TumorVAF_rna > 0.05 & rna_available==T,in_rna:='Yes']
dt[TumorVAF_rna <= 0.05 & rna_available==T,in_rna:='No']

t2 <- table.freq(dt$in_rna[dt$rna_available==T])
t2$prop <- t2$N/sum(t2$N)
t2$label <- paste0(prettyNum(t2$prop*100,digits=3),'% (N=',t2$N,')')
t2$label[2] <- NA
t2$group <- 'RNA-Seq\n(TCGA)'
x <- t2

x$prop <- 100*x$prop
cols <- c('#d2e8f0','#3f99bc')
names(cols) <- c('Yes','No')
p1 <- ggplot(x,aes(x=group,y=prop)) +
    scale_y_continuous(expand=c(0,0)) +
    geom_bar(stat='identity',aes(fill=value)) +
    scale_fill_manual(values=cols,name='Mutations\nvalidated:') +
    geom_text(aes(label=label),y=50,angle=90,size=3.5) +
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),legend.position='right') + 
    labs(x=NULL,y='% of mtDNA variants')
p1
```

## EDF 3d
```{r panel_edf3d, fig.width=5, fig.height=4, eval=T, echo=F} 

d <- copy(tcga_maf_data)
d <- d[t_depth >= 30 & t_depth_rna >= 30,]
d <- d[truncating==T & class=='unknown-truncating',]
dd <- d[,c('TumorVAF','TumorVAF_rna'),with=F]
dd$Heteroplasmy_DNA <- dd$TumorVAF * 100
dd$Heteroplasmy_RNA <- dd$TumorVAF_rna * 100
pearson_r <- cor(dd$Heteroplasmy_DNA,dd$Heteroplasmy_RNA,method='pearson')
statistic <- function(r,n) {
    t <- (r * sqrt(n-2))/sqrt(1-r^2)
    pval <- 2*pt(t,df=n-2,lower.tail=F)  
    list(pval=pval, statistic=t)
}

p2 <- ggplot(dd,aes(x=Heteroplasmy_DNA,y=Heteroplasmy_RNA)) +
    scale_x_continuous(limits=c(0,100),breaks=seq(0,100,25)) +
    scale_y_continuous(limits=c(0,100),breaks=seq(0,100,25)) +
    geom_point(alpha=0.3,color='steelblue',pch=19) +
    geom_abline(slope=1,intercept=0,color='black') +
    theme_std(base_size=12) +
    geom_text(data=dd[1,],label=paste('r =',round(pearson_r,3)),aes(x=80,y=7.5),size=3) +
    labs(x='% DNA heteroplasmy',y='% RNA heteroplasmy')
p2
```


## EDF 3e
```{r panel_EDF3e, fig.width=9, fig.height=4, eval=T} 

panel_EDF3e
if(get_pdfs) ggsave(here('figures/panel_EDF3e.pdf'),width=8,height=4)

```


## EDF 3f
```{r panel_EDF3f, fig.width=5.5, fig.height=2.2, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# compare the average TMB between MSI/MSS, 
# among non-cancergenes, cancer genes, and mtDNA
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

clin <- copy(tcga_clin_data)
dat <- clin[,c('PATIENT_ID','maintype','capture_area_tumornormal'),with=F]
x <- copy(pangi_clin_data)
dat <- merge(dat, x, by='PATIENT_ID', all.x=T)
dat <- dat[project %in% c('COAD','READ','STAD')]
dat <- dat[!is.na(msi_status),]
dat$nuclear_capture_area <- covered_region

d_mt <- copy(tcga_maf_data)
tbl_mt <- table.freq(strtrim(d_mt$Tumor_Sample_Barcode,12))
d_nu <- copy(mc3_maf_data)
tbl_nu <- table.freq(strtrim(d_nu$Tumor_Sample_Barcode,12))

tbl <- merge(tbl_mt, tbl_nu, by='value', all=T)
names(tbl) <- c('PATIENT_ID','N_mt','N_cancergene')
tbl[is.na(tbl)] <- 0
dat <- merge(dat, tbl, by='PATIENT_ID', all.x=T)
dat[project %in% c('COAD','READ'),project:='Colorectal']
dat[project %in% c('STAD'),project:='Stomach']

summarize <- function(x) {
    x_cancer <- sum(x$N_cancergene,na.rm=T)
    x_mt <- sum(x$N_mt,na.rm=T)
    N <- nrow(x)
    covd_bps_mt <- sum(x$capture_area_tumornormal,na.rm=T)
    list(x_cancer=x_cancer,x_mt=x_mt,N=N,covd_bps_mt=covd_bps_mt)
}
x <- dat[,summarize(.SD),by=c('project','msi_status')]
x <- x[order(project,msi_status),]
x$covd_bps_cancer <- (impact468region)*x$N

ci_mt <- adt(epitools::pois.exact(x$x_mt, pt = x$covd_bps_mt/1e6, conf.level = 0.95))
x$tmb_mt <- ci_mt$rate
x$tmb_mt_lwr <- ci_mt$lower
x$tmb_mt_upr <- ci_mt$upper
ci_cancer <- adt(epitools::pois.exact(x$x_cancer, pt = x$covd_bps_cancer/1e6, conf.level = 0.95))
x$tmb_cancer <- ci_cancer$rate
x$tmb_cancer_lwr <- ci_cancer$lower
x$tmb_cancer_upr <- ci_cancer$upper

x_tmb <- melt(x[,c('project','msi_status','tmb_mt','tmb_cancer'),with=F],
              id.vars=c('project','msi_status'))
setnames(x_tmb,'value','tmb')
x_tmb_lwr <- melt(x[,c('project','msi_status','tmb_mt_lwr','tmb_cancer_lwr'),with=F],
              id.vars=c('project','msi_status'))
x_tmb_lwr$variable <- gsub('_lwr','',x_tmb_lwr$variable)
setnames(x_tmb_lwr,'value','lwr')
x_tmb_upr <- melt(x[,c('project','msi_status','tmb_mt_upr','tmb_cancer_upr'),with=F],
              id.vars=c('project','msi_status'))
x_tmb_upr$variable <- gsub('_upr','',x_tmb_upr$variable)
setnames(x_tmb_upr,'value','upr')

pd <- merge(x_tmb, x_tmb_lwr, by=c('project','msi_status','variable'), all.x=T)
pd <- merge(pd, x_tmb_upr, by=c('project','msi_status','variable'), all.x=T)
pd[variable=='tmb_cancer',Region:='Cancer genes']
pd[variable=='tmb_mt',Region:='mtDNA']
pd$Region <- factor(pd$Region, levels=c('Cancer genes','mtDNA'))
pd <- pd[msi_status %nin% 'MSI-L']
pd$msi_status <- factor(pd$msi_status, levels=c('MSS','MSI-H'))
pd$group <- paste(pd$msi_status,pd$Region)
cols <- c('#d2e8f0','#3f99bc')
names(cols) <- c('Cancer genes','mtDNA')

panel_EDF3f <- ggplot(pd,aes(x=msi_status,y=tmb,group=Region)) +
    scale_y_continuous(expand=c(0,0)) + #breaks=seq(0,100,by=25)) +
    geom_bar(stat='identity',aes(fill=Region),position=position_dodge(width=1)) +
    scale_fill_manual(values=cols,name='Region') + 
    geom_errorbar(aes(min=lwr,max=upr),position=position_dodge(width=1),width=NA) +
    facet_wrap(facets=~project,ncol=2) + 
    theme_std(base_size=16) +
    labs(x='MSI-Status',y='Mutations/Mb')
panel_EDF3f
if(get_pdfs) ggsave(here('figures/panel_EDF3f.pdf'),width=5.5,height=2.2)

```


## EDF 4a
```{r panel_EDF4a, fig.width=4, fig.height=4, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# is there any correlation between Age and mtDNA mutations?

tmp <- demographics_info[,c('PATIENT_ID','Tumor_Sample_Barcode','maintype','capture_area_tumornormal','coverage_tumornormal','age_at_initial_pathologic_diagnosis'),with=F]
d <- copy(tcga_maf_data)
d <- d[class=='somatic',]
tbl <- table.freq(d$Tumor_Sample_Barcode)
tmp <- merge(tmp, tbl, by.x='Tumor_Sample_Barcode', by.y='value', all.x=T)
tmp[is.na(N),N:=0]
tmp$age_bin <- cut(tmp$age_at_initial_pathologic_diagnosis,c(0,20,35,45,55,65,75,85,130),include.lowest=T)
setnames(tmp,'age_at_initial_pathologic_diagnosis','age')
setnames(tmp,'capture_area_tumornormal','covd_bps')
tmp <- tmp[!is.na(age),]
tmp <- tmp[coverage_tumornormal >= 0.5]
tmp[maintype=='Non-Small Cell Lung Cancer',maintype:='NSC Lung']

collapse_type <- function(tmp) {
    n <- nrow(tmp)
    total_mutations <- sum(tmp$N)
    total_covd_bps <- sum(tmp$covd_bps)
    list(total_mutations=total_mutations, total_covd_bps=total_covd_bps, n=n)
}
res <- tmp[,collapse_type(.SD),by=c('age_bin','maintype')]
res <- res[n >= 3,]
res$tmb <- res$total_mutations / (res$total_covd_bps / 1e6)
res <- res[order(maintype,age_bin),]

test_type <- function(res) {
    r <- cor(res$tmb, as.integer(res$age_bin),method='spearman')
    n <- nrow(res)
    p <- tryCatch({
       cor.test(res$tmb, as.integer(res$age_bin),method='spearman',alternative='two.sided')$p.value
    },error=function(e) {
        as.numeric(NA)
    })
    list(n=n,r=r,p=p)
}
res2 <- res[,test_type(.SD),by=maintype]
res2$q <- p.adjust(res2$p, method='BH')
res2[q < 0.1,label:='*']
res2[q < 0.01,label:='**']
res2[q < 0.001,label:='***']
res2 <- res2[order(r,decreasing=T),]
res2$maintype <- factor(res2$maintype, levels=res2$maintype)
res <- res[maintype %in% res2$maintype,]
res$maintype <- factor(res$maintype, levels=res2$maintype)
res <- dcast(res,age_bin ~ maintype, value.var='tmb')
res <- melt(res, id.var='age_bin')
names(res) <- c('age_bin','maintype','tmb')
p1 <- ggplot(res,aes(x=age_bin,y=maintype)) +
    scale_y_discrete(expand=c(0,0)) + scale_x_discrete(expand=c(0,0)) +
    geom_tile(aes(fill=tmb),color='black') +
    scale_fill_gradient(low='white',high='blue',na.value='darkgrey') +
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1),legend.position='bottom') +
    labs(x='Age category',y=NULL)
p1 <- extract_gglegend(p1)

p2 <- ggplot(res2,aes(x=maintype,y=1)) +
    scale_y_continuous(position='right') +
    geom_tile(aes(fill=r),color='black') +
    geom_text(aes(label=label)) +
    scale_fill_gradient2(low='blue',mid='white',high='#e41a1c',midpoint=0) +
    theme_std(base_size=12) +
    theme(
          axis.text.x=element_blank(),axis.ticks.x=element_blank(),axis.line.x=element_blank(),
          axis.text.y=element_blank(),axis.ticks.y=element_blank(),axis.line.y=element_blank(),
          legend.position='bottom') +
    labs(y='Spearman r',x=NULL) +
    coord_flip()
p2 <- extract_gglegend(p2)

p_plot <- plot_grid(p1$plot, p2$plot, align='h',rel_widths=c(6,1))
p_legend <- plot_grid(p1$legend, p2$legend, align='v')
panel_EDF4a <- plot_grid(p_plot, p_legend, ncol=1, rel_heights=c(3,1))
panel_EDF4a
if(get_pdfs) ggsave(here('figures/panel_EDF4a.pdf'),width=4,height=4)
```


## EDF 4b
```{r panel_EDF4b, fig.width=4, fig.height=4, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# is there any correlation between stage and mtDNA mutations?

tmp <- demographics_info[,c('PATIENT_ID','Tumor_Sample_Barcode','maintype','stage','capture_area_tumornormal','coverage_tumornormal'),with=F]
d <- copy(tcga_maf_data)
d <- d[class=='somatic',]
tbl <- table.freq(d$Tumor_Sample_Barcode)
tmp <- merge(tmp, tbl, by.x='Tumor_Sample_Barcode', by.y='value', all.x=T)
tmp[is.na(N),N:=0]
tmp <- tmp[coverage_tumornormal >= 0.5]
setnames(tmp,'capture_area_tumornormal','covd_bps')
tmp <- tmp[stage %in% c('I','II','III','IV'),]
tmp$stage <- factor(tmp$stage, levels=c('I','II','III','IV'))
tmp[maintype=='Non-Small Cell Lung Cancer',maintype:='NSC Lung']

collapse_type <- function(tmp) {
    n <- nrow(tmp)
    total_mutations <- sum(tmp$N)
    total_covd_bps <- sum(tmp$covd_bps)
    list(total_mutations=total_mutations, total_covd_bps=total_covd_bps, n=n)
}
res <- tmp[,collapse_type(.SD),by=c('stage','maintype')]
res <- res[n >= 3,]
res$tmb <- res$total_mutations / (res$total_covd_bps / 1e6)
res <- res[order(maintype,stage),]

test_type <- function(res) {
    r <- cor(res$tmb, as.integer(res$stage),method='spearman')
    n <- nrow(res)
    p <- tryCatch({
       cor.test(res$tmb, as.integer(res$stage),method='spearman',alternative='two.sided')$p.value
    },error=function(e) {
        as.numeric(NA)
    })
    list(n=n,r=r,p=p)
}
res2 <- res[,test_type(.SD),by=maintype]
res2[is.na(p),r:=NA]
res2$q <- p.adjust(res2$p, method='BH')
res2[q < 0.1,label:='*']
res2[q < 0.01,label:='**']
res2[q < 0.001,label:='***']
res2 <- res2[order(r,decreasing=T),]
res2$maintype <- factor(res2$maintype, levels=res2$maintype)
res <- res[maintype %in% res2$maintype,]
res$maintype <- factor(res$maintype, levels=res2$maintype)
res <- dcast(res,stage ~ maintype, value.var='tmb')
res <- melt(res, id.var='stage')
names(res) <- c('stage','maintype','tmb')
p1 <- ggplot(res,aes(x=stage,y=maintype)) +
    scale_y_discrete(expand=c(0,0)) + scale_x_discrete(expand=c(0,0)) +
    geom_tile(aes(fill=tmb),color='black') +
    scale_fill_gradient(low='white',high='blue',na.value='darkgrey') +
    theme_std(base_size=12) +
    theme(legend.position='bottom') +
    labs(x='Pathologic stage',y=NULL)
p1 <- extract_gglegend(p1)

p2 <- ggplot(res2,aes(x=maintype,y=1)) +
    scale_y_continuous(position='right') +
    geom_tile(aes(fill=r),color='black') +
    geom_text(aes(label=label)) +
    scale_fill_gradient2(low='blue',mid='white',high='#e41a1c',midpoint=0) +
    theme_std(base_size=12) +
    theme(
          axis.text.x=element_blank(),axis.ticks.x=element_blank(),axis.line.x=element_blank(),
          axis.text.y=element_blank(),axis.ticks.y=element_blank(),axis.line.y=element_blank(),
          legend.position='bottom') +
    labs(y='Spearman r',x=NULL) +
    coord_flip()
p2 <- extract_gglegend(p2)

p_plot <- plot_grid(p1$plot, p2$plot, align='h',rel_widths=c(6,1))
p_legend <- plot_grid(p1$legend, p2$legend, align='v')
panel_EDF4b <- plot_grid(p_plot, p_legend, ncol=1, rel_heights=c(3,1))
panel_EDF4b
if(get_pdfs) ggsave(here('figures/panel_EDF4b.pdf'),width=4,height=4)
```



## EDF 5a
```{r panel_EDF5a, echo=F, eval=T, fig.width=2, fig.height=4}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# compare truncating rate/complex when using 20+ alt reads
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

gene_length <- copy(tcga_sample_gene_length20_data)
clin <- copy(tcga_clin_data)
maf <- copy(tcga_maf_data)
maf <- maf[t_alt_count >= 20,]

## for each sample, get its effective length per gene
genes <- copy(chrm_annotated_data)
genes <- table.freq(genes$symbol)
names(genes) <- c('symbol','length')
result_tcga <- get_variant_rates_per_complex(gene_length, maf, clin, min.coverage=0.9)

## plot truncating rates
x <- combinatorial_poisson_test('complex','total_truncating', 'total_complex_length_mb', result_tcga,'upper_trunc', ygap=1)
f <- function(i,x) c(x$a[i],x$b[i])
comparisons <- lapply(1:nrow(x),f,x)
panel_EDF5a <- ggplot(result_tcga,aes(x=complex,y=rate_trunc)) +
    geom_bar(stat='identity',fill='black') +
    geom_errorbar(aes(min=lower_trunc,max=rate_trunc),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_trunc,max=upper_trunc),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x$ypos,annotations=x$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Complex',y='Truncating/Mb') +
    scale_y_continuous(expand=c(0,0), limits=c(0,25),breaks=seq(0,20,by=5)) 
panel_EDF5a
if(get_pdfs) ggsave(here('figures/panel_EDF5a.pdf'),width=2,height=4)
```


## EDF 5b
```{r panel_EDF5b, echo=F, eval=T, fig.height=4, fig.width=2}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# compare non-truncating rate/complex when using 20+ alt reads
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## plot nontruncating rates
x <- combinatorial_poisson_test('complex','total_nontruncating','total_complex_length_mb',
                                result_tcga,'upper_nontrunc',ygap=4)
f <- function(i,x) c(x$a[i],x$b[i])
comparisons <- lapply(1:nrow(x),f,x)
panel_EDF5b <- ggplot(result_tcga,aes(x=complex,y=rate_nontrunc)) +
    geom_bar(stat='identity',fill='black') +
    geom_errorbar(aes(min=lower_nontrunc,max=rate_nontrunc),width=NA,color='white',size=0.5) +
    geom_errorbar(aes(min=rate_nontrunc,max=upper_nontrunc),width=NA,color='black',size=0.5) +
    geom_signif(comparisons=comparisons,tip_length=0,y_pos=x$ypos,annotations=x$label,size=0.25,textsize=2.5) +
    theme_std(base_size=12) +
    labs(x='Non-truncating',y='Mutations/Mb') +
    scale_y_continuous(expand=c(0,0),limits=c(0,100),breaks=seq(0,80,by=20)) 
panel_EDF5b
if(get_pdfs) ggsave(here('figures/panel_EDF5b.pdf'),width=2,height=4)
```


## EDF 5c
```{r panel_EDF5c, echo=F, eval=T, fig.height=4.5, fig.width=5}

## load padded homopolymer regions 
homopolymer <- copy(homopolymer_data)
setnames(homopolymer,'homopolymer','id')
homopolymer$homopolymer <- paste0(homopolymer$start,':',homopolymer$end,',',homopolymer$id)
homopolymer$start <- homopolymer$start-1; homopolymer$end <- homopolymer$end+1 ## pad around the homopolymer by +/- 1 bp
homopolymer$Chromosome <- 'MT'
setnames(homopolymer,'start','homopolymer_start')
setnames(homopolymer,'end','homopolymer_end')
setkey(homopolymer,'Chromosome','homopolymer_start','homopolymer_end')
expand <- function(homopolymer) {
    pos <- seq(homopolymer$homopolymer_start,homopolymer$homopolymer_end)
    list(pos=pos)
}
homopolymer_expanded <- homopolymer[,expand(.SD),by=c('homopolymer','id','sequence')]
collapse <- function(d) {
    out <- d[1,]
    out$homopolymer_start <- min(d$pos)
    out$homopolymer_end <- max(d$pos)
    out$rep <- unique(d$id)
    out
}
collapsed_homopolymer <- homopolymer_expanded[,collapse(.SD),by=homopolymer]
collapsed_homopolymer[,c('pos','id'):=NULL]
collapsed_homopolymer$Chromosome <- 'MT'
setkey(collapsed_homopolymer,'Chromosome','homopolymer_start','homopolymer_end')
collapsed_homopolymer$homopolymer_width <- nchar(collapsed_homopolymer$sequence)
setnames(collapsed_homopolymer,'rep','homopolymer_element')

## for each sample in each homopolymer, get the number of basepairs where is it covered
regions <- copy(cov20_t)
ids <- regions$Tumor_Sample_Barcode
regions[,Tumor_Sample_Barcode:=NULL]
regions <- t(regions)
regions <- cbind(pos=rownames(regions),adt(regions))
names(regions)[2:ncol(regions)] <- ids
regions$pos <- suppressWarnings(as.integer(regions$pos))
region_annotations <- copy(chrm_annotated_data)
region_annotations[,strand:=NULL]
setnames(region_annotations,'pos','start')
region_annotations$end <- region_annotations$start + 1
region_annotations$chr <- 'MT'
regions <- merge(region_annotations, regions, by.x='start', by.y='pos', all.x=T)

setkey(regions,'chr','start','end')
regions_annotated <- foverlaps(regions, collapsed_homopolymer, type='any')
regions_annotated <- regions_annotated[!is.na(homopolymer)]
regions_annotated <- regions_annotated[start >= homopolymer_start & start <= homopolymer_end]
sample_fields <- grep('TCGA-',names(regions),value=T)
regions_annotated <- regions_annotated[,c('homopolymer','homopolymer_start','homopolymer_end','homopolymer_width','homopolymer_element',sample_fields),with=F]
collapse_homopolymer <- function(d) {
    tmp <- d[,(sample_fields),with=F]
    nucleotides_covered <- colSums(tmp)
    out <- as.list(nucleotides_covered)
    out
}
coverage <- regions_annotated[,collapse_homopolymer(.SD), by=c('homopolymer','homopolymer_start','homopolymer_end','homopolymer_element','homopolymer_width')]
coverage_melted_full <- melt(coverage,id.vars=c('homopolymer','homopolymer_start', 'homopolymer_end','homopolymer_element', 'homopolymer_width'))
setnames(coverage_melted_full,'variable','Tumor_Sample_Barcode')

## subset the sample/homopolymer combos for those where we had sufficient coverage
coverage_melted_full <- coverage_melted_full[value==homopolymer_width+2,] 
coverage_melted_full[,value:=NULL]

## load indels, merge with homopolymers (note some samples/mutations may
## arise in several homopolymers at once)
d_full <- copy(tcga_maf_data)
d_full <- d_full[Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del'),] 
setkey(d_full,'Chromosome','Start_Position','End_Position')
d_full <- foverlaps(d_full, collapsed_homopolymer, type='any')
d_full <- d_full[!is.na(homopolymer)]
d_full <- d_full[Start_Position >= homopolymer_start & End_Position <= homopolymer_end]

d_full <- d_full[t_alt_count >= 20,]
d_full <- d_full[,c('Tumor_Sample_Barcode','maintype','Start_Position','Reference_Allele','Tumor_Seq_Allele2','Variant_Classification','HGVSp_Short','homopolymer','ShortVariantID'),with=F]
d_full <- d_full[!is.na(homopolymer),]
clin <- copy(tcga_clin_data)
all_samples <- clin$Tumor_Sample_Barcode

run_for_n <- function(n, d_full, coverage_melted_full) { 
    message(n)
    if(n>0) {
        downsample <- sample(all_samples,n,replace=T)
        d <- d_full[Tumor_Sample_Barcode %in% downsample,]
        coverage_melted <- coverage_melted_full[Tumor_Sample_Barcode %in% downsample,]
    } else {
        d <- d_full
        coverage_melted <- coverage_melted_full
    }
    ## annotate the coverage per homopolymer with mutations
    coverage_melted <- merge(coverage_melted, d, by=c('Tumor_Sample_Barcode','homopolymer'), all.x=T)
    x <- coverage_melted[,c('Tumor_Sample_Barcode','ShortVariantID','homopolymer','homopolymer_element','homopolymer_width'),with=F]
    x$affected <- !is.na(coverage_melted$ShortVariantID)
    x$nuc_repeat <- x$homopolymer_element %in% c('A','C','T','G')

    ## prep data for various potential enrichment tests
    get_number_affected_per_homopolymer <- function(x) {
        mutant <- length(unique(x$Tumor_Sample_Barcode[x$affected==T]))
        wildtype <- length(unique(x$Tumor_Sample_Barcode[x$affected==F]))
        list(mutant=mutant,wildtype=wildtype)
    }
    xx <- x[,get_number_affected_per_homopolymer(.SD),by=c('homopolymer_element','homopolymer')]
    xx$covered <- xx$wildtype+xx$mutant 
    xx$prop <- xx$mutant/xx$covered
    xx <- merge(xx, homopolymer, by='homopolymer', all.x=T)
    setnames(xx,'width','homopolymer_element_width')
    xx$homopolymer_width <- nchar(xx$sequence)

    ## add gene annotations to data
    genes <- copy(chrm_annotated_data)
    collapse_gene <- function(genes) {
        gene_start <- min(genes$pos)
        gene_end <- max(genes$pos)
        list(gene_start=gene_start, gene_end=gene_end)
    }
    genes <- genes[,collapse_gene(.SD),by=c('symbol','strand')]
    genes$chr <- 'MT'
    setkey(genes,'chr','gene_start','gene_end')
    setkey(xx,'Chromosome','homopolymer_start','homopolymer_end')
    xx <- foverlaps(xx, genes, type='within')
    xx <- xx[grepl('MT-R',symbol)==F & grepl('MT-T',symbol)==F,] ## exclude RNAs from this
    Ns <- length(sample_fields)
    xx$prop_callable <- xx$covered/Ns

    ## run binomial test
    tmp <- xx
    total_homopolymer_indels <- sum(tmp$mutant)
    total_homopolymer_regions_length <- sum(tmp$homopolymer_width) ## this is the padded width
    tmp <- tmp[order(homopolymer_start)]
    test <- function(tmp) {
        cnt_affected <- tmp$mutant
        homopolymer_length <- tmp$homopolymer_width
        probability <- homopolymer_length/total_homopolymer_regions_length
        p <- tryCatch({
            binom.test(cnt_affected,total_homopolymer_indels,probability,alternative='greater')$p.value
        },error=function(e) {
            as.numeric(NA)
        })
        list(x=cnt_affected,probability=probability,p=p)
    }
    l <- tmp[,test(.SD),by=homopolymer]
    l$q <- p.adjust(l$p,method='BH')
    l <- l[order(p,decreasing=F),]
    tmp <- merge(tmp, l, by='homopolymer', all.x=T)
    tmp$sample_size <- n
    tmp
}

result <- run_for_n(0, d_full, coverage_melted_full)
setnames(result,'q','q_new')
original_result <- copy(hotspots_indel_data)
original_result <- original_result[,c('homopolymer','q.value'),with=F]
setnames(original_result,'q.value','q_orig')
result <- merge(result, original_result, by='homopolymer', all=T)
result$nlog10q_new <- -log10(result$q_new)
result$nlog10q_orig <- -log10(result$q_orig)

dat <- result[,c('homopolymer','nlog10q_orig','nlog10q_new'),with=F]
dat$Significant <- 'Not significant'
dat[nlog10q_orig > 2 & nlog10q_new <= 2,Significant:='only min 5 alt reads']
dat[nlog10q_orig <= 2 & nlog10q_new > 2,Significant:='only min 20 alt reads']
dat[nlog10q_orig > 2 & nlog10q_new > 2,Significant:='Both']
dat[Significant!='Not significant',label:=homopolymer]

cols <- c('#bfbfbf',brewer.pal(3,'Set1'))
names(cols) <- c('Not significant','only min 5 alt reads','only min 20 alt reads','Both')

panel_EDF5c <- ggplot(dat, aes(x=nlog10q_orig, y=nlog10q_new)) +
    geom_abline(intercept=0,slope=1,color='#bfbfbf',linetype=2) +
    geom_point(aes(color=Significant)) +
    geom_text_repel(aes(label=label,color=Significant)) +
    scale_color_manual(values=cols,name='Significant when min alt reads:') + 
    theme_std(base_size=12) +
    theme(legend.position='bottom') 
panel_EDF5c
if(get_pdfs) ggsave(here('figures/panel_EDF5c.pdf'),width=5,height=4.5)
```


## EDF 5d
```{r panel_EDF5d, fig.width=6, fig.height=4, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# barplot of homopolymer indel hotspots by cancer type
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## load homopolymer indel hotspot data
ms <- copy(hotspots_indel_data)
ms <- ms[order(Start_Position),]
ms$homopolymer <- paste(ms$Hugo_Symbol,ms$homopolymer)
hotspots <- ms$homopolymer[ms$q.value < 0.01]

d <- copy(tcga_maf_data)
d$homopolymer <- paste(d$Hugo_Symbol,d$homopolymer)
d$indel_hotspot <- d$homopolymer %in% hotspots & d$Variant_Classification %in% c('Frame_Shift_Del','Frame_Shift_Ins')
dd <- d[indel_hotspot==T,c('Tumor_Sample_Barcode','homopolymer'),with=F]

clin <- copy(tcga_clin_data)
clin <- clin[coverage_tumor >= 0.9]
clin <- merge(clin, dd, by='Tumor_Sample_Barcode', all.x=T)
summarize_sample <- function(clin) {
    homopolymer <- clin$homopolymer
    if(length(homopolymer) > 1) {
        homopolymer <- '2+'
    } else if(is.na(homopolymer)) {
        homopolymer <- 'None'
    }
    list(homopolymer=homopolymer)
}
info <- clin[,summarize_sample(.SD),by=c('maintype','Tumor_Sample_Barcode')]
m <- as.data.frame.matrix(xtabs(~maintype + homopolymer, data=info))
tmp <- data.table(maintype=rownames(m),prop_affected=1-(m$None/rowSums(m)),affected=rowSums(m)-m$None,overall=rowSums(m))
tmp <- tmp[order(prop_affected,decreasing=F)]
tmp$label <- paste0(tmp$affected,'/',tmp$overall)
tmp <- tmp[overall >= 20,]

m <- m/rowSums(m)
m <- cbind(maintype=rownames(m),adt(m))
m <- as.data.table(reshape2::melt(m,id.var='maintype'))
m$variable <- factor(m$variable, levels=c('None',hotspots,'2+'))
m$maintype <- factor(m$maintype,levels=tmp$maintype)
m <- m[!is.na(maintype),]

cols <- c('#bfbfbf',brewer.pal(6,'Accent'),'black')
names(cols) <- levels(m$variable)

panel_EDF5d <- ggplot(m, aes(x=maintype, y=value)) +
    scale_y_continuous(expand=c(0,0),limits=c(0,0.20),position='right',labels=scales::percent) + 
    geom_bar(stat='identity',aes(fill=variable)) +
    geom_text(data=tmp,aes(x=maintype,y=prop_affected,label=label),hjust=-0.2,size=3) +
    scale_fill_manual(values=cols,name='Homopolymer hotspot') +
    coord_flip() +
    theme_std(base_size=12) +
    labs(x=NULL,y='% samples affected')

panel_EDF5d
if(get_pdfs) ggsave(here('figures/panel_EDF5d.pdf'),width=6,height=4)
```



## EDF 5e
```{r panel_EDF5e, fig.width=4.7, fig.height=2.5, eval=T}

ms <- copy(hotspots_indel_data)
setnames(ms,'mutant','samples_mutated')
setnames(ms,'covered','samples_covered')
setnames(ms,'homopolymer_element','nt')
setnames(ms,'homopolymer_width','width')
ms <- ms[,c('Hugo_Symbol','nt','width','prop','q.value','homopolymer'),with=F]
ms$nlog10q <- -log10(ms$q.value)
## get prop of PCAWG samples with frameshift indels per hotspot

pcawg <- copy(pcawg_maf_data[TumorVAF >= 0.05 & Hugo_Symbol != 'Control_Region',])
excluded <- grep('TCGA-',pcawg$Tumor_Sample_Barcode)
pcawg <- pcawg[-excluded,]
pcawg <- pcawg[homopolymer != '' & Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del'),]
pcawg_clin <- copy(pcawg_clin_data)
pcawg_clin <- pcawg_clin[grepl('TCGA-',Tumor_Sample_Barcode)==F,]
N_pcawg <- length(unique(pcawg_clin$PATIENT_ID))
tbl <- table.freq(pcawg$homopolymer)
tbl$prop <- tbl$N / N_pcawg
ms <- merge(ms, tbl, by.x='homopolymer', by.y='value', all.x=T)
setnames(ms,'prop.x','prop_tcga')
setnames(ms,'prop.y','prop_pcawg')
panel_edf5e_data <- ms[order(prop_tcga,decreasing=T),c('Hugo_Symbol','homopolymer','prop_tcga','prop_pcawg','q.value'),with=F]
panel_edf5e_data[is.na(panel_edf5e_data)] <- 0
panel_edf5e_data$Significance <- 'Not significant'
panel_edf5e_data[q.value < 0.01,Significance:='Q<0.01']
panel_edf5e_data$label <- paste(panel_edf5e_data$Hugo_Symbol,panel_edf5e_data$homopolymer)
panel_edf5e_data[q.value >= 0.01,label:=NA]
r <- cor(panel_edf5e_data$prop_tcga, panel_edf5e_data$prop_pcawg, method='pearson')
r <- paste0('Pearson r = ',prettyNum(r,digits=2))

cols <- c('#bfbfbf','#e41a1c')
names(cols) <- c('Not significant','Q<0.01')
panel_EDF5e <- ggplot(panel_edf5e_data,aes(x=prop_tcga,y=prop_pcawg)) +
    scale_y_continuous(limits=c(0,0.015),labels=scales::percent) +
    scale_x_continuous(limits=c(0,0.015),labels=scales::percent) +
    geom_abline(intercept=0,slope=1,color='black') +
    geom_text_repel(aes(label=label,color=Significance),size=2) + 
    geom_text(data=panel_edf5e_data[1,], x=0.012, y=0.00075, aes(label=r),size=3, color='black') +
    geom_point(aes(color=Significance),pch=19,size=2) +
    scale_color_manual(values=cols) + 
    labs(x='% TCGA cases affected',y='% PCAWG cases affected\n(Excludes TCGA samples)') +
    theme_std(base_size=12) 
panel_EDF5e
if(get_pdfs) ggsave(here('figures/panel_EDF5e.pdf'),width=4.7,height=2.5)
```


## EDF 5f
```{r panel_EDF5f, fig.width=6, fig.height=5, eval=T, echo=F}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Compare length of homopolymeric tracts across complexes
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

d <- copy(hotspots_indel_data)
d[grep('MT-ND',Hugo_Symbol),Complex:='CI']
d[grep('MT-CY',Hugo_Symbol),Complex:='CIII']
d[grep('MT-CO',Hugo_Symbol),Complex:='CIV']
d[grep('MT-ATP',Hugo_Symbol),Complex:='CV']
d[grep('MT-RNR',Hugo_Symbol),Complex:='rRNA']
d[grep('MT-T',Hugo_Symbol),Complex:='tRNA']
d$Complex <- factor(d$Complex, levels=rev(c('CI','CIII','CIV','CV')))
d <- d[,c('Complex','homopolymer','homopolymer_width','covered','mutant','q.value'),with=F]
d$hotspot <- d$q.value < 0.01

summarize <- function(d) {
    total_covered <- sum(d$covered)
    total_mutant <- sum(d$mutant)
    prop <- total_mutant / total_covered
    n_loci <- nrow(d)
    list(total_covered=total_covered, total_mutant=total_mutant, prop=prop, n_loci=n_loci)
}
info <- d[,summarize(.SD),by=c('Complex','homopolymer_width')]
info$homopolymer_width <- factor(info$homopolymer_width, levels=8:5)
info2 <- dcast(info, Complex ~ homopolymer_width, value.var='prop')
info2 <- melt(info2, id.vars=c('Complex'))
info2[is.na(value),value:=0]
setnames(info2,c('variable','value'),c('homopolymer_width','prop'))
info2 <- merge(info2, info[,c('Complex','homopolymer_width','total_mutant','n_loci'),with=F],
               by=c('Complex','homopolymer_width'), all.x=T)
info2[is.na(total_mutant),total_mutant:=0]

tmp <- reshape2::melt(xtabs(~homopolymer_width + Complex, data=d))
tmp2 <- reshape2::melt(xtabs(~homopolymer_width + Complex, data=d[hotspot==T]))
tmp <- as.data.table(merge(tmp, tmp2, by=c('homopolymer_width','Complex'), all=T))
tmp[is.na(tmp)] <- 0
setnames(tmp,c('value.x','value.y'),c('loci','hotspots'))
tmp$label <- paste0(tmp$hotspots,'/',tmp$loci)
tmp[loci==0,label:='']
tmp$homopolymer_width <- factor(tmp$homopolymer_width, levels=8:5)
info2 <- merge(info2, tmp[,c('homopolymer_width','Complex','label'),with=F],by=c('Complex','homopolymer_width'),all.x=T)
info2$Complex <- factor(info2$Complex, levels=c('CI','CIII','CIV','CV'))
info2$rate <- info2$prop * 100

p1 <- ggplot(info2, aes(x=Complex,y=homopolymer_width)) +
    scale_x_discrete(expand=c(0,0)) + scale_y_discrete(expand=c(0,0)) +
    geom_tile(aes(fill=rate),color='black') +
    geom_text(aes(label=label)) +
    scale_fill_gradient(low='white',high='#1F78B4',name='Mutation rate') +
    theme_std(base_size=12) +
    theme(legend.position='right') +
    labs(y='Homopolymer width (bp)',x='Complex')
p1 <- extract_gglegend(p1)

cols <- brewer.pal(4,'Paired')
names(cols) <- c('CI','CIII','CIV','CV')

p2 <- ggplot(info,aes(x=homopolymer_width,y=n_loci)) +
    scale_y_continuous(expand=c(0,0)) +
    geom_bar(stat='identity',aes(fill=Complex)) +
    scale_fill_manual(values=cols,name='ETC Complex') +
    theme_std(base_size=12) +
    theme(axis.text.y=element_blank(),axis.ticks.y=element_blank()) +
    labs(x=NULL,y='N homopolymers') +
    coord_flip()
p2 <- extract_gglegend(p2)

p <- plot_grid(p1$plot, p2$plot, align='h', rel_widths=c(1,1))
p.legend <- plot_grid(p1$legend, p2$legend, align='h', rel_widths=c(1,1))
panel_EDF5f <- plot_grid(p, p.legend, align='v', ncol=1, rel_heights=c(3,2))
panel_EDF5f
if(get_pdfs) ggsave(here('figures/panel_EDF5f.pdf'),width=6,height=5)
```


## EDF 5g
```{r panel_EDF5g, fig.width=4, fig.height=2.5, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# fraction of truncating arising at homopolymer hotspots in helix vs tumor
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

hotspots <- copy(hotspots_indel_data)
hotspot_homopolymers <- hotspots$homopolymer[hotspots$q.value < 0.01]

dt <- copy(helix_maf_data)
dt[Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del') & homopolymer!='',ShortVariantID:=homopolymer]
dt$hotspot <- dt$homopolymer %in% hotspot_homopolymers
dt <- dt[Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del','Nonsense_Mutation')]
n <- sum(dt$TOTAL_COUNT)
x <- sum(dt$TOTAL_COUNT[dt$hotspot==T])
ci_helix <- binom.confint(x,n,method='exact')
ci_helix <- adt(ci_helix)
ci_helix$cohort <- 'HelixMTdb'

dt <- copy(tcga_maf_data)
dt[Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del') & homopolymer!='',ShortVariantID:=homopolymer]
dt$hotspot <- dt$homopolymer %in% hotspot_homopolymers
dt <- dt[Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del','Nonsense_Mutation')]
n <- nrow(dt)
x <- sum(dt$hotspot==T)
ci_tcga <- binom.confint(x,n,method='exact')
ci_tcga <- adt(ci_tcga)
ci_tcga$cohort <- 'TCGA tumors'
ci <- rbind(ci_helix, ci_tcga)
ci$pct <- 100*ci$mean
ci$lwr <- 100*ci$lower
ci$upr <- 100*ci$upper

pctinfo <- break_axis(ci$pct,maxlower=3,lowerticksize=0.5,minupper=20,upperticksize=10,ratio_lower_to_upper=0.9)
lwrinfo <- break_axis(ci$lwr,maxlower=3,lowerticksize=0.5,minupper=20,upperticksize=10,ratio_lower_to_upper=0.9)
uprinfo <- break_axis(ci$upr,maxlower=3,lowerticksize=0.5,minupper=20,upperticksize=10,ratio_lower_to_upper=0.9)
ci$newpct <- pctinfo$newy
ci$newlower <- lwrinfo$newy
ci$newupper <- uprinfo$newy

panel_EDF5g <- ggplot(ci,aes(x=cohort,y=newpct)) + 
    scale_y_continuous(expand=c(0,0),labels=pctinfo$labels,limits=c(pctinfo$limits[1],5),breaks=pctinfo$breaks,position='right') +
    geom_bar(stat='identity',fill='#3f99bc') +
    geom_errorbar(aes(min=newlower,max=newpct),width=NA,color='white') +
    geom_errorbar(aes(min=newpct,max=newupper),width=NA,color='black') +
    theme_std(base_size=12) +
    theme(legend.position='none') + 
    labs(x=NULL,y='% cases with truncating variant') +
    coord_flip()
panel_EDF5g
if(get_pdfs) ggsave(here('figures/panel_EDF5g.pdf'),width=4,height=2.5)
```

## EDF 6a
```{r panel_EDF6a, echo=F, eval=T,fig.width=6,fig.height=5} 

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# compare indel hotspots between IMPACT and TCGA
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## load padded homopolymer regions
homopolymer <- copy(homopolymer_data)
setnames(homopolymer,'homopolymer','id')
homopolymer$homopolymer <- paste0(homopolymer$start,':',homopolymer$end,',',homopolymer$id)
homopolymer$start <- homopolymer$start-1; homopolymer$end <- homopolymer$end+1 ## pad around the homopolymer by +/- 1 bp
homopolymer$Chromosome <- 'MT'
setnames(homopolymer,'start','homopolymer_start')
setnames(homopolymer,'end','homopolymer_end')
setkey(homopolymer,'Chromosome','homopolymer_start','homopolymer_end')
expand <- function(homopolymer) {
    pos <- seq(homopolymer$homopolymer_start,homopolymer$homopolymer_end)
    list(pos=pos)
}
homopolymer_expanded <- homopolymer[,expand(.SD),by=c('homopolymer','id','sequence')]

collapse <- function(d) {
    out <- d[1,]
    out$homopolymer_start <- min(d$pos)
    out$homopolymer_end <- max(d$pos)
    out$rep <- unique(d$id)
    out
}
collapsed_homopolymer <- homopolymer_expanded[,collapse(.SD),by=homopolymer]
collapsed_homopolymer[,c('pos','id'):=NULL]
collapsed_homopolymer$Chromosome <- 'MT'
setkey(collapsed_homopolymer,'Chromosome','homopolymer_start','homopolymer_end')
collapsed_homopolymer$homopolymer_width <- nchar(collapsed_homopolymer$sequence)
setnames(collapsed_homopolymer,'rep','homopolymer_element')

## for each sample in each homopolymer, get the number of basepairs where is it covered
regions <- copy(cov_t_impact)
regions <- regions[!duplicated(Tumor_Sample_Barcode),]
ids <- regions$Tumor_Sample_Barcode
regions[,Tumor_Sample_Barcode:=NULL]
regions <- t(regions)
regions <- cbind(pos=rownames(regions),as.data.table(regions))
names(regions)[2:ncol(regions)] <- ids
regions$pos <- suppressWarnings(as.integer(regions$pos))

region_annotations <- copy(chrm_annotated_data)
region_annotations[,strand:=NULL]
setnames(region_annotations,'pos','start')
region_annotations$end <- region_annotations$start + 1
region_annotations$chr <- 'MT'
regions <- merge(region_annotations, regions, by.x='start', by.y='pos', all.x=T)
setkey(regions,'chr','start','end')
regions_annotated <- foverlaps(regions, collapsed_homopolymer, type='any')
regions_annotated <- regions_annotated[!is.na(homopolymer)]
regions_annotated <- regions_annotated[start >= homopolymer_start & start <= homopolymer_end]

sample_fields <- grep('P-00',names(regions),value=T)
regions_annotated <- regions_annotated[,c('homopolymer','homopolymer_start','homopolymer_end','homopolymer_width','homopolymer_element',sample_fields),with=F]
collapse_homopolymer <- function(d) {
    tmp <- d[,(sample_fields),with=F]
    nucleotides_covered <- colSums(tmp)
    out <- as.list(nucleotides_covered)
    out
}
coverage <- regions_annotated[,collapse_homopolymer(.SD), by=c('homopolymer','homopolymer_start','homopolymer_end','homopolymer_element','homopolymer_width')]

coverage_melted_full <- melt(coverage,id.vars=c('homopolymer','homopolymer_start','homopolymer_end','homopolymer_element','homopolymer_width'))
setnames(coverage_melted_full,'variable','Tumor_Sample_Barcode')

## subset the sample/homopolymer combos for those where we had sufficient coverage
coverage_melted_full <- coverage_melted_full[value==homopolymer_width+2,]
coverage_melted_full[,value:=NULL]

## load indels, merge with homopolymers (note some samples/mutations may
## arise in several homopolymers at once)
d_full <- copy(impact_maf_data)
d_full <- d_full[Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del'),]
setkey(d_full,'Chromosome','Start_Position','End_Position')
d_full <- foverlaps(d_full, collapsed_homopolymer, type='any')
d_full <- d_full[!is.na(homopolymer)]
d_full <- d_full[Start_Position >= homopolymer_start & End_Position <= homopolymer_end]
d_full <- d_full[,c('Tumor_Sample_Barcode','maintype','Start_Position','Reference_Allele','Tumor_Seq_Allele2','Variant_Classification','HGVSp_Short','homopolymer','ShortVariantID'),with=F]
d_full <- d_full[!is.na(homopolymer),]
d <- d_full

## annotate the coverage per homopolymer with mutations
coverage_melted <- coverage_melted_full
coverage_melted <- merge(coverage_melted, d, by=c('Tumor_Sample_Barcode','homopolymer'), all.x=T)
x <- coverage_melted[,c('Tumor_Sample_Barcode','ShortVariantID','homopolymer','homopolymer_element','homopolymer_width'),with=F]
x$affected <- !is.na(coverage_melted$ShortVariantID)
x$nuc_repeat <- x$homopolymer_element %in% c('A','C','T','G')

## prep data for various potential enrichment tests
get_number_affected_per_homopolymer <- function(x) {
    mutant <- length(unique(x$Tumor_Sample_Barcode[x$affected==T]))
    wildtype <- length(unique(x$Tumor_Sample_Barcode[x$affected==F]))
    list(mutant=mutant,wildtype=wildtype)
}
xx <- x[,get_number_affected_per_homopolymer(.SD),by=c('homopolymer_element','homopolymer')]
xx$covered <- xx$wildtype+xx$mutant
xx$prop <- xx$mutant/xx$covered
xx <- merge(xx, homopolymer, by='homopolymer', all.x=T)
setnames(xx,'width','homopolymer_element_width')
xx$homopolymer_width <- nchar(xx$sequence)

## add gene annotations to data
genes <- copy(chrm_annotated_data)
collapse_gene <- function(genes) {
    gene_start <- min(genes$pos)
    gene_end <- max(genes$pos)
    list(gene_start=gene_start, gene_end=gene_end)
}
genes <- genes[,collapse_gene(.SD),by=c('symbol','strand')]
genes$chr <- 'MT'
setkey(genes,'chr','gene_start','gene_end')
setkey(xx,'Chromosome','homopolymer_start','homopolymer_end')
xx <- foverlaps(xx, genes, type='within')
xx <- xx[grepl('MT-R',symbol)==F & grepl('MT-T',symbol)==F,] ## exclude RNAs from this
Ns <- length(sample_fields)
xx$prop_callable <- xx$covered/Ns

## run binomial test
tmp <- xx
total_homopolymer_indels <- sum(tmp$mutant)
total_homopolymer_regions_length <- sum(tmp$homopolymer_width) ## this is the padded width
tmp <- tmp[order(homopolymer_start)]
test <- function(tmp) {
    cnt_affected <- tmp$mutant
    homopolymer_length <- tmp$homopolymer_width
    probability <- homopolymer_length/total_homopolymer_regions_length
    p <- tryCatch({
        binom.test(cnt_affected,total_homopolymer_indels,probability,alternative='greater')$p.value
    },error=function(e) {
        as.numeric(NA)
    })
    list(x=cnt_affected,probability=probability,p=p)
}
l <- tmp[,test(.SD),by=homopolymer]
l$q <- p.adjust(l$p,method='BH')
l <- l[order(p,decreasing=F),]
tmp <- merge(tmp, l, by='homopolymer', all.x=T)
tmp <- tmp[,c('homopolymer','p','q','symbol','homopolymer_width','homopolymer_element'),with=F]
tmp <- tmp[order(p,decreasing=F),]
# write.tsv(tmp,here('data/processed_data/hotspots_indel_impact.txt'))
setnames(tmp,'q','q_impact')
setnames(tmp,'symbol','Hugo_Symbol')
tmp <- tmp[,c('Hugo_Symbol','homopolymer','q_impact'),with=F]

## load TCGA hotspots for comparison
tmp_tcga <- copy(hotspots_indel_data)
tmp_tcga <- tmp_tcga[,c('Hugo_Symbol','homopolymer','q.value'),with=F]
setnames(tmp_tcga,'q.value','q_tcga')

result <- merge(tmp, tmp_tcga, by=c('Hugo_Symbol','homopolymer'), all=T)
result$nlog10q_impact <- -log10(result$q_impact)
result$nlog10q_tcga <- -log10(result$q_tcga)
result[q_tcga < 0.01 | q_impact < 0.01,label:=homopolymer]
result$significant <- 'Not significant'
result[q_tcga < 0.01 & (q_impact >= 0.01 | is.na(q_impact)),significant:='TCGA']
result[q_tcga >= 0.01 & q_impact < 0.01,significant:='MSK-IMPACT']
result[q_tcga < 0.01 & q_impact < 0.01,significant:='Both']
result$significant <- factor(result$significant, levels=c('Both','TCGA','MSK-IMPACT','Not significant'))

panel_EDF6a <- ggplot(result,aes(x=nlog10q_tcga,y=nlog10q_impact)) +
    geom_abline(slope=1,intercept=0,linetype=2,color='#bfbfbf') +
    geom_point(aes(color=significant)) +
    geom_text_repel(aes(label=label,color=significant)) +
    scale_color_brewer(palette='Dark2',name='Hotspot in') +
    scale_x_continuous(limits=c(0,80),breaks=seq(0,80,by=20)) +
    scale_y_continuous(limits=c(0,80),breaks=seq(0,80,by=20)) +
    theme_std(base_size=12) +
    labs(x='-log10(Q-value), TCGA',y='-log10(Q-value), IMPACT',subtitle='Homopolymer hotspots')
panel_EDF6a
if(get_pdfs) ggsave(here('figures/panel_EDF6a.pdf'),width=6,height=5)
```

## EDF 6b (i)
```{r panel_EDF6b_i, fig.width=9, fig.height=6, eval=T, echo=F}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# heatmap hotspot indels across cancer types
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

library(pheatmap)
d_tcga <- copy(tcga_maf_data)
d_tcga$cohort <- 'TCGA'
d_impact <- copy(impact_maf_data)
d_impact$cohort <- 'MSK-IMPACT'
setnames(d_impact,'microsatellite','homopolymer')
d <- rbind(d_tcga,d_impact,fill=T)
hotspots <- result$label[!is.na(result$label)]
d <- d[homopolymer %in% hotspots,]
#d <- d[homopolymer %in% c(CI_hotspots, CIV_hotspots),]
d <- d[Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del'),]
d <- d[,c('Tumor_Sample_Barcode','cohort','Hugo_Symbol','homopolymer','maintype'),with=F]
d[grepl('MT-ND',Hugo_Symbol),Complex:='CI']
d[grepl('MT-CO',Hugo_Symbol),Complex:='CIV']
d$id <- paste(d$Hugo_Symbol,d$homopolymer)
x <- as.data.frame.matrix(xtabs(~id + maintype, data=d))
x <- x / rowSums(x)
p1 <- pheatmap(x)
p1
```

## EDF 6b (ii)
```{r panel_EDF6b_ii, fig.width=5, fig.height=6, eval=T, echo=F}
hotspots <- p1$tree_row$labels[p1$tree_row$order]
tmp <- as.data.frame.matrix(xtabs(~id + cohort,data=d))
tmp <- cbind(homopolymer=rownames(tmp), as.data.table(tmp))
tmp <- melt(tmp,id.vars='homopolymer')
tmp$homopolymer <- factor(tmp$homopolymer, levels=rev(hotspots))
tmp$variable <- factor(tmp$variable, levels=c('MSK-IMPACT','TCGA'))
p2 <- ggplot(tmp,aes(x=homopolymer,y=value)) +
    scale_y_continuous(expand=c(0,0),position='right') +
    geom_bar(stat='identity',aes(fill=variable),color='black',size=0.25) +
    scale_fill_brewer(palette='Accent') +
    theme_std(base_size=12) +
    coord_flip() +
    labs(x=NULL,y='n cases')
p2
```

## EDF 7a
```{r panel_EDF7a, fig.width=6, fig.height=5, echo=F, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# what to say about pathogenicity of 
# observed somatic mutations vs never arising in tumors?
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

library(DescTools)
d <- copy(tcga_maf_data)
helix <- copy(helix_maf_data)
mitimpact <- copy(mitimpact_data)

## load the APOGEE data from MitImpact, annotate hotspots, check if they're predicted more pathogenic
mitimpact$ShortVariantID <- paste0(mitimpact$Ref,mitimpact$Start,mitimpact$Alt)
mitimpact$APOGEE_boost_mean_prob <- as.numeric(gsub(',','.',mitimpact$APOGEE_boost_mean_prob))
apogee <- mitimpact[,c('Gene_symbol','OXPHOS_complex','ShortVariantID','Start',
                       'APOGEE_boost_mean_prob',
                       'APOGEE_boost_consensus','ClinVar_Oct2017_CLINSIG'),with=F]
apogee <- apogee[order(Start)]
setnames(apogee,'APOGEE_boost_mean_prob','prob')
setnames(apogee,'APOGEE_boost_consensus','consensus')

## get frequency of observed somatic SNVs
d <- d[Variant_Type=='SNP',]
d <- d[Variant_Classification %in% nontruncating_classes,] ## make sure to exclude silent
variants_ever_in_germline <- sortunique(d$ShortVariantID[d$ShortVariantID %in% helix$ShortVariantID])
d[ShortVariantID %in% variants_ever_in_germline,class:='Observed in germline']
d[ShortVariantID %nin% variants_ever_in_germline,class:='Somatic only']
getN <- function(d) {
    N <- length(unique(d$PATIENT_ID))
    list(N=N)
}
info <- d[,getN(.SD),by=c('class','ShortVariantID')]
x <- merge(apogee, info, by='ShortVariantID', all.x=T)
x[is.na(N),N:=0]
x[N==0,class:='Never observed']
x$clinvar_na <- F
x$pathogenic <- F
x$likely_pathogenic <- F
x$likely_benign <- F
x$benign <- F
x$uncertain_significance <- F
x$other <- F

x[ClinVar_Oct2017_CLINSIG=='.',clinvar_na:=T]
x[grepl('likely pathogenic',tolower(ClinVar_Oct2017_CLINSIG)),likely_pathogenic:=T]
x[grepl('likely benign',tolower(ClinVar_Oct2017_CLINSIG)),likely_benign:=T]
x[grepl('uncertain significance',tolower(ClinVar_Oct2017_CLINSIG)),uncertain_significance:=T]
x[grepl('pathogenic',tolower(ClinVar_Oct2017_CLINSIG)) & likely_pathogenic==F,pathogenic:=T]
x[grepl('benign',tolower(ClinVar_Oct2017_CLINSIG)) & likely_benign==F,benign:=T]
x[grepl('other',tolower(ClinVar_Oct2017_CLINSIG)),other:=T]

x <- x[,c('ShortVariantID','class','pathogenic','likely_pathogenic','likely_benign','benign','uncertain_significance','other','clinvar_na'),with=F]
x <- x[clinvar_na==F,]
conflicts <- which(rowSums(x[,c(3:ncol(x)),with=F]) > 1)
x <- x[!conflicts,]
xm <- melt(x, id.vars=c('class','ShortVariantID'))
xm <- xm[class!='Observed in germline',]
xm <- xm[value==T,]
xm <- xm[variable!='other',]
xm$variable <- factor(xm$variable, levels=c('pathogenic','likely_pathogenic','uncertain_significance','likely_benign','benign'))
tbl <- table.freq(xm$class)
xm <- merge(xm, tbl, by.x='class', by.y='value', all.x=T)
xm$class <- factor(xm$class, levels=c('Observed in germline','Never observed','Somatic only'))
xm <- xm[order(class),]
xm$class <- paste0(xm$class,' (N=',xm$N,')')
xm$class <- factor(xm$class, levels=unique(xm$class))

count <- function(xm,prop=T) {
    tbl <- table(xm$variable)
    if(prop) tbl <- tbl/sum(tbl)
    out <- as.list(tbl)
    names(out) <- names(tbl)
    out
}
cnt <- xm[,count(.SD),by=class]
N <- xm[,count(.SD,prop=F),by=class]
classes <- N$class
N <- as.matrix(N[,c(2:ncol(N)),with=F])
rownames(N) <- classes
pval <- plabel(DescTools::CochranArmitageTest(N, alternative = c("two.sided"))$p.value)
cnt <- cnt[order(class),]
cnt <- reshape2::melt(cnt,id.vars='class')
cols <- brewer.pal(11,'RdBu')[c(2,4,6,8,10)]
names(cols) <- levels(cnt$variable)
cnt$value <- 100*cnt$value

panel_EDF7a <- ggplot(cnt,aes(x=class,y=value)) +
    scale_y_continuous(position='right',expand=c(0,0)) +
    geom_bar(stat='identity',aes(fill=variable)) +
    scale_fill_manual(values=cols, name='ClinVar CLINSIG') +
    coord_flip() +
    theme_std(base_size=12) +
    labs(x=NULL,y='Percentage') +
    theme(legend.position='right')
panel_EDF7a <- extract_gglegend(panel_EDF7a)
panel_EDF7a <- plot_grid(panel_EDF7a$plot, panel_EDF7a$legend, ncol=1, nrow=2)
panel_EDF7a
if(get_pdfs) ggsave(here('figures/panel_EDF7a.pdf'),width=6,height=5)

```

## EDF 7b
```{r, panel_EDF7b, fig.height=4, fig.width=9, eval=T, echo=F}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# validation of tRNA aligned position hotspots in PCAWG
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## find any positions in aligned trnas with positive/negative selection
## based on permutation tests

res <- copy(hotspots_trna_pcawg_data)
collapse_pos <- function(res) {
    qval <- res$q.value[1]
    total_mutated <- sum(res$samples_mutated)
    list(qval=qval, total_mutated=total_mutated)
}
info <- res[,collapse_pos(.SD),by=c('trna_region','trna_pos','index')]
info[trna_region=='',trna_region:='n/a']
info <- info[order(index)]
region_levels <- c(unique(info$trna_region[info$trna_region!='n/a']),'n/a')
region_cols <- c(brewer.pal(8,'Paired'),'black')
names(region_cols) <- region_levels
mylabels <- info$trna_pos
mylabels[2:(length(mylabels)-1)] <- ''
info$nlog10q <- -log10(info$qval)

scaling_factor <- 5
pd <- info[,c('trna_region','trna_pos','total_mutated','nlog10q'),with=F]
suppressWarnings(pd <- melt(pd, id.var=c('trna_region','trna_pos')))
pd[variable=='total_mutated',value:=-1*value]
pd$trna_pos <- factor(pd$trna_pos, levels=info$trna_pos)
pd[variable=='nlog10q',value:=scaling_factor*value]
pd[variable=='total_mutated',variable:=trna_region]
pd$variable <- factor(pd$variable, levels=c(region_levels,'nlog10q'))

cols <- c(brewer.pal(8,'Accent'),'white','black')
names(cols) <- c(region_levels,'nlog10q')
mybreaks <- seq(-15,15,by=5)
mylabels <- c(seq(-15,0,by=5),seq(5,15,by=5)/scaling_factor)
panel_EDF7b <- ggplot(pd,aes(x=trna_pos,value)) +
    geom_hline(yintercept=2*scaling_factor,color='red',linetype=2,size=0.5) +
    geom_hline(yintercept=0) + 
    scale_y_continuous(expand=c(0,0),breaks=mybreaks,labels=mylabels) +
    scale_x_discrete(labels=mylabels) + 
    scale_fill_manual(values=cols,name='tRNA region') + 
    geom_bar(stat='identity',aes(fill=variable),color='black') +
    theme_std(base_size=12) +
    labs(x='Aligned cloverleaf position',y='N cases          -log10(Q)') +
    theme( 
          axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          axis.line.x=element_blank(),legend.position='right')
panel_EDF7b
if(get_pdfs) ggsave(here('figures/panel_EDF7b.pdf'),width=9,height=4)

```





## EDF 8a
```{r panel_EDF8a, fig.width=10, fig.height=5, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# heatmap of cancer type vs gsea for truncating vs WT
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

x <- copy(rnaseq_truncating_gsea_data)
x <- x[,c('pathway','type','padj','NES'),with=F]
x$stat <- -log10(x$padj) * sign(x$NES)
x_trunc <- x
m <- x[,c('pathway','type','stat'),with=F]
summarize_signature <- function(m) {
    n_sig <- sum(abs(m$stat) >= 2)
    up_minus_down <- sum(m$stat >= 2) - sum(m$stat <= -2)
    list(up_minus_down=up_minus_down, n_sig=n_sig)
}
siginfo <- m[,summarize_signature(.SD), by=pathway]
siginfo <- siginfo[n_sig > 0,]
siginfo <- siginfo[order(up_minus_down,n_sig,decreasing=T),]
m <- m[pathway %in% siginfo$pathway]

m <- reshape(m, idvar='pathway', timevar='type', direction='wide')
pathways <- m$pathway
m[,pathway:=NULL]
m <- as.matrix(m)
rownames(m) <- pathways
colnames(m) <- gsub('stat[.]','',colnames(m))
m <- t(m)
m <- m[,siginfo$pathway]
colnames(m) <- gsub('_',' ',colnames(m))


## get % of samples with truncating vars per tumor type
x <- copy(rnaseq_truncating_histogram_data)
x <- x[order(N,decreasing=T),]
m <- m[x$type,]
x[,N:=NULL]
x <- as.data.table(reshape2::melt(x,id.var='type'))
x[variable=='wt',variable:='Wildtype']
x[variable=='affected',variable:='Truncating']
x$type <- factor(x$type, levels=rev(rownames(m)))
cols <- c('black','#bfbfbf')
names(cols) <- c('Truncating','Wildtype')
m2 <- adt(reshape2::melt(m))
names(m2) <- c('type','geneset','value')
m2$type <- factor(m2$type, levels=rev(rownames(m)))
m2$geneset <- factor(m2$geneset, levels=colnames(m))

p_bar <- ggplot(x,aes(x=type,y=value)) + 
    scale_y_continuous(expand=c(0,0),position='right') + 
    geom_bar(aes(fill=variable),stat='identity') +
    scale_fill_manual(values=cols,name='mtDNA Status') + 
    theme_std(base_size=12) +
    labs(x=NULL,y='Samples') +
    theme(axis.text.y=element_blank(),axis.ticks.y=element_blank(),axis.line.y=element_blank()) +
    coord_flip()
p_bar <- extract_gglegend(p_bar)

p_heatmap <- ggplot(m2, aes(x=geneset,y=type)) +
    geom_tile(aes(fill=value)) +
    scale_fill_gradient2(low='#4575b4',mid='white',high='#d73027',midpoint=0,name='-log10(Q)') +
    theme_std(base_size=12) +
    labs(x=NULL,y=NULL) +
    theme(
          axis.text.x=element_text(angle=90,hjust=1,vjust=0.5), 
          axis.ticks.x=element_blank(), 
          axis.ticks.y=element_blank(),
          axis.line.x=element_blank(), 
          axis.line.y=element_blank())
p_heatmap <- extract_gglegend(p_heatmap)

p_plot <- plot_grid(p_heatmap$plot, p_bar$plot, align='h', nrow=1, rel_widths=c(5,1))
p_legend <- plot_grid(p_heatmap$legend, p_bar$legend, ncol=1)
panel_EDF8a <- plot_grid(p_plot, p_legend, nrow=1, rel_widths=c(5,1))
panel_EDF8a
if(get_pdfs) ggsave(here('figures/panel_EDF8a.pdf'),width=10,height=5)

```


## EDF 8b

```{r panel_EDF8b, fig.width=10, fig.height=5, eval=T}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# panel: heatmap of cancer type vs gsea for truncating vs WT
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

x <- copy(rnaseq_vus_gsea_data)
x <- x[,c('pathway','type','padj','NES'),with=F]
x$stat <- -log10(x$padj) * sign(x$NES)
m <- x[,c('pathway','type','stat'),with=F]
x_vus <- x

summarize_signature <- function(m) {
    n_sig <- sum(abs(m$stat) >= 2)
    up_minus_down <- sum(m$stat >= 2) - sum(m$stat <= -2)
    list(up_minus_down=up_minus_down, n_sig=n_sig)
}
siginfo <- m[,summarize_signature(.SD), by=pathway]
siginfo <- siginfo[n_sig > 0,]
siginfo <- siginfo[order(up_minus_down,n_sig,decreasing=T),]
m <- m[pathway %in% siginfo$pathway]

m <- reshape(m, idvar='pathway', timevar='type', direction='wide')
pathways <- m$pathway
m[,pathway:=NULL]
m <- as.matrix(m)
rownames(m) <- pathways
colnames(m) <- gsub('stat[.]','',colnames(m))
m <- t(m)
m <- m[,siginfo$pathway]
colnames(m) <- gsub('_',' ',colnames(m))


## get % of samples with vus vars per tumor type
x <- copy(rnaseq_vus_histogram_data)
x <- x[order(N,decreasing=T),]
m <- m[x$type,]
x[,N:=NULL]
x <- adt(reshape2::melt(x,id.var='type'))
x[variable=='wt',variable:='Wildtype']
x[variable=='affected',variable:='VUS']
x$type <- factor(x$type, levels=rev(rownames(m)))
cols <- c('black','#bfbfbf')
names(cols) <- c('VUS','Wildtype')
m2 <- adt(reshape2::melt(m))
names(m2) <- c('type','geneset','value')
m2$type <- factor(m2$type, levels=rev(rownames(m)))
m2$geneset <- factor(m2$geneset, levels=colnames(m))

p_bar <- ggplot(x,aes(x=type,y=value)) + 
    scale_y_continuous(expand=c(0,0),limits=c(0,600),breaks=c(0,200,400,600),position='right') + 
    geom_bar(aes(fill=variable),stat='identity') +
    scale_fill_manual(values=cols,name='mtDNA Status') + 
    theme_std(base_size=12) +
    labs(x=NULL,y='Samples') +
    theme(axis.text.y=element_blank(),axis.ticks.y=element_blank(),axis.line.y=element_blank()) +
    coord_flip()
p_bar <- extract_gglegend(p_bar)

p_heatmap <- ggplot(m2, aes(x=geneset,y=type)) +
    geom_tile(aes(fill=value)) +
    scale_fill_gradient2(low='#4575b4',mid='white',high='#d73027',midpoint=0,name='-log10(Q)') +
    theme_std(base_size=12) +
    labs(x=NULL,y=NULL) +
    theme(
          axis.text.x=element_text(angle=90,hjust=1,vjust=0.5), 
          axis.ticks.x=element_blank(), 
          axis.ticks.y=element_blank(),
          axis.line.x=element_blank(), 
          axis.line.y=element_blank())
p_heatmap <- extract_gglegend(p_heatmap)

p_plot <- plot_grid(p_heatmap$plot, p_bar$plot, align='h', nrow=1, rel_widths=c(5,1))
p_legend <- plot_grid(p_heatmap$legend, p_bar$legend, ncol=1)
panel_EDF8b <- plot_grid(p_plot, p_legend, nrow=1, rel_widths=c(5,1))
panel_EDF8b
if(get_pdfs) ggsave(here('figures/panel_EDF8b.pdf'),width=10,height=5)

```


## EDF 8c

```{r panel_EDF8c, fig.width=6.5, fig.height=3}

cms <- copy(tcga_cms)
cms <- cms[,c('sample','CMS_final_network_plus_RFclassifier_in_nonconsensus_samples'),with=F]
names(cms) <- c('PATIENT_ID','cms_type')
cms <- cms[cms_type %nin% c('UNK','NOLBL'),]

info <- copy(tcga_sample_hotspot_status_data)
info$PATIENT_ID <- strtrim(info$Tumor_Sample_Barcode,12)
info <- merge(info, cms, by='PATIENT_ID', all.x=T)
info <- info[!is.na(cms_type),]
info[,maintype:=NULL]
setnames(info,'cms_type','maintype')
setnames(info,'annotation_prefilter','label')
info <- info[label!='Unknown']
info <- info[,c('PATIENT_ID','label','maintype'),with=F]

get_barplot <- function(l,tcga=T) {
    cols <- c('#bfbfbf','#ea948d','#d2e7ef')
    names(cols) <- c('Wildtype','Truncating','VUS')

    l <- info
    xo <- xtabs(~maintype + label, data=l[label!='Unknown'])
    xo <- adt(reshape2::melt(xo))
    xo$label <- factor(xo$label, levels=rev(c('Wildtype','VUS','Truncating')))

    getci <- function(xo) {
        xo <- xo[order(label,decreasing=T),]
        ci <- as.data.frame(binom.confint(xo$value,sum(xo$value),method='exact')) 
        xo$total <- sum(xo$value)
        xo$prop <- 100*ci$mean
        xo$propminus1 <- cumsum(xo$prop)-xo$prop
        xo$lwr <- xo$propminus1 + 100*ci$lower
        xo$upr <- xo$propminus1 + 100*ci$upper
        xo
    }
    xo <- xo[,getci(.SD),by=maintype]
    xo$maintype <- factor(xo$maintype, levels=c('CMS4','CMS3','CMS2','CMS1'))

    p1 <- ggplot(xo,aes(x=maintype,y=prop,group=label)) +
        scale_y_continuous(expand=c(0,0),position='right',breaks=seq(0,100,by=25)) +
        geom_bar(stat='identity',aes(fill=label)) + 
        geom_errorbar(aes(min=lwr,max=upr),width=NA) +
        scale_fill_manual(values=cols,name=NULL) + 
        coord_flip() +
        theme_std(base_size=12) +
        theme(legend.position='bottom') +
        labs(x=NULL,y='Percentage of samples')
   p1 <- extract_gglegend(p1)

    p2 <- ggplot(xo[!duplicated(maintype)],aes(x=maintype,y=total)) +
        scale_y_continuous(expand=c(0,0),position='right') + 
        geom_bar(stat='identity') +
        coord_flip() +
        theme_std(base_size=12) +
        theme(axis.text.y=element_blank(),axis.ticks.y=element_blank()) +
        labs(x=NULL,y='N Samples')

    p <- plot_grid(p1$plot, p2, align='h',rel_widths=c(4,1))
    p <- plot_grid(p, p1$legend, ncol=1,rel_heights=c(4,1))
    list(plot=p, data=xo)
}

res <- get_barplot(info)
panel_EDF8c <- res$plot
tmp <- res$data
m <- reshape(tmp[,c('maintype','label','value'),with=F],idvar='maintype',timevar='label',direction='wide')
types <- m$maintype
m <- as.matrix(m[,(2:ncol(m)),with=F])
rownames(m) <- types
pval <- chisq.test(m)

panel_EDF8c
if(get_pdfs) ggsave(here('figures/panel_EDF8c.pdf'),width=6.5,height=3)

```


## EDF 9a
```{r panel_EDF9a, fig.height=7,fig.width=8, eval=T, echo=F}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# forest plot with IMPACT data
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## load/format nuclear driver mutations
muts <- copy(impact_maf_nuclear_data)
muts$important <- muts$oncogenic %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
    muts$`is-a-hotspot`=='Y' |
    (muts$Variant_Classification %in% c('Nonsense_Mutation','Frame_Shift_Del','Frame_Shift_Ins','Splice_Site') & muts$Hugo_Symbol %in% c('APC','TP53','SMAD4'))
muts <- muts[important==T,c('Tumor_Sample_Barcode','Hugo_Symbol'),with=F]
collapse <- function(muts) {
    RAS <- any(muts$Hugo_Symbol %in% c('KRAS','HRAS','NRAS'))
    RAF <- any(muts$Hugo_Symbol %in% c('BRAF'))
    APC <- any(muts$Hugo_Symbol %in% c('APC'))
    SMAD4 <- any(muts$Hugo_Symbol %in% c('SMAD4'))
    TP53 <- any(muts$Hugo_Symbol %in% c('TP53'))
    list(RAS=RAS, RAF=RAF, APC=APC, SMAD4=SMAD4, TP53=TP53)
}
muts <- muts[,collapse(.SD),by=Tumor_Sample_Barcode]

## load hotspot/sample annotations
tmp_clin <- copy(patient_clin_data_impact)
tmp_clin <- tmp_clin[,c('PATIENT_ID','OS_MONTHS','OS_STATUS'),with=F]

## load clinical data from MSK CRC 2017
clin_sample <- copy(crc_msk_2017_sample_clin_data)
clin_pt <- copy(crc_msk_2017_patient_clin_data)

clin <- merge(clin_sample, clin_pt, by='PATIENT_ID', all.x=T)
clin <- merge(clin, muts, by.x='SAMPLE_ID', by.y='Tumor_Sample_Barcode', all.x=T)
clin[is.na(RAS),RAS:=F]
clin[is.na(RAF),RAF:=F]
clin[is.na(APC),APC:=F]
clin[is.na(SMAD4),SMAD4:=F]
clin[is.na(TP53),TP53:=F]

## remove Stage IV
clin <- clin[STAGE_AT_DIAGNOSIS!='IV']

## annotate outcome
setnames(clin,c('OS_STATUS','OS_MONTHS'),c('OS_STATUS_2017','OS_MONTHS_2017'))
clin <- merge(clin, tmp_clin, by='PATIENT_ID', all.x=T)
clin$OS_MONTHS <- as.numeric(clin$OS_MONTHS)
clin[OS_STATUS=='DECEASED',OS_STATUS:='1']
clin[OS_STATUS=='LIVING',OS_STATUS:='0']
clin$OS_STATUS <- as.integer(clin$OS_STATUS)
clin <- clin[!is.na(OS_MONTHS),]

# add other covariates
setnames(clin,'PRIMARY_TUMOR_LOCATION','site')
clin[PRIMARY_SITE=='Rectum',site:='Rectum']
clin <- clin[!is.na(site),]
clin$msi_status <- as.character(NA)
clin[MSI_SCORE < 3,msi_status:='MSS']
clin[MSI_SCORE >= 3 & MSI_SCORE < 10,msi_status:='MSI-L']
clin[MSI_SCORE >= 10,msi_status:='MSI-H']
clin$msi_status <- factor(clin$msi_status, levels=c('MSS','MSI-L','MSI-H'))
clin$msi_status2 <- clin$msi_status
clin[msi_status2 %in% c('MSI-L','MSI-H'),msi_status2:='MSI-L/H']
clin$msi_status2 <- factor(clin$msi_status2, levels=c('MSS','MSI-L/H'))
clin <- clin[!is.na(msi_status),]

purity <- copy(impact_colorectal_purity) 
clin <- merge(clin, purity, by.x='SAMPLE_ID', by.y='Tumor_Sample_Barcode', all.x=T)
clin$purity_impact <- clin$purity_facets
clin <- clin[!is.na(purity_impact),]
clin$STAGE_AT_DIAGNOSIS <- factor(clin$STAGE_AT_DIAGNOSIS, levels=c('III','II','I'))

## merge mtstatus into clin data
info <- copy(impact_sample_hotspot_status_data)
setnames(info,'annotation_prefilter','mtstatus_prefilter')
setnames(info,'annotation_postfilter','mtstatus')
info <- info[maintype=='Colorectal Cancer']
x <- merge(clin, info, by.x='SAMPLE_ID', by.y='Tumor_Sample_Barcode', all.x=T)
x <- x[mtstatus!='Unknown',]

## reduce patients with multiple samples to single sample
## - If any samples have mtstatus %in% WT, VUS, Truncating,
##   randomly select one as representative sample
## - Otherwise, randomly select 1 from all
set.seed(42)
get_rep_sample <- function(info) {
    n <- nrow(info)
    if(n > 1) {
        random_choice <- sample(n)[1]
        out <- info[random_choice,]
    } else {
        out <- info
    }
    out
}
x <- x[,get_rep_sample(.SD),by=PATIENT_ID]
x$mtstatus2 <- x$mtstatus
x[mtstatus2 != 'Wildtype', mtstatus2:='Affected']
x$mtstatus2 <- factor(x$mtstatus2, levels=c('Wildtype','Affected'))
x$mtstatus <- factor(x$mtstatus, levels=c('Wildtype','VUS','Truncating'))

## 3-level
res.cox <- coxph(Surv(OS_MONTHS, OS_STATUS) ~ mtstatus + STAGE_AT_DIAGNOSIS +  AGE_AT_DIAGNOSIS +  site + msi_status + RAS + RAF + APC + SMAD4 + TP53 + SEX + purity_impact, data = x, iter.max=100)
out <- summary(res.cox)$coef
out <- cbind(covariate=rownames(out),as.data.table(out))
out <- cbind(out, as.data.table(exp(confint(res.cox))))
n_wt <- sum(x$mtstatus=='Wildtype')
n_vus <- sum(x$mtstatus=='VUS')
n_trunc <- sum(x$mtstatus=='Truncating')
n_age <- nrow(x)
n_stageI <- sum(x$STAGE_AT_DIAGNOSIS=='I')
n_stageII <- sum(x$STAGE_AT_DIAGNOSIS=='II')
n_stageIII <- sum(x$STAGE_AT_DIAGNOSIS=='III')
n_siteLeft <- sum(x$site=='Left',na.rm=T)
n_siteRectum <- sum(x$site=='Rectum',na.rm=T)
n_siteRight <- sum(x$site=='Right',na.rm=T)
n_RAS <- sum(x$RAS==T)
n_RAF <- sum(x$RAF==T)
n_APC <- sum(x$APC==T)
n_SMAD4 <- sum(x$SMAD4==T)
n_TP53 <- sum(x$TP53==T)
n_SEXFemale <- sum(x$SEX=='Female')
n_SEXMale <- sum(x$SEX=='Male')
n_msi_statusMSS <- sum(x$msi_status=='MSS')
n_msi_statusMSIl <- sum(x$msi_status=='MSI-L')
n_msi_statusMSIh <- sum(x$msi_status=='MSI-H')
n_purity <- sum(!is.na(x$purity_impact))

tmp <- list(mtstatusWT=n_wt, mtstatusVUS=n_vus, mtstatusTruncating=n_trunc,
            AGE_AT_DIAGNOSIS=n_age,
            STAGE_AT_DIAGNOSISIII=n_stageIII,
            STAGE_AT_DIAGNOSISII=n_stageII,
            STAGE_AT_DIAGNOSISI=n_stageI,
            siteLeft=n_siteLeft, siteRectum=n_siteRectum, siteRight=n_siteRight,
            RASTRUE=n_RAS, RAFTRUE=n_RAF, APCTRUE=n_APC, SMAD4TRUE=n_SMAD4, TP53TRUE=n_TP53,
            SEXFemale=n_SEXFemale, SEXMale=n_SEXMale,
            msi_statusMSS=n_msi_statusMSS,
            `msi_statusMSI-L`=n_msi_statusMSIl,
            `msi_statusMSI-H`=n_msi_statusMSIh,
            purity_impact=n_purity)
Ns <- data.table(covariate=names(tmp), N=unlist(tmp))
Ns$pos <- 1:nrow(Ns)

dat1 <- merge(Ns, out[,c(1,3,7,8,6),with=F], by='covariate', all.x=T)
dat1$covariate <- gsub(' ','_',dat1$covariate)
dat1$covariate <- gsub('-','_',dat1$covariate)
names(dat1) <- c('covariate','N','pos','HR','lwr','upr','p')
dat1[is.na(HR),covariate:=paste0(covariate,'_Reference')]
dat1[is.na(HR),HR:=1]
dat1 <- dat1[order(pos),]
dat1$covariate <- gsub('mtstatus','',dat1$covariate)
dat1$covariate <- gsub('AGE','Age',dat1$covariate)
dat1$covariate <- gsub('STAGE','',dat1$covariate)
dat1$covariate <- gsub('site','',dat1$covariate)
dat1$covariate <- gsub('TRUE','',dat1$covariate)
dat1$covariate <- gsub('SEX','',dat1$covariate)
dat1$covariate <- gsub('msi_status','',dat1$covariate)
dat1$covariate <- gsub('cms_type','',dat1$covariate)
dat1$covariate <- factor(dat1$covariate, levels=rev(dat1$covariate))
dat1[lwr > 1 | upr < 1,Significant:='Yes']
dat1[lwr < 1 & upr > 1,Significant:='No']
dat1[is.na(p),Significant:='NA (Reference)']
cols <- c('#bfbfbf','black','#4675b4')
names(cols) <- c('NA (Reference)','No','Yes')
dat1$logHR <- log(dat1$HR)
dat1$loglwr <- log(dat1$lwr)
dat1$logupr <- log(dat1$upr)
minval <- log(0.0001)
maxval <- log(10000)
dat1[logHR >= maxval,logHR:=maxval]
dat1[loglwr >= maxval,loglwr:=maxval]
dat1[logupr >= maxval,logupr:=maxval]
dat1[logHR <= minval,logHR:=minval]
dat1[loglwr <= minval,loglwr:=minval]
dat1[logupr <= minval,logupr:=minval]
dat1$HR <- exp(dat1$logHR)
dat1$lwr <- exp(dat1$loglwr)
dat1$upr <- exp(dat1$logupr)

## 2-level
res.cox <- coxph(Surv(OS_MONTHS, OS_STATUS) ~ mtstatus2 + STAGE_AT_DIAGNOSIS + AGE_AT_DIAGNOSIS +
                 site + msi_status + RAS + RAF + APC + SMAD4 + TP53 + SEX + purity_impact,
             data = x, iter.max=100)

out <- summary(res.cox)$coef
out <- cbind(covariate=rownames(out),as.data.table(out))
out <- cbind(out, as.data.table(exp(confint(res.cox))))
n_wt <- sum(x$mtstatus2=='Wildtype')
n_affected <- sum(x$mtstatus2=='Affected')
n_age <- nrow(x)
n_stageI <- sum(x$STAGE_AT_DIAGNOSIS=='I')
n_stageII <- sum(x$STAGE_AT_DIAGNOSIS=='II')
n_stageIII <- sum(x$STAGE_AT_DIAGNOSIS=='III')
n_siteLeft <- sum(x$site=='Left',na.rm=T)
n_siteRectum <- sum(x$site=='Rectum',na.rm=T)
n_siteRight <- sum(x$site=='Right',na.rm=T)
n_RAS <- sum(x$RAS==T)
n_RAF <- sum(x$RAF==T)
n_APC <- sum(x$APC==T)
n_SMAD4 <- sum(x$SMAD4==T)
n_TP53 <- sum(x$TP53==T)
n_SEXFemale <- sum(x$SEX=='Female')
n_SEXMale <- sum(x$SEX=='Male')
n_msi_statusMSS <- sum(x$msi_status=='MSS')
n_msi_statusMSIl <- sum(x$msi_status=='MSI-L')
n_msi_statusMSIh <- sum(x$msi_status=='MSI-H')
n_purity <- sum(!is.na(x$purity_impact))

tmp <- list(mtstatus2WT=n_wt, mtstatus2Affected=n_affected,
            AGE_AT_DIAGNOSIS=n_age,
            STAGE_AT_DIAGNOSISIII=n_stageIII,
            STAGE_AT_DIAGNOSISII=n_stageII,
            STAGE_AT_DIAGNOSISI=n_stageI,
            siteLeft=n_siteLeft, siteRectum=n_siteRectum, siteRight=n_siteRight,
            RASTRUE=n_RAS, RAFTRUE=n_RAF, APCTRUE=n_APC, SMAD4TRUE=n_SMAD4, TP53TRUE=n_TP53,
            SEXFemale=n_SEXFemale, SEXMale=n_SEXMale,
            msi_statusMSS=n_msi_statusMSS,
            `msi_statusMSI-L`=n_msi_statusMSIl,
            `msi_statusMSI-H`=n_msi_statusMSIh,
            purity_impact=n_purity)
Ns <- data.table(covariate=names(tmp), N=unlist(tmp))
Ns$pos <- 1:nrow(Ns)

dat2 <- merge(Ns, out[,c(1,3,7,8,6),with=F], by='covariate', all.x=T)
dat2$covariate <- gsub(' ','_',dat2$covariate)
dat2$covariate <- gsub('-','_',dat2$covariate)
names(dat2) <- c('covariate','N','pos','HR','lwr','upr','p')
dat2[is.na(HR),covariate:=paste0(covariate,'_Reference')]
dat2[is.na(HR),HR:=1]
dat2 <- dat2[order(pos),]
dat2$covariate <- gsub('mtstatus','',dat2$covariate)
dat2$covariate <- gsub('AGE','Age',dat2$covariate)
dat2$covariate <- gsub('STAGE','',dat2$covariate)
dat2$covariate <- gsub('site','',dat2$covariate)
dat2$covariate <- gsub('TRUE','',dat2$covariate)
dat2$covariate <- gsub('SEX','',dat2$covariate)
dat2$covariate <- gsub('msi_status','',dat2$covariate)
dat2$covariate <- gsub('cms_type','',dat2$covariate)
dat2$covariate <- factor(dat2$covariate, levels=rev(dat2$covariate))
dat2[lwr > 1 | upr < 1,Significant:='Yes']
dat2[lwr < 1 & upr > 1,Significant:='No']
dat2[is.na(p),Significant:='NA (Reference)']
cols <- c('#bfbfbf','black','#4675b4')
names(cols) <- c('NA (Reference)','No','Yes')
dat2$logHR <- log(dat2$HR)
dat2$loglwr <- log(dat2$lwr)
dat2$logupr <- log(dat2$upr)
minval <- log(0.0001)
maxval <- log(10000)
dat2[logHR >= maxval,logHR:=maxval]
dat2[loglwr >= maxval,loglwr:=maxval]
dat2[logupr >= maxval,logupr:=maxval]
dat2[logHR <= minval,logHR:=minval]
dat2[loglwr <= minval,loglwr:=minval]
dat2[logupr <= minval,logupr:=minval]
dat2$HR <- exp(dat2$logHR)
dat2$lwr <- exp(dat2$loglwr)
dat2$upr <- exp(dat2$logupr)
dat1$version <- '3-level'
dat2$version <- '2-level'
dat <- rbind(dat1, dat2)

panel_EDF9a <- ggplot(dat[version=='3-level'],aes(x=covariate, y=HR)) +
    geom_hline(yintercept=1) +
    geom_point(aes(size=N,fill=NA,color=Significant),pch=21) +
    geom_errorbar(aes(min=lwr,max=upr,color=Significant),width=0.333) +
    scale_color_manual(values=cols,name='Significant') +
    scale_fill_manual(values=cols,name='Significant') +
    coord_flip() +
    theme_std(base_size=12) +
    scale_y_continuous(trans=scales::log_trans(),labels=scales::comma,
                       breaks=c(0.01,0.1,1,10,100),limits=c(0.01,100),
                       position='right') +
    labs(x=NULL,y='Hazard ratio (overall survival)')
panel_EDF9a
if(get_pdfs) ggsave(here('figures/panel_EDF9a.pdf'),width=8,height=7)
```

## EDF 9b 
```{r panel_EDF9b, fig.height=7,fig.width=8, eval=T, echo=F}
panel_EDF9b <- ggplot(dat[version=='2-level'],aes(x=covariate, y=HR)) +
    geom_hline(yintercept=1) +
    geom_point(aes(size=N,fill=NA,color=Significant),pch=21) +
    geom_errorbar(aes(min=lwr,max=upr,color=Significant),width=0.333) +
    scale_color_manual(values=cols,name='Significant') +
    scale_fill_manual(values=cols,name='Significant') +
    coord_flip() +
    theme_std(base_size=12) +
    scale_y_continuous(trans=scales::log_trans(),labels=scales::comma,
                       breaks=c(0.01,0.1,1,10,100),limits=c(0.01,100),
                       position='right') +
    labs(x=NULL,y='Hazard ratio (overall survival)')
panel_EDF9b
if(get_pdfs) ggsave(here('figures/panel_EDF9b.pdf'),width=8,height=7)
```


## EDF 10a
```{r panel_EDF10a, echo=F, eval=T, fig.width=6, fig.height=6} 

pcawg <- copy(pcawg_maf_data)
pcawg$cohort <- 'PCAWG'
tcga <- copy(tcga_maf_data)
tcga$cohort <- 'TCGA'
d <- rbind(pcawg, tcga, fill=T)
d[Variant_Type %in% c('INS','DEL') & homopolymer!='',ShortVariantID:=homopolymer]
d[Variant_Type %in% c('INS','DEL') & homopolymer!='',Start_Position:=homopolymer_start]

d <- d[,c('Tumor_Sample_Barcode','ShortVariantID','cohort','Hugo_Symbol','Variant_Classification','TumorVAF','t_depth','t_alt_count','Start_Position'),with=F]
d$type <- 'Silent/Other'
d[Variant_Classification %in% c('Control_Region'),type:='Control Region']
d[Variant_Classification %in% c('rRNA','tRNA'),type:='tRNA/rRNA']
d[Variant_Classification %in% c('Frame_Shift_Ins','Frame_Shift_Del','Nonsense_Mutation'),type:='Truncating']
d[Variant_Classification %in% c('In_Frame_Ins','In_Frame_Del','Missense_Mutation'),type:='Missense']
d <- d[order(Start_Position,ShortVariantID),]
d$ShortVariantID <- factor(d$ShortVariantID, levels=unique(d$ShortVariantID))
d$cohort <- factor(d$cohort,levels=c('TCGA','PCAWG'))
types <- sort(unique(d$type))

cols <- c('#93c7ea','salmon')
names(cols) <- c('PCAWG','TCGA')
summarize_variant <- function(d) {
    n <- nrow(d)
    list(n=n)
}

## split by VAF
info1 <- d[TumorVAF >= 0.3,summarize_variant(.SD), by=c('cohort','type')]
toadd1 <- info1[type=='Control Region']
toadd1$cohort <- 'TCGA'; toadd1$n <- 0
info1 <- rbind(info1, toadd1)
info1$VAF <- 'VAF >= 30%'
info2 <- d[TumorVAF < 0.3,summarize_variant(.SD), by=c('cohort','type')]
toadd2 <- info2[type=='Control Region']
toadd2$cohort <- 'TCGA'; toadd2$n <- 0
info2 <- rbind(info2, toadd2)
info2$VAF <- 'VAF < 30%'
info <- rbind(info1, info2)
info$VAF <- factor(info$VAF, levels=c('VAF >= 30%','VAF < 30%'))

panel_EDF10a <- ggplot(info, aes(x=type,y=n,group=cohort)) +
    scale_y_continuous(expand=c(0,0)) + 
    geom_bar(stat='identity', aes(fill=cohort),position=position_dodge()) + 
    #geom_text(aes(label=n)) + 
    scale_fill_manual(values=cols,name='Cohort') + 
    theme_std(base_size=12) +
    facet_wrap(ncol=1,facets=~VAF,scale='free_y') +
    theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1)) +
    labs(x='Variant Class',y='N variants detected')
panel_EDF10a
if(get_pdfs) ggsave(here('figures/panel_EDF10a.pdf'),width=6,height=6)
```

## EDF 10b
```{r panel_EDF10b, eval=T, echo=F, fig.height=5, fig.width=6}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# compare rna-seq between <30% and >30%
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

low <- copy(rnaseq_truncating_gsea_lowvaf_data)
low[,genes:=NULL]
low$stat <- sign(low$NES) * -log10(low$padj)
low$vaf <- '<30%'

high <- copy(rnaseq_truncating_gsea_highvaf_data)
high[,genes:=NULL]
high$stat <- sign(high$NES) * -log10(high$padj)
high$vaf <- '>=30%'
vaf <- rbind(low[,c('type','pathway','vaf','stat'),with=F],high[,c('type','pathway','vaf','stat'),with=F])
vaf <- as.data.table(dcast(vaf, type + pathway ~ vaf, value.var='stat'))
vaf <- vaf[!is.na('<30%') & !is.na('>=30%'),]
vaf <- vaf[pathway %in% c('TNFA_signaling_via_NFKB','Oxidative_phosphorylation')]
vaf$pathway <- factor(vaf$pathway, levels=c('TNFA_signaling_via_NFKB','Oxidative_phosphorylation'))

vaf$significant <- 'Not significant'
vaf[abs(`<30%`) > 2 & abs(`>=30%`) <= 2,significant:='VAF < 30%']
vaf[abs(`<30%`) <= 2 & abs(`>=30%`) > 2,significant:='VAF >= 30%']
vaf[abs(`<30%`) > 2 & abs(`>=30%`) > 2,significant:='Any VAF']
cols <- c('gray','blue','red','purple')
names(cols) <- c('Not significant','VAF < 30%','VAF >= 30%','Any VAF')

panel_EDF10b<- ggplot(vaf,aes(x=`>=30%`,y=`<30%`)) +
    geom_hline(yintercept=0,linetype=2,size=0.5,color='#bfbfbf') +
    geom_vline(xintercept=0,linetype=2,size=0.5,color='#bfbfbf') +
    geom_point(size=2,aes(color=significant)) +
    scale_color_manual(values=cols, name='Significant when') + 
    facet_wrap(facets=~pathway,ncol=1) +
    theme_std(base_size=10) +
    theme(panel.grid.major=element_line(color='grey92',size=0.25)) +
    labs(x='Sign(NES) X -log10(Q), VAF >= 30%',y='Sign(NES) X -log10(Q), VAF < 30%')
panel_EDF10b
if(get_pdfs) ggsave(here('figures/panel_EDF10b.pdf'),width=4,height=5)

```


